<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>api</th>
      <th>api version</th>
      <th>organization name</th>
      <th>environment</th>
      <th>url</th>
      <th>updatedAt</th>
      <th>name</th>
      <th>id</th>
      <th>version</th>
      <th>description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>11</th>
      <td>drs</td>
      <td>1.1.0</td>
      <td>NCI Data Commons Framework Services</td>
      <td>None</td>
      <td>https://nci-crdc.datacommons.io/</td>
      <td>2020-07-24T12:00:00Z</td>
      <td>NCI DCFS</td>
      <td>gov.nih.nci.dcfs.drs</td>
      <td>2.12.0</td>
      <td>DRS server that follows the Gen3 implementation. Gen3 is a GA4GH compliant open source platform for developing framework services and data commons. Data commons accelerate and democratize the process of scientific discovery, especially over large or complex datasets. Gen3 is maintained by the Center for Translational Data Science at the University of Chicago. https://gen3.org</td>
    </tr>
    <tr>
      <th>4</th>
      <td>drs</td>
      <td>1.0.0</td>
      <td>Seven Bridges</td>
      <td>production</td>
      <td>https://cavatica-ga4gh-api.sbgenomics.com</td>
      <td>2020-09-10T00:41:57Z</td>
      <td>Cavatica DRS API</td>
      <td>com.sb.cavatica.drs</td>
      <td>1.0.0</td>
      <td>The Data Repository Service (DRS) API provides a generic interface to data repositories so data consumers, including workflow systems, can access data in a single, standard way regardless of where it's stored and how it's managed. The primary functionality of DRS is to map a logical ID to a means for physically retrieving the data represented by the ID.</td>
    </tr>
    <tr>
      <th>5</th>
      <td>drs</td>
      <td>1.0.0</td>
      <td>Seven Bridges</td>
      <td>production</td>
      <td>https://cgc-ga4gh-api.sbgenomics.com</td>
      <td>2020-09-10T00:41:57Z</td>
      <td>Cancer Genomics Cloud DRS API</td>
      <td>com.sb.cgc.drs</td>
      <td>1.0.0</td>
      <td>The Data Repository Service (DRS) API provides a generic interface to data repositories so data consumers, including workflow systems, can access data in a single, standard way regardless of where it's stored and how it's managed. The primary functionality of DRS is to map a logical ID to a means for physically retrieving the data represented by the ID.</td>
    </tr>
    <tr>
      <th>13</th>
      <td>drs</td>
      <td>1.1.0</td>
      <td>AnVIL</td>
      <td>None</td>
      <td>https://gen3.theanvil.io/</td>
      <td>2020-07-24T12:00:00Z</td>
      <td>AnVIL</td>
      <td>io.theanvil.drs</td>
      <td>2.12.0</td>
      <td>DRS server that follows the Gen3 implementation. Gen3 is a GA4GH compliant open source platform for developing framework services and data commons. Data commons accelerate and democratize the process of scientific discovery, especially over large or complex datasets. Gen3 is maintained by the Center for Translational Data Science at the University of Chicago. https://gen3.org</td>
    </tr>
    <tr>
      <th>8</th>
      <td>drs</td>
      <td>1.1.0</td>
      <td>Childrens Hospital of Philadelphia</td>
      <td>production</td>
      <td>https://data.kidsfirstdrc.org</td>
      <td>2020-10-06T13:11:38Z</td>
      <td>The Kids First Data Resource Center (KFDRC)</td>
      <td>edu.chop.kfdrc</td>
      <td>2020.08</td>
      <td>Full implementation of the DRS 1.1 standard with support for persistent identifiers. Open source DRS server that follows the Gen3 implementation. Gen3 is a GA4GH compliant open source platform for developing framework services and data commons. Data commons accelerate and democratize the process of scientific discovery, especially over large or complex datasets. Gen3 is maintained by the Center for Translational Data Science at the University of Chicago. https://gen3.org</td>
    </tr>
    <tr>
      <th>12</th>
      <td>drs</td>
      <td>1.1.0</td>
      <td>BioData Catalyst</td>
      <td>None</td>
      <td>https://gen3.biodatacatalyst.nhlbi.nih.gov/</td>
      <td>2020-07-24T12:00:00Z</td>
      <td>BioData Catalyst</td>
      <td>gov.nih.nhlbi.biodatacatalyst.drs</td>
      <td>2.12.0</td>
      <td>Full implementation of the DRS 1.1 standard with support for persistent identifiers. Open source DRS server that follows the Gen3 implementation. Gen3 is a GA4GH compliant open source platform for developing framework services and data commons. Data commons accelerate and democratize the process of scientific discovery, especially over large or complex datasets. Gen3 is maintained by the Center for Translational Data Science at the University of Chicago. https://gen3.org</td>
    </tr>
    <tr>
      <th>17</th>
      <td>htsget</td>
      <td>1.2.0</td>
      <td>Global Alliance for Genomics and Health</td>
      <td>production</td>
      <td>https://htsget.ga4gh.org/variants</td>
      <td>2020-09-04T14:40:00Z</td>
      <td>GA4GH htsget reference server variants API</td>
      <td>org.ga4gh.htsgetreference.variants</td>
      <td>1.3.0</td>
      <td>Reference web service of the GA4GH htsget protocol. Streams variant data from public datasets, such as Tabula Muris and 1000 Genomes.</td>
    </tr>
    <tr>
      <th>16</th>
      <td>htsget</td>
      <td>1.2.0</td>
      <td>Global Alliance for Genomics and Health</td>
      <td>production</td>
      <td>https://htsget.ga4gh.org/reads</td>
      <td>2020-09-04T14:40:00Z</td>
      <td>GA4GH htsget reference server reads API</td>
      <td>org.ga4gh.htsgetreference.reads</td>
      <td>1.3.0</td>
      <td>Reference web service of the GA4GH htsget protocol. Streams alignment data from public datasets, such as Tabula Muris and 1000 Genomes.</td>
    </tr>
    <tr>
      <th>21</th>
      <td>refget</td>
      <td>1.0.0</td>
      <td>European Bioinformatics Institute</td>
      <td>None</td>
      <td>https://refget-insdc.jeremy-codes.com</td>
      <td>None</td>
      <td>INSDC Refget Mirror</td>
      <td>uk.ac.ebi.insdc.refget</td>
      <td>0.3.0</td>
      <td>This web service serves genomic reference sequences from the International Nucleotide Sequence Database Collaboration (INSDC) according to the Reference Sequence Retrieval (Refget) API protocol. Refget is an open protocol developed and maintained by the Global Alliance for Genomics and Health (GA4GH). The Refget protocol prescribes the access of sequences by an unambiguous, vendor-neutral identifier that is derived from a hash digest of the raw sequence content. One or more hashing functions can be supported. The following checksum identifier formats are supported by this service: GA4GH Variant Representation sequence checksum, Refget protocol Trunc512, MD5</td>
    </tr>
    <tr>
      <th>20</th>
      <td>refget</td>
      <td>1.0.0</td>
      <td>European Bioinformatics Institute</td>
      <td>None</td>
      <td>https://www.ebi.ac.uk/ena/cram</td>
      <td>2020-08-07T13:19:49Z</td>
      <td>CRAM Reference Registry</td>
      <td>uk.ac.ebi.cram-reference-registry</td>
      <td>1.0.0</td>
      <td>ENA’s CRAM Reference Registry is the original refget deployment serving CRAM’s requirements for a centralised server for reference sequence retrieval. It served as the blueprint for the refget specification. All sequences held in the INSDC resources (ENA, Genbank, DDBJ) have been checksumed against the MD5 checksum, as noted in the CRAM specification. This server hosts over 2.5 billion sequences in an Oracle database. To reduce database and server load, the CRAM reference registry is hosted behind a Squid proxy server, to provide fast access to frequently accessed sequences.</td>
    </tr>
    <tr>
      <th>22</th>
      <td>refget</td>
      <td>1.0.0</td>
      <td>European Bioinformatics Institute</td>
      <td>None</td>
      <td>https://refget.herokuapp.com</td>
      <td>2020-08-07T13:19:49Z</td>
      <td>Refget reference implementation</td>
      <td>uk.ac.ebi.refget.ref-impl</td>
      <td>1.0.0</td>
      <td>Refget’s reference implementation is implemented in Perl and is updated with each approved change to the refget specification. The server stores metadata (e.g. sequence length, known aliases) within a RDBMS such as PostgreSQL, MySQL or SQLite through the DBIx::Class (https://metacpan.org/pod/DBIx::Class) object relational mapper, implements its application logic using the Mojolicious framework (https://metacpan.org/pod/Mojolicious) and features an extensive library of tests.</td>
    </tr>
    <tr>
      <th>23</th>
      <td>refget</td>
      <td>1.0.0</td>
      <td>European Bioinformatics Institute</td>
      <td>None</td>
      <td>https://spjb6ejone.execute-api.us-east-2.amazonaws.com/Prod</td>
      <td>2020-08-07T13:19:49Z</td>
      <td>Serverless refget on AWS</td>
      <td>uk.ac.ebi.refget-serverless</td>
      <td>1.0.0</td>
      <td>The refget-cloud codebase used in the AWS ECS Fargate deployment has been repurposed to work in a serverless context. In this case, a refget service has been created without any provisioning of servers, VMs, or container instances. Instead, the service consists of three AWS Lambda functions (one for each endpoint in the refget specification); each function is associated with an HTTPS request route via API Gateway. The serverless refget points to the same INSDC public S3 bucket. This service is best considered as a technical demonstration to complement the above containerized service; together they illustrate how refget can be deployed in multiple contexts using cloud resources.</td>
    </tr>
    <tr>
      <th>9</th>
      <td>rnaget</td>
      <td>1.0.0</td>
      <td>Centre for Genomic Regulation</td>
      <td>production</td>
      <td>https://genome.crg.cat/rnaget</td>
      <td>None</td>
      <td>CRG RNAget Reference Implementation</td>
      <td>eu.crg.rnaget</td>
      <td>1.0.0</td>
      <td>RNAget implementation serving compliance test dataset</td>
    </tr>
    <tr>
      <th>7</th>
      <td>rnaget</td>
      <td>1.0.0</td>
      <td>California Institute of Technology</td>
      <td>production</td>
      <td>https://felcat.caltech.edu/rnaget</td>
      <td>None</td>
      <td>Caltech RNAget Reference Implementation</td>
      <td>edu.caltech.rnaget</td>
      <td>1.0.0</td>
      <td>RNAget implementation serving compliance test dataset</td>
    </tr>
    <tr>
      <th>2</th>
      <td>search</td>
      <td>1.0.0</td>
      <td>DNAstack</td>
      <td>dev</td>
      <td>https://storage.googleapis.com/ga4gh-tables-example/tables</td>
      <td>None</td>
      <td>DNAstack Discovery Search test</td>
      <td>com.dnastack.search.test</td>
      <td>0.1.0</td>
      <td>None</td>
    </tr>
    <tr>
      <th>1</th>
      <td>search</td>
      <td>1.0.0</td>
      <td>DNAstack</td>
      <td>prod</td>
      <td>https://ga4gh-search-adapter-presto-public.prod.dnastack.com</td>
      <td>2020-09-14T18:41:09Z</td>
      <td>ga4gh-search-adapter-presto-public</td>
      <td>com.dnastack.search</td>
      <td>0.4-189-gbfc2688</td>
      <td>Supports data discovery by searching heterogeneous genomic and clinical data using a uniform query language.</td>
    </tr>
    <tr>
      <th>14</th>
      <td>service-registry</td>
      <td>1.0.0</td>
      <td>Autism Speaks</td>
      <td>prod</td>
      <td>https://ga4gh-service-registry.mss.ng</td>
      <td>None</td>
      <td>MSSNG GA4GH Service Registry</td>
      <td>org.autismspeaks.mssng.service-registry</td>
      <td>2.0-113-g274a2aa</td>
      <td>Provides links to GA4GH services offered by the MSSNG project</td>
    </tr>
    <tr>
      <th>18</th>
      <td>service-registry</td>
      <td>1.0.0</td>
      <td>Global Alliance for Genomics and Health</td>
      <td>None</td>
      <td>https://registry.ga4gh.org/v1</td>
      <td>None</td>
      <td>GA4GH Standards and Implementations Registry</td>
      <td>org.ga4gh.registry</td>
      <td>1.0.0</td>
      <td>Flexible API registry of all GA4GH standards and associated implementations</td>
    </tr>
    <tr>
      <th>10</th>
      <td>service-registry</td>
      <td>1.0.0</td>
      <td>CSC - IT Center for Science Ltd.</td>
      <td>dev</td>
      <td>https://registry-beacon.rahtiapp.fi</td>
      <td>2020-07-23T18:03:14Z</td>
      <td>ELIXIR-FI Beacon Registry</td>
      <td>fi.rahtiapp.registry-beacon</td>
      <td>1.1.0</td>
      <td>ELIXIR-FI Beacon Registry at CSC for Beacon network</td>
    </tr>
    <tr>
      <th>0</th>
      <td>service-registry</td>
      <td>1.0.0</td>
      <td>DNAstack</td>
      <td>prod</td>
      <td>https://beacon-network.org/api/v2</td>
      <td>None</td>
      <td>DNAstack Beacon Network</td>
      <td>com.dnastack.beacon-network</td>
      <td>1.1-149-ga8b0cf5</td>
      <td>The Beacon Network is a search engine across the world's public beacons. It enables global discovery of genetic mutations, federated across a large and growing network of shared genetic datasets.The Beacon Network is developed, maintained, and hosted by DNAstack.</td>
    </tr>
    <tr>
      <th>15</th>
      <td>trs</td>
      <td>2.0.0</td>
      <td>Dockstore</td>
      <td>None</td>
      <td>https://dockstore.org/api/api/ga4gh/v2/</td>
      <td>None</td>
      <td>Dockstore</td>
      <td>org.dockstore.dockstoreapi</td>
      <td>1.8.1</td>
      <td>Dockstore TRS implementation</td>
    </tr>
    <tr>
      <th>19</th>
      <td>trs</td>
      <td>2.0.0</td>
      <td>Biocontainers</td>
      <td>None</td>
      <td>https://api.biocontainers.pro/api/ga4gh/v2</td>
      <td>None</td>
      <td>Biocontainers</td>
      <td>pro.biocontainers.biocontainersapi</td>
      <td>1.0.0</td>
      <td>Biocontainers TRS implementation</td>
    </tr>
    <tr>
      <th>6</th>
      <td>wes</td>
      <td>1.0.0</td>
      <td>Seven Bridges</td>
      <td>production</td>
      <td>https://cgc-ga4gh-api.sbgenomics.com/ga4gh/wes/v1/</td>
      <td>2020-11-10T18:45:39Z</td>
      <td>CGC GA4GH WES API</td>
      <td>com.sbgenomics.cgc-ga4gh-api.wes</td>
      <td>1.0.1.4</td>
      <td>This service provides a standard way for users to submit workflow requests to workflow execution systems, and to monitor their execution.</td>
    </tr>
    <tr>
      <th>3</th>
      <td>wes</td>
      <td>1.0.0</td>
      <td>DNAstack</td>
      <td>demo</td>
      <td>https://ddap-wes-service.prod.dnastack.com/ga4gh/wes/v1</td>
      <td>2020-09-14T18:45:21Z</td>
      <td>DNAstack Demo WES Service</td>
      <td>com.dnastack.wes.demo</td>
      <td>1.0.0</td>
      <td>Demo Workflow Execution Service (WES)</td>
    </tr>
  </tbody>
</table>