{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Demonstration of GA4GH schema repository with SRA sequencing Experiment Metadata\n",
    "\n",
    "NCBI SRA publishes attributes of sequencing runs and experiments in the Cloud via BigQuery and Athena. Details are provided [here](https://www.ncbi.nlm.nih.gov/sra/docs/sra-cloud-based-metadata-table/)\n",
    "\n",
    "The data in the table above derives from BioProject and BioSample. Besides a set of columns available for all studies in the table, the attributes column allows key-value pairs to be provided for custom attributes for a given study (BioPeoject).\n",
    "\n",
    "This notebook explores how a schema registry can assist in describing the custom metadata for specific studies. With this information it is possible to reconcile attributes across studies.\n",
    "\n",
    "For a set of six NCBI and EBI studies a schema was generated to list the project/study specific attributes.\n",
    "\n",
    "The studies are as follows:\n",
    "* sra_PRJEB10573 - [Characterization of RBP9 and RBP10, two developmentally regulated RNA-binding proteins in Trypanosoma brucei](https://www.ebi.ac.uk/ena/browser/view/PRJEB14125)\n",
    "* sra_PRJEB37886 - [COVID-19 Genomics UK (COG-UK) consortium](https://www.ebi.ac.uk/ena/browser/view/PRJEB37886)\n",
    "* sra_PRJEB1985 - [Whole genome re-sequencing of a single Bos taurus animal for SNP discovery](https://www.ebi.ac.uk/ena/browser/view/PRJEB1985)\n",
    "* sra_phs001554 - [dbGaP Colon Cancer study (BioProject PRJNA437032)](https://www.ncbi.nlm.nih.gov/bioproject/437032)\n",
    "* sra_scr_icac - [dbGaP Childhood Asthma study (BioProject PRJNA837686)](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA837686)\n",
    "* sra_scr_udn_v5 - [dbGaP Undiagnosed Disease Network study (BioProject PRJNA350185)](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA350185)\n",
    "\n",
    "These are a largely unrelated set of studies selected for the purpose of this demonstration. Others may be added.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests;\n",
    "\n",
    "import json;\n",
    "\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "def prettyprint(a_dict):\n",
    "    print(json.dumps(a_dict, indent=3))\n",
    "\n",
    "def printline(char=\"_\"):\n",
    "    print(char*80)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "base = \"http://localhost:8080\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dbGaP\n",
      "dataconnect-demo\n",
      "expt-metadata\n"
     ]
    }
   ],
   "source": [
    "url = f\"{base}/namespaces\"\n",
    "response = requests.get(url)\n",
    "namespaces = response.json()['namespaces']\n",
    "for namespace in namespaces:\n",
    "    print( namespace['namespace_name'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fetch all schemas for the expt-metadata namespace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================================================================================\n",
      "getting schemas for namespace expt-metadata\n",
      "________________________________________________________________________________\n",
      "\n",
      "fetching schema sra_PRJEB10573\n",
      "________________________________________________________________________________\n",
      "\n",
      "fetching schema sra_PRJEB1985\n",
      "________________________________________________________________________________\n",
      "\n",
      "fetching schema sra_PRJEB37886\n",
      "________________________________________________________________________________\n",
      "\n",
      "fetching schema sra_phs001554\n",
      "________________________________________________________________________________\n",
      "\n",
      "fetching schema sra_scr_icac\n",
      "________________________________________________________________________________\n",
      "\n",
      "fetching schema sra_scr_udn_v5\n",
      "________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "namespaces = [\"expt-metadata\"]\n",
    "expt_schemas = {}\n",
    "for namespace in namespaces:\n",
    "    printline(\"=\")\n",
    "    print(f\"getting schemas for namespace {namespace}\")\n",
    "    url = f\"{base}/schemas/{namespace}/\"\n",
    "    response = requests.get(url)\n",
    "    schemas = response.json()['schemas']\n",
    "    printline()\n",
    "    for schema in schemas:\n",
    "        print( )\n",
    "        url = f\"{base}/schemas/{namespace}/{schema['schema_name']}/versions/v2\"\n",
    "        print(f\"fetching schema {schema['schema_name']}\")\n",
    "        response = requests.get(url)\n",
    "        response_schema = response.json()\n",
    "        expt_schemas[schema['schema_name']] = response_schema\n",
    "        #prettyprint( schema)\n",
    "        printline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bases</th>\n",
       "      <th>broker_name_sam</th>\n",
       "      <th>bytes</th>\n",
       "      <th>cell_line_sam</th>\n",
       "      <th>cell_type_sam_ss_dpl37</th>\n",
       "      <th>common_name_sam</th>\n",
       "      <th>ena_first_public_sam</th>\n",
       "      <th>ena_last_update_sam</th>\n",
       "      <th>experimental_factor__over_expression_dazl_exp</th>\n",
       "      <th>experimental_factor__over_expression_nanos3_exp</th>\n",
       "      <th>...</th>\n",
       "      <th>gap_accession_sam</th>\n",
       "      <th>is_tumor_sam</th>\n",
       "      <th>molecular_data_type_sam</th>\n",
       "      <th>phs</th>\n",
       "      <th>sex_calc</th>\n",
       "      <th>study_design_sam</th>\n",
       "      <th>study_disease_sam</th>\n",
       "      <th>study_name_sam</th>\n",
       "      <th>subject_is_affected_sam</th>\n",
       "      <th>submitted_subject_id_sam</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>study</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sra_PRJEB10573</th>\n",
       "      <td>9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_PRJEB1985</th>\n",
       "      <td>49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_PRJEB37886</th>\n",
       "      <td>7511</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7511</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>465.0</td>\n",
       "      <td>270.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_phs001554</th>\n",
       "      <td>2892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2892.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_scr_icac</th>\n",
       "      <td>1035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1035.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_scr_udn_v5</th>\n",
       "      <td>9847</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10028</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6469.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows Ã— 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                bases  broker_name_sam  bytes  cell_line_sam  \\\n",
       "study                                                          \n",
       "sra_PRJEB10573      9              1.0      9            1.0   \n",
       "sra_PRJEB1985      49              NaN     49            NaN   \n",
       "sra_PRJEB37886   7511              3.0   7511            NaN   \n",
       "sra_phs001554    2892              NaN   2892            NaN   \n",
       "sra_scr_icac     1035              NaN   1035            NaN   \n",
       "sra_scr_udn_v5   9847              NaN  10028            NaN   \n",
       "\n",
       "                cell_type_sam_ss_dpl37  common_name_sam  ena_first_public_sam  \\\n",
       "study                                                                           \n",
       "sra_PRJEB10573                     1.0              1.0                   1.0   \n",
       "sra_PRJEB1985                      NaN              NaN                   NaN   \n",
       "sra_PRJEB37886                     NaN              NaN                 465.0   \n",
       "sra_phs001554                      NaN              NaN                   NaN   \n",
       "sra_scr_icac                       NaN              NaN                   NaN   \n",
       "sra_scr_udn_v5                     NaN              NaN                   NaN   \n",
       "\n",
       "                ena_last_update_sam  \\\n",
       "study                                 \n",
       "sra_PRJEB10573                  3.0   \n",
       "sra_PRJEB1985                   NaN   \n",
       "sra_PRJEB37886                270.0   \n",
       "sra_phs001554                   NaN   \n",
       "sra_scr_icac                    NaN   \n",
       "sra_scr_udn_v5                  NaN   \n",
       "\n",
       "                experimental_factor__over_expression_dazl_exp  \\\n",
       "study                                                           \n",
       "sra_PRJEB10573                                            2.0   \n",
       "sra_PRJEB1985                                             NaN   \n",
       "sra_PRJEB37886                                            NaN   \n",
       "sra_phs001554                                             NaN   \n",
       "sra_scr_icac                                              NaN   \n",
       "sra_scr_udn_v5                                            NaN   \n",
       "\n",
       "                experimental_factor__over_expression_nanos3_exp  ...  \\\n",
       "study                                                            ...   \n",
       "sra_PRJEB10573                                              2.0  ...   \n",
       "sra_PRJEB1985                                               NaN  ...   \n",
       "sra_PRJEB37886                                              NaN  ...   \n",
       "sra_phs001554                                               NaN  ...   \n",
       "sra_scr_icac                                                NaN  ...   \n",
       "sra_scr_udn_v5                                              NaN  ...   \n",
       "\n",
       "                gap_accession_sam  is_tumor_sam  molecular_data_type_sam  phs  \\\n",
       "study                                                                           \n",
       "sra_PRJEB10573                NaN           NaN                      NaN  NaN   \n",
       "sra_PRJEB1985                 NaN           NaN                      NaN  NaN   \n",
       "sra_PRJEB37886                NaN           NaN                      NaN  NaN   \n",
       "sra_phs001554                 1.0           1.0                      1.0  5.0   \n",
       "sra_scr_icac                  1.0           1.0                      NaN  1.0   \n",
       "sra_scr_udn_v5                1.0           NaN                      2.0  1.0   \n",
       "\n",
       "                sex_calc  study_design_sam  study_disease_sam  study_name_sam  \\\n",
       "study                                                                           \n",
       "sra_PRJEB10573       NaN               NaN                NaN             NaN   \n",
       "sra_PRJEB1985        NaN               NaN                NaN             NaN   \n",
       "sra_PRJEB37886       NaN               NaN                NaN             NaN   \n",
       "sra_phs001554        2.0               1.0                2.0             1.0   \n",
       "sra_scr_icac         2.0               1.0                NaN             1.0   \n",
       "sra_scr_udn_v5       2.0               1.0                2.0             1.0   \n",
       "\n",
       "                subject_is_affected_sam  submitted_subject_id_sam  \n",
       "study                                                              \n",
       "sra_PRJEB10573                      NaN                       NaN  \n",
       "sra_PRJEB1985                       NaN                       NaN  \n",
       "sra_PRJEB37886                      NaN                       NaN  \n",
       "sra_phs001554                       2.0                    2892.0  \n",
       "sra_scr_icac                        NaN                    1035.0  \n",
       "sra_scr_udn_v5                      3.0                    6469.0  \n",
       "\n",
       "[6 rows x 66 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "tab_list = []\n",
    "include_single_values = False\n",
    "for study_name, study_schema in expt_schemas.items():\n",
    "    tab_row = {\"study\":study_name}\n",
    "    for pname, prop_dict in study_schema['properties'].items():\n",
    "        if include_single_values and prop_dict['unique_count'] == 1:\n",
    "            tab_row[pname] = prop_dict['enums'][0]\n",
    "        else:\n",
    "            tab_row[pname] = prop_dict['$unique_count']\n",
    "    tab_list.append(tab_row)\n",
    "summary_df = pd.DataFrame.from_dict(tab_list)\n",
    "summary_df.set_index('study', inplace=True)\n",
    "summary_df        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Schema comparison across studies\n",
    "\n",
    "In the following table...\n",
    "* Each column represents a different study\n",
    "* Each row represents a different attribute key\n",
    "* The numbers indicate the number of unique values the given attribute takes for each study\n",
    "* A value of 0 indicates that the the attribute is not present in the schema for that study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>study</th>\n",
       "      <th>sra_PRJEB10573</th>\n",
       "      <th>sra_PRJEB1985</th>\n",
       "      <th>sra_PRJEB37886</th>\n",
       "      <th>sra_phs001554</th>\n",
       "      <th>sra_scr_icac</th>\n",
       "      <th>sra_scr_udn_v5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>bases</th>\n",
       "      <td>9</td>\n",
       "      <td>49</td>\n",
       "      <td>7511</td>\n",
       "      <td>2892</td>\n",
       "      <td>1035</td>\n",
       "      <td>9847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>broker_name_sam</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bytes</th>\n",
       "      <td>9</td>\n",
       "      <td>49</td>\n",
       "      <td>7511</td>\n",
       "      <td>2892</td>\n",
       "      <td>1035</td>\n",
       "      <td>10028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line_sam</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_type_sam_ss_dpl37</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>common_name_sam</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ena_first_public_sam</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>465</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ena_last_update_sam</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>270</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experimental_factor__over_expression_dazl_exp</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experimental_factor__over_expression_nanos3_exp</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>external_id_sam</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>7439</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genetic_modification_sam</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype_sam_ss_dpl92</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insdc_center_name_sam</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insdc_first_public_sam</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6734</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insdc_last_update_sam</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>6728</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insdc_status_sam</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run_file_create_date</th>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>5106</td>\n",
       "      <td>0</td>\n",
       "      <td>950</td>\n",
       "      <td>8808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scientific_name_sam</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>submitter_id_sam</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>7440</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alias_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>description_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_accession_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>title_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>artic_primer_version_exp</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>artic_protocol_version_exp</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assemblyname</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>collecting_institution_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>collection_date_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>748</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>collector_name_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>definition_for_seropositive_sample_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ena_last_update_sam_ss_dpl537</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>199</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geographic_location__country_and_or_sea__sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geographic_location__region_and_locality__sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>host_common_name_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>host_health_state_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>host_scientific_name_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>host_sex_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>host_subject_id_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insdc_center_alias_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isolate_sam_ss_dpl100</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max_cycle_threshold_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>330</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min_cycle_threshold_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>328</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organism_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>receipt_date_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>337</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_capture_status_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>serotype__required_for_a_seropositive_sample__sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>virus_identifier_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alignment_software_exp</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>analyte_type_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>avgreadlength_run</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biospecimen_repository_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biospecimen_repository_sample_id_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2892</td>\n",
       "      <td>1035</td>\n",
       "      <td>7991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>body_site_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>consent_code</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coverage_run</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>130</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gap_accession_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_tumor_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>molecular_data_type_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phs</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sex_calc</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>study_design_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>study_disease_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>study_name_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subject_is_affected_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>submitted_subject_id_sam</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2892</td>\n",
       "      <td>1035</td>\n",
       "      <td>6469</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "study                                              sra_PRJEB10573  \\\n",
       "bases                                                           9   \n",
       "broker_name_sam                                                 1   \n",
       "bytes                                                           9   \n",
       "cell_line_sam                                                   1   \n",
       "cell_type_sam_ss_dpl37                                          1   \n",
       "common_name_sam                                                 1   \n",
       "ena_first_public_sam                                            1   \n",
       "ena_last_update_sam                                             3   \n",
       "experimental_factor__over_expression_dazl_exp                   2   \n",
       "experimental_factor__over_expression_nanos3_exp                 2   \n",
       "external_id_sam                                                 9   \n",
       "genetic_modification_sam                                        1   \n",
       "genotype_sam_ss_dpl92                                           3   \n",
       "insdc_center_name_sam                                           1   \n",
       "insdc_first_public_sam                                          1   \n",
       "insdc_last_update_sam                                           3   \n",
       "insdc_status_sam                                                1   \n",
       "run_file_create_date                                            9   \n",
       "scientific_name_sam                                             1   \n",
       "submitter_id_sam                                                9   \n",
       "alias_sam                                                       0   \n",
       "description_sam                                                 0   \n",
       "sra_accession_sam                                               0   \n",
       "title_sam                                                       0   \n",
       "artic_primer_version_exp                                        0   \n",
       "artic_protocol_version_exp                                      0   \n",
       "assemblyname                                                    0   \n",
       "collecting_institution_sam                                      0   \n",
       "collection_date_sam                                             0   \n",
       "collector_name_sam                                              0   \n",
       "definition_for_seropositive_sample_sam                          0   \n",
       "ena_last_update_sam_ss_dpl537                                   0   \n",
       "geographic_location__country_and_or_sea__sam                    0   \n",
       "geographic_location__region_and_locality__sam                   0   \n",
       "host_common_name_sam                                            0   \n",
       "host_health_state_sam                                           0   \n",
       "host_scientific_name_sam                                        0   \n",
       "host_sex_sam                                                    0   \n",
       "host_subject_id_sam                                             0   \n",
       "insdc_center_alias_sam                                          0   \n",
       "isolate_sam_ss_dpl100                                           0   \n",
       "max_cycle_threshold_sam                                         0   \n",
       "min_cycle_threshold_sam                                         0   \n",
       "organism_sam                                                    0   \n",
       "receipt_date_sam                                                0   \n",
       "sample_capture_status_sam                                       0   \n",
       "serotype__required_for_a_seropositive_sample__sam               0   \n",
       "virus_identifier_sam                                            0   \n",
       "alignment_software_exp                                          0   \n",
       "analyte_type_sam                                                0   \n",
       "avgreadlength_run                                               0   \n",
       "biospecimen_repository_sam                                      0   \n",
       "biospecimen_repository_sample_id_sam                            0   \n",
       "body_site_sam                                                   0   \n",
       "consent_code                                                    0   \n",
       "coverage_run                                                    0   \n",
       "gap_accession_sam                                               0   \n",
       "is_tumor_sam                                                    0   \n",
       "molecular_data_type_sam                                         0   \n",
       "phs                                                             0   \n",
       "sex_calc                                                        0   \n",
       "study_design_sam                                                0   \n",
       "study_disease_sam                                               0   \n",
       "study_name_sam                                                  0   \n",
       "subject_is_affected_sam                                         0   \n",
       "submitted_subject_id_sam                                        0   \n",
       "\n",
       "study                                              sra_PRJEB1985  \\\n",
       "bases                                                         49   \n",
       "broker_name_sam                                                0   \n",
       "bytes                                                         49   \n",
       "cell_line_sam                                                  0   \n",
       "cell_type_sam_ss_dpl37                                         0   \n",
       "common_name_sam                                                0   \n",
       "ena_first_public_sam                                           0   \n",
       "ena_last_update_sam                                            0   \n",
       "experimental_factor__over_expression_dazl_exp                  0   \n",
       "experimental_factor__over_expression_nanos3_exp                0   \n",
       "external_id_sam                                                0   \n",
       "genetic_modification_sam                                       0   \n",
       "genotype_sam_ss_dpl92                                          0   \n",
       "insdc_center_name_sam                                          1   \n",
       "insdc_first_public_sam                                         1   \n",
       "insdc_last_update_sam                                          1   \n",
       "insdc_status_sam                                               1   \n",
       "run_file_create_date                                           8   \n",
       "scientific_name_sam                                            0   \n",
       "submitter_id_sam                                               0   \n",
       "alias_sam                                                      1   \n",
       "description_sam                                                1   \n",
       "sra_accession_sam                                              1   \n",
       "title_sam                                                      1   \n",
       "artic_primer_version_exp                                       0   \n",
       "artic_protocol_version_exp                                     0   \n",
       "assemblyname                                                   0   \n",
       "collecting_institution_sam                                     0   \n",
       "collection_date_sam                                            0   \n",
       "collector_name_sam                                             0   \n",
       "definition_for_seropositive_sample_sam                         0   \n",
       "ena_last_update_sam_ss_dpl537                                  0   \n",
       "geographic_location__country_and_or_sea__sam                   0   \n",
       "geographic_location__region_and_locality__sam                  0   \n",
       "host_common_name_sam                                           0   \n",
       "host_health_state_sam                                          0   \n",
       "host_scientific_name_sam                                       0   \n",
       "host_sex_sam                                                   0   \n",
       "host_subject_id_sam                                            0   \n",
       "insdc_center_alias_sam                                         0   \n",
       "isolate_sam_ss_dpl100                                          0   \n",
       "max_cycle_threshold_sam                                        0   \n",
       "min_cycle_threshold_sam                                        0   \n",
       "organism_sam                                                   0   \n",
       "receipt_date_sam                                               0   \n",
       "sample_capture_status_sam                                      0   \n",
       "serotype__required_for_a_seropositive_sample__sam              0   \n",
       "virus_identifier_sam                                           0   \n",
       "alignment_software_exp                                         0   \n",
       "analyte_type_sam                                               0   \n",
       "avgreadlength_run                                              0   \n",
       "biospecimen_repository_sam                                     0   \n",
       "biospecimen_repository_sample_id_sam                           0   \n",
       "body_site_sam                                                  0   \n",
       "consent_code                                                   0   \n",
       "coverage_run                                                   0   \n",
       "gap_accession_sam                                              0   \n",
       "is_tumor_sam                                                   0   \n",
       "molecular_data_type_sam                                        0   \n",
       "phs                                                            0   \n",
       "sex_calc                                                       0   \n",
       "study_design_sam                                               0   \n",
       "study_disease_sam                                              0   \n",
       "study_name_sam                                                 0   \n",
       "subject_is_affected_sam                                        0   \n",
       "submitted_subject_id_sam                                       0   \n",
       "\n",
       "study                                              sra_PRJEB37886  \\\n",
       "bases                                                        7511   \n",
       "broker_name_sam                                                 3   \n",
       "bytes                                                        7511   \n",
       "cell_line_sam                                                   0   \n",
       "cell_type_sam_ss_dpl37                                          0   \n",
       "common_name_sam                                                 0   \n",
       "ena_first_public_sam                                          465   \n",
       "ena_last_update_sam                                           270   \n",
       "experimental_factor__over_expression_dazl_exp                   0   \n",
       "experimental_factor__over_expression_nanos3_exp                 0   \n",
       "external_id_sam                                              7439   \n",
       "genetic_modification_sam                                        0   \n",
       "genotype_sam_ss_dpl92                                           0   \n",
       "insdc_center_name_sam                                          38   \n",
       "insdc_first_public_sam                                       6734   \n",
       "insdc_last_update_sam                                        6728   \n",
       "insdc_status_sam                                                2   \n",
       "run_file_create_date                                         5106   \n",
       "scientific_name_sam                                             2   \n",
       "submitter_id_sam                                             7440   \n",
       "alias_sam                                                       0   \n",
       "description_sam                                                 0   \n",
       "sra_accession_sam                                               0   \n",
       "title_sam                                                       0   \n",
       "artic_primer_version_exp                                        9   \n",
       "artic_protocol_version_exp                                     26   \n",
       "assemblyname                                                    3   \n",
       "collecting_institution_sam                                     43   \n",
       "collection_date_sam                                           748   \n",
       "collector_name_sam                                              2   \n",
       "definition_for_seropositive_sample_sam                          2   \n",
       "ena_last_update_sam_ss_dpl537                                 199   \n",
       "geographic_location__country_and_or_sea__sam                    2   \n",
       "geographic_location__region_and_locality__sam                   6   \n",
       "host_common_name_sam                                            3   \n",
       "host_health_state_sam                                           2   \n",
       "host_scientific_name_sam                                        3   \n",
       "host_sex_sam                                                    2   \n",
       "host_subject_id_sam                                             2   \n",
       "insdc_center_alias_sam                                         31   \n",
       "isolate_sam_ss_dpl100                                           2   \n",
       "max_cycle_threshold_sam                                       330   \n",
       "min_cycle_threshold_sam                                       328   \n",
       "organism_sam                                                    2   \n",
       "receipt_date_sam                                              337   \n",
       "sample_capture_status_sam                                       2   \n",
       "serotype__required_for_a_seropositive_sample__sam               2   \n",
       "virus_identifier_sam                                            2   \n",
       "alignment_software_exp                                          0   \n",
       "analyte_type_sam                                                0   \n",
       "avgreadlength_run                                               0   \n",
       "biospecimen_repository_sam                                      0   \n",
       "biospecimen_repository_sample_id_sam                            0   \n",
       "body_site_sam                                                   0   \n",
       "consent_code                                                    0   \n",
       "coverage_run                                                    0   \n",
       "gap_accession_sam                                               0   \n",
       "is_tumor_sam                                                    0   \n",
       "molecular_data_type_sam                                         0   \n",
       "phs                                                             0   \n",
       "sex_calc                                                        0   \n",
       "study_design_sam                                                0   \n",
       "study_disease_sam                                               0   \n",
       "study_name_sam                                                  0   \n",
       "subject_is_affected_sam                                         0   \n",
       "submitted_subject_id_sam                                        0   \n",
       "\n",
       "study                                              sra_phs001554  \\\n",
       "bases                                                       2892   \n",
       "broker_name_sam                                                0   \n",
       "bytes                                                       2892   \n",
       "cell_line_sam                                                  0   \n",
       "cell_type_sam_ss_dpl37                                         0   \n",
       "common_name_sam                                                0   \n",
       "ena_first_public_sam                                           0   \n",
       "ena_last_update_sam                                            0   \n",
       "experimental_factor__over_expression_dazl_exp                  0   \n",
       "experimental_factor__over_expression_nanos3_exp                0   \n",
       "external_id_sam                                                0   \n",
       "genetic_modification_sam                                       0   \n",
       "genotype_sam_ss_dpl92                                          0   \n",
       "insdc_center_name_sam                                          0   \n",
       "insdc_first_public_sam                                         0   \n",
       "insdc_last_update_sam                                          0   \n",
       "insdc_status_sam                                               0   \n",
       "run_file_create_date                                           0   \n",
       "scientific_name_sam                                            0   \n",
       "submitter_id_sam                                               0   \n",
       "alias_sam                                                      0   \n",
       "description_sam                                                0   \n",
       "sra_accession_sam                                              0   \n",
       "title_sam                                                      0   \n",
       "artic_primer_version_exp                                       0   \n",
       "artic_protocol_version_exp                                     0   \n",
       "assemblyname                                                   1   \n",
       "collecting_institution_sam                                     0   \n",
       "collection_date_sam                                            0   \n",
       "collector_name_sam                                             0   \n",
       "definition_for_seropositive_sample_sam                         0   \n",
       "ena_last_update_sam_ss_dpl537                                  0   \n",
       "geographic_location__country_and_or_sea__sam                   0   \n",
       "geographic_location__region_and_locality__sam                  0   \n",
       "host_common_name_sam                                           0   \n",
       "host_health_state_sam                                          0   \n",
       "host_scientific_name_sam                                       0   \n",
       "host_sex_sam                                                   0   \n",
       "host_subject_id_sam                                            0   \n",
       "insdc_center_alias_sam                                         0   \n",
       "isolate_sam_ss_dpl100                                          0   \n",
       "max_cycle_threshold_sam                                        0   \n",
       "min_cycle_threshold_sam                                        0   \n",
       "organism_sam                                                   0   \n",
       "receipt_date_sam                                               0   \n",
       "sample_capture_status_sam                                      0   \n",
       "serotype__required_for_a_seropositive_sample__sam              0   \n",
       "virus_identifier_sam                                           0   \n",
       "alignment_software_exp                                         1   \n",
       "analyte_type_sam                                               1   \n",
       "avgreadlength_run                                             58   \n",
       "biospecimen_repository_sam                                     1   \n",
       "biospecimen_repository_sample_id_sam                        2892   \n",
       "body_site_sam                                                  1   \n",
       "consent_code                                                   5   \n",
       "coverage_run                                                 130   \n",
       "gap_accession_sam                                              1   \n",
       "is_tumor_sam                                                   1   \n",
       "molecular_data_type_sam                                        1   \n",
       "phs                                                            5   \n",
       "sex_calc                                                       2   \n",
       "study_design_sam                                               1   \n",
       "study_disease_sam                                              2   \n",
       "study_name_sam                                                 1   \n",
       "subject_is_affected_sam                                        2   \n",
       "submitted_subject_id_sam                                    2892   \n",
       "\n",
       "study                                              sra_scr_icac  \\\n",
       "bases                                                      1035   \n",
       "broker_name_sam                                               0   \n",
       "bytes                                                      1035   \n",
       "cell_line_sam                                                 0   \n",
       "cell_type_sam_ss_dpl37                                        0   \n",
       "common_name_sam                                               0   \n",
       "ena_first_public_sam                                          0   \n",
       "ena_last_update_sam                                           0   \n",
       "experimental_factor__over_expression_dazl_exp                 0   \n",
       "experimental_factor__over_expression_nanos3_exp               0   \n",
       "external_id_sam                                               0   \n",
       "genetic_modification_sam                                      0   \n",
       "genotype_sam_ss_dpl92                                         0   \n",
       "insdc_center_name_sam                                         0   \n",
       "insdc_first_public_sam                                        0   \n",
       "insdc_last_update_sam                                         0   \n",
       "insdc_status_sam                                              0   \n",
       "run_file_create_date                                        950   \n",
       "scientific_name_sam                                           0   \n",
       "submitter_id_sam                                              0   \n",
       "alias_sam                                                     0   \n",
       "description_sam                                               0   \n",
       "sra_accession_sam                                             0   \n",
       "title_sam                                                     0   \n",
       "artic_primer_version_exp                                      0   \n",
       "artic_protocol_version_exp                                    0   \n",
       "assemblyname                                                  1   \n",
       "collecting_institution_sam                                    0   \n",
       "collection_date_sam                                           0   \n",
       "collector_name_sam                                            0   \n",
       "definition_for_seropositive_sample_sam                        0   \n",
       "ena_last_update_sam_ss_dpl537                                 0   \n",
       "geographic_location__country_and_or_sea__sam                  0   \n",
       "geographic_location__region_and_locality__sam                 0   \n",
       "host_common_name_sam                                          0   \n",
       "host_health_state_sam                                         0   \n",
       "host_scientific_name_sam                                      0   \n",
       "host_sex_sam                                                  0   \n",
       "host_subject_id_sam                                           0   \n",
       "insdc_center_alias_sam                                        0   \n",
       "isolate_sam_ss_dpl100                                         0   \n",
       "max_cycle_threshold_sam                                       0   \n",
       "min_cycle_threshold_sam                                       0   \n",
       "organism_sam                                                  0   \n",
       "receipt_date_sam                                              0   \n",
       "sample_capture_status_sam                                     0   \n",
       "serotype__required_for_a_seropositive_sample__sam             0   \n",
       "virus_identifier_sam                                          0   \n",
       "alignment_software_exp                                        1   \n",
       "analyte_type_sam                                              1   \n",
       "avgreadlength_run                                             0   \n",
       "biospecimen_repository_sam                                    1   \n",
       "biospecimen_repository_sample_id_sam                       1035   \n",
       "body_site_sam                                                 1   \n",
       "consent_code                                                  1   \n",
       "coverage_run                                                  0   \n",
       "gap_accession_sam                                             1   \n",
       "is_tumor_sam                                                  1   \n",
       "molecular_data_type_sam                                       0   \n",
       "phs                                                           1   \n",
       "sex_calc                                                      2   \n",
       "study_design_sam                                              1   \n",
       "study_disease_sam                                             0   \n",
       "study_name_sam                                                1   \n",
       "subject_is_affected_sam                                       0   \n",
       "submitted_subject_id_sam                                   1035   \n",
       "\n",
       "study                                              sra_scr_udn_v5  \n",
       "bases                                                        9847  \n",
       "broker_name_sam                                                 0  \n",
       "bytes                                                       10028  \n",
       "cell_line_sam                                                   0  \n",
       "cell_type_sam_ss_dpl37                                          0  \n",
       "common_name_sam                                                 0  \n",
       "ena_first_public_sam                                            0  \n",
       "ena_last_update_sam                                             0  \n",
       "experimental_factor__over_expression_dazl_exp                   0  \n",
       "experimental_factor__over_expression_nanos3_exp                 0  \n",
       "external_id_sam                                                 0  \n",
       "genetic_modification_sam                                        0  \n",
       "genotype_sam_ss_dpl92                                           0  \n",
       "insdc_center_name_sam                                           0  \n",
       "insdc_first_public_sam                                          0  \n",
       "insdc_last_update_sam                                           0  \n",
       "insdc_status_sam                                                0  \n",
       "run_file_create_date                                         8808  \n",
       "scientific_name_sam                                             0  \n",
       "submitter_id_sam                                                0  \n",
       "alias_sam                                                       0  \n",
       "description_sam                                                 0  \n",
       "sra_accession_sam                                               0  \n",
       "title_sam                                                       0  \n",
       "artic_primer_version_exp                                        0  \n",
       "artic_protocol_version_exp                                      0  \n",
       "assemblyname                                                    6  \n",
       "collecting_institution_sam                                      0  \n",
       "collection_date_sam                                             0  \n",
       "collector_name_sam                                              0  \n",
       "definition_for_seropositive_sample_sam                          0  \n",
       "ena_last_update_sam_ss_dpl537                                   0  \n",
       "geographic_location__country_and_or_sea__sam                    0  \n",
       "geographic_location__region_and_locality__sam                   0  \n",
       "host_common_name_sam                                            0  \n",
       "host_health_state_sam                                           0  \n",
       "host_scientific_name_sam                                        0  \n",
       "host_sex_sam                                                    0  \n",
       "host_subject_id_sam                                             0  \n",
       "insdc_center_alias_sam                                          0  \n",
       "isolate_sam_ss_dpl100                                           0  \n",
       "max_cycle_threshold_sam                                         0  \n",
       "min_cycle_threshold_sam                                         0  \n",
       "organism_sam                                                    0  \n",
       "receipt_date_sam                                                0  \n",
       "sample_capture_status_sam                                       0  \n",
       "serotype__required_for_a_seropositive_sample__sam               0  \n",
       "virus_identifier_sam                                            0  \n",
       "alignment_software_exp                                          3  \n",
       "analyte_type_sam                                                2  \n",
       "avgreadlength_run                                               0  \n",
       "biospecimen_repository_sam                                      1  \n",
       "biospecimen_repository_sample_id_sam                         7991  \n",
       "body_site_sam                                                   3  \n",
       "consent_code                                                    1  \n",
       "coverage_run                                                    0  \n",
       "gap_accession_sam                                               1  \n",
       "is_tumor_sam                                                    0  \n",
       "molecular_data_type_sam                                         2  \n",
       "phs                                                             1  \n",
       "sex_calc                                                        2  \n",
       "study_design_sam                                                1  \n",
       "study_disease_sam                                               2  \n",
       "study_name_sam                                                  1  \n",
       "subject_is_affected_sam                                         3  \n",
       "submitted_subject_id_sam                                     6469  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "attribute_summary_df = summary_df.transpose()\n",
    "attribute_summary_df.fillna(0, inplace=True)\n",
    "for scname in expt_schemas.keys():\n",
    "    attribute_summary_df[scname] = attribute_summary_df[scname].astype(int)\n",
    "attribute_summary_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Show the table as a heat map\n",
    "A quick overview of which attributes are used in which studies can be seen in the following heat map."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_d0e83_row0_col0, #T_d0e83_row0_col1, #T_d0e83_row0_col2, #T_d0e83_row0_col3, #T_d0e83_row0_col4, #T_d0e83_row0_col5, #T_d0e83_row1_col0, #T_d0e83_row1_col2, #T_d0e83_row2_col0, #T_d0e83_row2_col1, #T_d0e83_row2_col2, #T_d0e83_row2_col3, #T_d0e83_row2_col4, #T_d0e83_row2_col5, #T_d0e83_row3_col0, #T_d0e83_row4_col0, #T_d0e83_row5_col0, #T_d0e83_row6_col0, #T_d0e83_row6_col2, #T_d0e83_row7_col0, #T_d0e83_row7_col2, #T_d0e83_row8_col0, #T_d0e83_row9_col0, #T_d0e83_row10_col0, #T_d0e83_row10_col2, #T_d0e83_row11_col0, #T_d0e83_row12_col0, #T_d0e83_row13_col0, #T_d0e83_row13_col1, #T_d0e83_row13_col2, #T_d0e83_row14_col0, #T_d0e83_row14_col1, #T_d0e83_row14_col2, #T_d0e83_row15_col0, #T_d0e83_row15_col1, #T_d0e83_row15_col2, #T_d0e83_row16_col0, #T_d0e83_row16_col1, #T_d0e83_row16_col2, #T_d0e83_row17_col0, #T_d0e83_row17_col1, #T_d0e83_row17_col2, #T_d0e83_row17_col4, #T_d0e83_row17_col5, #T_d0e83_row18_col0, #T_d0e83_row18_col2, #T_d0e83_row19_col0, #T_d0e83_row19_col2, #T_d0e83_row20_col1, #T_d0e83_row21_col1, #T_d0e83_row22_col1, #T_d0e83_row23_col1, #T_d0e83_row24_col2, #T_d0e83_row25_col2, #T_d0e83_row26_col2, #T_d0e83_row26_col3, #T_d0e83_row26_col4, #T_d0e83_row26_col5, #T_d0e83_row27_col2, #T_d0e83_row28_col2, #T_d0e83_row29_col2, #T_d0e83_row30_col2, #T_d0e83_row31_col2, #T_d0e83_row32_col2, #T_d0e83_row33_col2, #T_d0e83_row34_col2, #T_d0e83_row35_col2, #T_d0e83_row36_col2, #T_d0e83_row37_col2, #T_d0e83_row38_col2, #T_d0e83_row39_col2, #T_d0e83_row40_col2, #T_d0e83_row41_col2, #T_d0e83_row42_col2, #T_d0e83_row43_col2, #T_d0e83_row44_col2, #T_d0e83_row45_col2, #T_d0e83_row46_col2, #T_d0e83_row47_col2, #T_d0e83_row48_col3, #T_d0e83_row48_col4, #T_d0e83_row48_col5, #T_d0e83_row49_col3, #T_d0e83_row49_col4, #T_d0e83_row49_col5, #T_d0e83_row50_col3, #T_d0e83_row51_col3, #T_d0e83_row51_col4, #T_d0e83_row51_col5, #T_d0e83_row52_col3, #T_d0e83_row52_col4, #T_d0e83_row52_col5, #T_d0e83_row53_col3, #T_d0e83_row53_col4, #T_d0e83_row53_col5, #T_d0e83_row54_col3, #T_d0e83_row54_col4, #T_d0e83_row54_col5, #T_d0e83_row55_col3, #T_d0e83_row56_col3, #T_d0e83_row56_col4, #T_d0e83_row56_col5, #T_d0e83_row57_col3, #T_d0e83_row57_col4, #T_d0e83_row58_col3, #T_d0e83_row58_col5, #T_d0e83_row59_col3, #T_d0e83_row59_col4, #T_d0e83_row59_col5, #T_d0e83_row60_col3, #T_d0e83_row60_col4, #T_d0e83_row60_col5, #T_d0e83_row61_col3, #T_d0e83_row61_col4, #T_d0e83_row61_col5, #T_d0e83_row62_col3, #T_d0e83_row62_col5, #T_d0e83_row63_col3, #T_d0e83_row63_col4, #T_d0e83_row63_col5, #T_d0e83_row64_col3, #T_d0e83_row64_col5, #T_d0e83_row65_col3, #T_d0e83_row65_col4, #T_d0e83_row65_col5 {\n",
       "  background-color: #fde725;\n",
       "  color: #000000;\n",
       "  font-size: 1px;\n",
       "}\n",
       "#T_d0e83_row1_col1, #T_d0e83_row1_col3, #T_d0e83_row1_col4, #T_d0e83_row1_col5, #T_d0e83_row3_col1, #T_d0e83_row3_col2, #T_d0e83_row3_col3, #T_d0e83_row3_col4, #T_d0e83_row3_col5, #T_d0e83_row4_col1, #T_d0e83_row4_col2, #T_d0e83_row4_col3, #T_d0e83_row4_col4, #T_d0e83_row4_col5, #T_d0e83_row5_col1, #T_d0e83_row5_col2, #T_d0e83_row5_col3, #T_d0e83_row5_col4, #T_d0e83_row5_col5, #T_d0e83_row6_col1, #T_d0e83_row6_col3, #T_d0e83_row6_col4, #T_d0e83_row6_col5, #T_d0e83_row7_col1, #T_d0e83_row7_col3, #T_d0e83_row7_col4, #T_d0e83_row7_col5, #T_d0e83_row8_col1, #T_d0e83_row8_col2, #T_d0e83_row8_col3, #T_d0e83_row8_col4, #T_d0e83_row8_col5, #T_d0e83_row9_col1, #T_d0e83_row9_col2, #T_d0e83_row9_col3, #T_d0e83_row9_col4, #T_d0e83_row9_col5, #T_d0e83_row10_col1, #T_d0e83_row10_col3, #T_d0e83_row10_col4, #T_d0e83_row10_col5, #T_d0e83_row11_col1, #T_d0e83_row11_col2, #T_d0e83_row11_col3, #T_d0e83_row11_col4, #T_d0e83_row11_col5, #T_d0e83_row12_col1, #T_d0e83_row12_col2, #T_d0e83_row12_col3, #T_d0e83_row12_col4, #T_d0e83_row12_col5, #T_d0e83_row13_col3, #T_d0e83_row13_col4, #T_d0e83_row13_col5, #T_d0e83_row14_col3, #T_d0e83_row14_col4, #T_d0e83_row14_col5, #T_d0e83_row15_col3, #T_d0e83_row15_col4, #T_d0e83_row15_col5, #T_d0e83_row16_col3, #T_d0e83_row16_col4, #T_d0e83_row16_col5, #T_d0e83_row17_col3, #T_d0e83_row18_col1, #T_d0e83_row18_col3, #T_d0e83_row18_col4, #T_d0e83_row18_col5, #T_d0e83_row19_col1, #T_d0e83_row19_col3, #T_d0e83_row19_col4, #T_d0e83_row19_col5, #T_d0e83_row20_col0, #T_d0e83_row20_col2, #T_d0e83_row20_col3, #T_d0e83_row20_col4, #T_d0e83_row20_col5, #T_d0e83_row21_col0, #T_d0e83_row21_col2, #T_d0e83_row21_col3, #T_d0e83_row21_col4, #T_d0e83_row21_col5, #T_d0e83_row22_col0, #T_d0e83_row22_col2, #T_d0e83_row22_col3, #T_d0e83_row22_col4, #T_d0e83_row22_col5, #T_d0e83_row23_col0, #T_d0e83_row23_col2, #T_d0e83_row23_col3, #T_d0e83_row23_col4, #T_d0e83_row23_col5, #T_d0e83_row24_col0, #T_d0e83_row24_col1, #T_d0e83_row24_col3, #T_d0e83_row24_col4, #T_d0e83_row24_col5, #T_d0e83_row25_col0, #T_d0e83_row25_col1, #T_d0e83_row25_col3, #T_d0e83_row25_col4, #T_d0e83_row25_col5, #T_d0e83_row26_col0, #T_d0e83_row26_col1, #T_d0e83_row27_col0, #T_d0e83_row27_col1, #T_d0e83_row27_col3, #T_d0e83_row27_col4, #T_d0e83_row27_col5, #T_d0e83_row28_col0, #T_d0e83_row28_col1, #T_d0e83_row28_col3, #T_d0e83_row28_col4, #T_d0e83_row28_col5, #T_d0e83_row29_col0, #T_d0e83_row29_col1, #T_d0e83_row29_col3, #T_d0e83_row29_col4, #T_d0e83_row29_col5, #T_d0e83_row30_col0, #T_d0e83_row30_col1, #T_d0e83_row30_col3, #T_d0e83_row30_col4, #T_d0e83_row30_col5, #T_d0e83_row31_col0, #T_d0e83_row31_col1, #T_d0e83_row31_col3, #T_d0e83_row31_col4, #T_d0e83_row31_col5, #T_d0e83_row32_col0, #T_d0e83_row32_col1, #T_d0e83_row32_col3, #T_d0e83_row32_col4, #T_d0e83_row32_col5, #T_d0e83_row33_col0, #T_d0e83_row33_col1, #T_d0e83_row33_col3, #T_d0e83_row33_col4, #T_d0e83_row33_col5, #T_d0e83_row34_col0, #T_d0e83_row34_col1, #T_d0e83_row34_col3, #T_d0e83_row34_col4, #T_d0e83_row34_col5, #T_d0e83_row35_col0, #T_d0e83_row35_col1, #T_d0e83_row35_col3, #T_d0e83_row35_col4, #T_d0e83_row35_col5, #T_d0e83_row36_col0, #T_d0e83_row36_col1, #T_d0e83_row36_col3, #T_d0e83_row36_col4, #T_d0e83_row36_col5, #T_d0e83_row37_col0, #T_d0e83_row37_col1, #T_d0e83_row37_col3, #T_d0e83_row37_col4, #T_d0e83_row37_col5, #T_d0e83_row38_col0, #T_d0e83_row38_col1, #T_d0e83_row38_col3, #T_d0e83_row38_col4, #T_d0e83_row38_col5, #T_d0e83_row39_col0, #T_d0e83_row39_col1, #T_d0e83_row39_col3, #T_d0e83_row39_col4, #T_d0e83_row39_col5, #T_d0e83_row40_col0, #T_d0e83_row40_col1, #T_d0e83_row40_col3, #T_d0e83_row40_col4, #T_d0e83_row40_col5, #T_d0e83_row41_col0, #T_d0e83_row41_col1, #T_d0e83_row41_col3, #T_d0e83_row41_col4, #T_d0e83_row41_col5, #T_d0e83_row42_col0, #T_d0e83_row42_col1, #T_d0e83_row42_col3, #T_d0e83_row42_col4, #T_d0e83_row42_col5, #T_d0e83_row43_col0, #T_d0e83_row43_col1, #T_d0e83_row43_col3, #T_d0e83_row43_col4, #T_d0e83_row43_col5, #T_d0e83_row44_col0, #T_d0e83_row44_col1, #T_d0e83_row44_col3, #T_d0e83_row44_col4, #T_d0e83_row44_col5, #T_d0e83_row45_col0, #T_d0e83_row45_col1, #T_d0e83_row45_col3, #T_d0e83_row45_col4, #T_d0e83_row45_col5, #T_d0e83_row46_col0, #T_d0e83_row46_col1, #T_d0e83_row46_col3, #T_d0e83_row46_col4, #T_d0e83_row46_col5, #T_d0e83_row47_col0, #T_d0e83_row47_col1, #T_d0e83_row47_col3, #T_d0e83_row47_col4, #T_d0e83_row47_col5, #T_d0e83_row48_col0, #T_d0e83_row48_col1, #T_d0e83_row48_col2, #T_d0e83_row49_col0, #T_d0e83_row49_col1, #T_d0e83_row49_col2, #T_d0e83_row50_col0, #T_d0e83_row50_col1, #T_d0e83_row50_col2, #T_d0e83_row50_col4, #T_d0e83_row50_col5, #T_d0e83_row51_col0, #T_d0e83_row51_col1, #T_d0e83_row51_col2, #T_d0e83_row52_col0, #T_d0e83_row52_col1, #T_d0e83_row52_col2, #T_d0e83_row53_col0, #T_d0e83_row53_col1, #T_d0e83_row53_col2, #T_d0e83_row54_col0, #T_d0e83_row54_col1, #T_d0e83_row54_col2, #T_d0e83_row55_col0, #T_d0e83_row55_col1, #T_d0e83_row55_col2, #T_d0e83_row55_col4, #T_d0e83_row55_col5, #T_d0e83_row56_col0, #T_d0e83_row56_col1, #T_d0e83_row56_col2, #T_d0e83_row57_col0, #T_d0e83_row57_col1, #T_d0e83_row57_col2, #T_d0e83_row57_col5, #T_d0e83_row58_col0, #T_d0e83_row58_col1, #T_d0e83_row58_col2, #T_d0e83_row58_col4, #T_d0e83_row59_col0, #T_d0e83_row59_col1, #T_d0e83_row59_col2, #T_d0e83_row60_col0, #T_d0e83_row60_col1, #T_d0e83_row60_col2, #T_d0e83_row61_col0, #T_d0e83_row61_col1, #T_d0e83_row61_col2, #T_d0e83_row62_col0, #T_d0e83_row62_col1, #T_d0e83_row62_col2, #T_d0e83_row62_col4, #T_d0e83_row63_col0, #T_d0e83_row63_col1, #T_d0e83_row63_col2, #T_d0e83_row64_col0, #T_d0e83_row64_col1, #T_d0e83_row64_col2, #T_d0e83_row64_col4, #T_d0e83_row65_col0, #T_d0e83_row65_col1, #T_d0e83_row65_col2 {\n",
       "  background-color: #440154;\n",
       "  color: #f1f1f1;\n",
       "  font-size: 1px;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_d0e83\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >study</th>\n",
       "      <th id=\"T_d0e83_level0_col0\" class=\"col_heading level0 col0\" >sra_PRJEB10573</th>\n",
       "      <th id=\"T_d0e83_level0_col1\" class=\"col_heading level0 col1\" >sra_PRJEB1985</th>\n",
       "      <th id=\"T_d0e83_level0_col2\" class=\"col_heading level0 col2\" >sra_PRJEB37886</th>\n",
       "      <th id=\"T_d0e83_level0_col3\" class=\"col_heading level0 col3\" >sra_phs001554</th>\n",
       "      <th id=\"T_d0e83_level0_col4\" class=\"col_heading level0 col4\" >sra_scr_icac</th>\n",
       "      <th id=\"T_d0e83_level0_col5\" class=\"col_heading level0 col5\" >sra_scr_udn_v5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row0\" class=\"row_heading level0 row0\" >bases</th>\n",
       "      <td id=\"T_d0e83_row0_col0\" class=\"data row0 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row0_col1\" class=\"data row0 col1\" >1</td>\n",
       "      <td id=\"T_d0e83_row0_col2\" class=\"data row0 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row0_col3\" class=\"data row0 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row0_col4\" class=\"data row0 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row0_col5\" class=\"data row0 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row1\" class=\"row_heading level0 row1\" >broker_name_sam</th>\n",
       "      <td id=\"T_d0e83_row1_col0\" class=\"data row1 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row1_col1\" class=\"data row1 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row1_col2\" class=\"data row1 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row1_col3\" class=\"data row1 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row1_col4\" class=\"data row1 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row1_col5\" class=\"data row1 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row2\" class=\"row_heading level0 row2\" >bytes</th>\n",
       "      <td id=\"T_d0e83_row2_col0\" class=\"data row2 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row2_col1\" class=\"data row2 col1\" >1</td>\n",
       "      <td id=\"T_d0e83_row2_col2\" class=\"data row2 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row2_col3\" class=\"data row2 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row2_col4\" class=\"data row2 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row2_col5\" class=\"data row2 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row3\" class=\"row_heading level0 row3\" >cell_line_sam</th>\n",
       "      <td id=\"T_d0e83_row3_col0\" class=\"data row3 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row3_col1\" class=\"data row3 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row3_col2\" class=\"data row3 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row3_col3\" class=\"data row3 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row3_col4\" class=\"data row3 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row3_col5\" class=\"data row3 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row4\" class=\"row_heading level0 row4\" >cell_type_sam_ss_dpl37</th>\n",
       "      <td id=\"T_d0e83_row4_col0\" class=\"data row4 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row4_col1\" class=\"data row4 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row4_col2\" class=\"data row4 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row4_col3\" class=\"data row4 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row4_col4\" class=\"data row4 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row4_col5\" class=\"data row4 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row5\" class=\"row_heading level0 row5\" >common_name_sam</th>\n",
       "      <td id=\"T_d0e83_row5_col0\" class=\"data row5 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row5_col1\" class=\"data row5 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row5_col2\" class=\"data row5 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row5_col3\" class=\"data row5 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row5_col4\" class=\"data row5 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row5_col5\" class=\"data row5 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row6\" class=\"row_heading level0 row6\" >ena_first_public_sam</th>\n",
       "      <td id=\"T_d0e83_row6_col0\" class=\"data row6 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row6_col1\" class=\"data row6 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row6_col2\" class=\"data row6 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row6_col3\" class=\"data row6 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row6_col4\" class=\"data row6 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row6_col5\" class=\"data row6 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row7\" class=\"row_heading level0 row7\" >ena_last_update_sam</th>\n",
       "      <td id=\"T_d0e83_row7_col0\" class=\"data row7 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row7_col1\" class=\"data row7 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row7_col2\" class=\"data row7 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row7_col3\" class=\"data row7 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row7_col4\" class=\"data row7 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row7_col5\" class=\"data row7 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row8\" class=\"row_heading level0 row8\" >experimental_factor__over_expression_dazl_exp</th>\n",
       "      <td id=\"T_d0e83_row8_col0\" class=\"data row8 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row8_col1\" class=\"data row8 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row8_col2\" class=\"data row8 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row8_col3\" class=\"data row8 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row8_col4\" class=\"data row8 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row8_col5\" class=\"data row8 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row9\" class=\"row_heading level0 row9\" >experimental_factor__over_expression_nanos3_exp</th>\n",
       "      <td id=\"T_d0e83_row9_col0\" class=\"data row9 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row9_col1\" class=\"data row9 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row9_col2\" class=\"data row9 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row9_col3\" class=\"data row9 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row9_col4\" class=\"data row9 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row9_col5\" class=\"data row9 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row10\" class=\"row_heading level0 row10\" >external_id_sam</th>\n",
       "      <td id=\"T_d0e83_row10_col0\" class=\"data row10 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row10_col1\" class=\"data row10 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row10_col2\" class=\"data row10 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row10_col3\" class=\"data row10 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row10_col4\" class=\"data row10 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row10_col5\" class=\"data row10 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row11\" class=\"row_heading level0 row11\" >genetic_modification_sam</th>\n",
       "      <td id=\"T_d0e83_row11_col0\" class=\"data row11 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row11_col1\" class=\"data row11 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row11_col2\" class=\"data row11 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row11_col3\" class=\"data row11 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row11_col4\" class=\"data row11 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row11_col5\" class=\"data row11 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row12\" class=\"row_heading level0 row12\" >genotype_sam_ss_dpl92</th>\n",
       "      <td id=\"T_d0e83_row12_col0\" class=\"data row12 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row12_col1\" class=\"data row12 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row12_col2\" class=\"data row12 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row12_col3\" class=\"data row12 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row12_col4\" class=\"data row12 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row12_col5\" class=\"data row12 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row13\" class=\"row_heading level0 row13\" >insdc_center_name_sam</th>\n",
       "      <td id=\"T_d0e83_row13_col0\" class=\"data row13 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row13_col1\" class=\"data row13 col1\" >1</td>\n",
       "      <td id=\"T_d0e83_row13_col2\" class=\"data row13 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row13_col3\" class=\"data row13 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row13_col4\" class=\"data row13 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row13_col5\" class=\"data row13 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row14\" class=\"row_heading level0 row14\" >insdc_first_public_sam</th>\n",
       "      <td id=\"T_d0e83_row14_col0\" class=\"data row14 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row14_col1\" class=\"data row14 col1\" >1</td>\n",
       "      <td id=\"T_d0e83_row14_col2\" class=\"data row14 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row14_col3\" class=\"data row14 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row14_col4\" class=\"data row14 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row14_col5\" class=\"data row14 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row15\" class=\"row_heading level0 row15\" >insdc_last_update_sam</th>\n",
       "      <td id=\"T_d0e83_row15_col0\" class=\"data row15 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row15_col1\" class=\"data row15 col1\" >1</td>\n",
       "      <td id=\"T_d0e83_row15_col2\" class=\"data row15 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row15_col3\" class=\"data row15 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row15_col4\" class=\"data row15 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row15_col5\" class=\"data row15 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row16\" class=\"row_heading level0 row16\" >insdc_status_sam</th>\n",
       "      <td id=\"T_d0e83_row16_col0\" class=\"data row16 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row16_col1\" class=\"data row16 col1\" >1</td>\n",
       "      <td id=\"T_d0e83_row16_col2\" class=\"data row16 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row16_col3\" class=\"data row16 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row16_col4\" class=\"data row16 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row16_col5\" class=\"data row16 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row17\" class=\"row_heading level0 row17\" >run_file_create_date</th>\n",
       "      <td id=\"T_d0e83_row17_col0\" class=\"data row17 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row17_col1\" class=\"data row17 col1\" >1</td>\n",
       "      <td id=\"T_d0e83_row17_col2\" class=\"data row17 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row17_col3\" class=\"data row17 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row17_col4\" class=\"data row17 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row17_col5\" class=\"data row17 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row18\" class=\"row_heading level0 row18\" >scientific_name_sam</th>\n",
       "      <td id=\"T_d0e83_row18_col0\" class=\"data row18 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row18_col1\" class=\"data row18 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row18_col2\" class=\"data row18 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row18_col3\" class=\"data row18 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row18_col4\" class=\"data row18 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row18_col5\" class=\"data row18 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row19\" class=\"row_heading level0 row19\" >submitter_id_sam</th>\n",
       "      <td id=\"T_d0e83_row19_col0\" class=\"data row19 col0\" >1</td>\n",
       "      <td id=\"T_d0e83_row19_col1\" class=\"data row19 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row19_col2\" class=\"data row19 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row19_col3\" class=\"data row19 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row19_col4\" class=\"data row19 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row19_col5\" class=\"data row19 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row20\" class=\"row_heading level0 row20\" >alias_sam</th>\n",
       "      <td id=\"T_d0e83_row20_col0\" class=\"data row20 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row20_col1\" class=\"data row20 col1\" >1</td>\n",
       "      <td id=\"T_d0e83_row20_col2\" class=\"data row20 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row20_col3\" class=\"data row20 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row20_col4\" class=\"data row20 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row20_col5\" class=\"data row20 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row21\" class=\"row_heading level0 row21\" >description_sam</th>\n",
       "      <td id=\"T_d0e83_row21_col0\" class=\"data row21 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row21_col1\" class=\"data row21 col1\" >1</td>\n",
       "      <td id=\"T_d0e83_row21_col2\" class=\"data row21 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row21_col3\" class=\"data row21 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row21_col4\" class=\"data row21 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row21_col5\" class=\"data row21 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row22\" class=\"row_heading level0 row22\" >sra_accession_sam</th>\n",
       "      <td id=\"T_d0e83_row22_col0\" class=\"data row22 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row22_col1\" class=\"data row22 col1\" >1</td>\n",
       "      <td id=\"T_d0e83_row22_col2\" class=\"data row22 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row22_col3\" class=\"data row22 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row22_col4\" class=\"data row22 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row22_col5\" class=\"data row22 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row23\" class=\"row_heading level0 row23\" >title_sam</th>\n",
       "      <td id=\"T_d0e83_row23_col0\" class=\"data row23 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row23_col1\" class=\"data row23 col1\" >1</td>\n",
       "      <td id=\"T_d0e83_row23_col2\" class=\"data row23 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row23_col3\" class=\"data row23 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row23_col4\" class=\"data row23 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row23_col5\" class=\"data row23 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row24\" class=\"row_heading level0 row24\" >artic_primer_version_exp</th>\n",
       "      <td id=\"T_d0e83_row24_col0\" class=\"data row24 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row24_col1\" class=\"data row24 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row24_col2\" class=\"data row24 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row24_col3\" class=\"data row24 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row24_col4\" class=\"data row24 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row24_col5\" class=\"data row24 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row25\" class=\"row_heading level0 row25\" >artic_protocol_version_exp</th>\n",
       "      <td id=\"T_d0e83_row25_col0\" class=\"data row25 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row25_col1\" class=\"data row25 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row25_col2\" class=\"data row25 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row25_col3\" class=\"data row25 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row25_col4\" class=\"data row25 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row25_col5\" class=\"data row25 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row26\" class=\"row_heading level0 row26\" >assemblyname</th>\n",
       "      <td id=\"T_d0e83_row26_col0\" class=\"data row26 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row26_col1\" class=\"data row26 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row26_col2\" class=\"data row26 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row26_col3\" class=\"data row26 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row26_col4\" class=\"data row26 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row26_col5\" class=\"data row26 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row27\" class=\"row_heading level0 row27\" >collecting_institution_sam</th>\n",
       "      <td id=\"T_d0e83_row27_col0\" class=\"data row27 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row27_col1\" class=\"data row27 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row27_col2\" class=\"data row27 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row27_col3\" class=\"data row27 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row27_col4\" class=\"data row27 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row27_col5\" class=\"data row27 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row28\" class=\"row_heading level0 row28\" >collection_date_sam</th>\n",
       "      <td id=\"T_d0e83_row28_col0\" class=\"data row28 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row28_col1\" class=\"data row28 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row28_col2\" class=\"data row28 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row28_col3\" class=\"data row28 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row28_col4\" class=\"data row28 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row28_col5\" class=\"data row28 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row29\" class=\"row_heading level0 row29\" >collector_name_sam</th>\n",
       "      <td id=\"T_d0e83_row29_col0\" class=\"data row29 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row29_col1\" class=\"data row29 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row29_col2\" class=\"data row29 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row29_col3\" class=\"data row29 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row29_col4\" class=\"data row29 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row29_col5\" class=\"data row29 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row30\" class=\"row_heading level0 row30\" >definition_for_seropositive_sample_sam</th>\n",
       "      <td id=\"T_d0e83_row30_col0\" class=\"data row30 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row30_col1\" class=\"data row30 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row30_col2\" class=\"data row30 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row30_col3\" class=\"data row30 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row30_col4\" class=\"data row30 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row30_col5\" class=\"data row30 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row31\" class=\"row_heading level0 row31\" >ena_last_update_sam_ss_dpl537</th>\n",
       "      <td id=\"T_d0e83_row31_col0\" class=\"data row31 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row31_col1\" class=\"data row31 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row31_col2\" class=\"data row31 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row31_col3\" class=\"data row31 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row31_col4\" class=\"data row31 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row31_col5\" class=\"data row31 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row32\" class=\"row_heading level0 row32\" >geographic_location__country_and_or_sea__sam</th>\n",
       "      <td id=\"T_d0e83_row32_col0\" class=\"data row32 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row32_col1\" class=\"data row32 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row32_col2\" class=\"data row32 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row32_col3\" class=\"data row32 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row32_col4\" class=\"data row32 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row32_col5\" class=\"data row32 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row33\" class=\"row_heading level0 row33\" >geographic_location__region_and_locality__sam</th>\n",
       "      <td id=\"T_d0e83_row33_col0\" class=\"data row33 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row33_col1\" class=\"data row33 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row33_col2\" class=\"data row33 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row33_col3\" class=\"data row33 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row33_col4\" class=\"data row33 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row33_col5\" class=\"data row33 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row34\" class=\"row_heading level0 row34\" >host_common_name_sam</th>\n",
       "      <td id=\"T_d0e83_row34_col0\" class=\"data row34 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row34_col1\" class=\"data row34 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row34_col2\" class=\"data row34 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row34_col3\" class=\"data row34 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row34_col4\" class=\"data row34 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row34_col5\" class=\"data row34 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row35\" class=\"row_heading level0 row35\" >host_health_state_sam</th>\n",
       "      <td id=\"T_d0e83_row35_col0\" class=\"data row35 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row35_col1\" class=\"data row35 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row35_col2\" class=\"data row35 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row35_col3\" class=\"data row35 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row35_col4\" class=\"data row35 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row35_col5\" class=\"data row35 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row36\" class=\"row_heading level0 row36\" >host_scientific_name_sam</th>\n",
       "      <td id=\"T_d0e83_row36_col0\" class=\"data row36 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row36_col1\" class=\"data row36 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row36_col2\" class=\"data row36 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row36_col3\" class=\"data row36 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row36_col4\" class=\"data row36 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row36_col5\" class=\"data row36 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row37\" class=\"row_heading level0 row37\" >host_sex_sam</th>\n",
       "      <td id=\"T_d0e83_row37_col0\" class=\"data row37 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row37_col1\" class=\"data row37 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row37_col2\" class=\"data row37 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row37_col3\" class=\"data row37 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row37_col4\" class=\"data row37 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row37_col5\" class=\"data row37 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row38\" class=\"row_heading level0 row38\" >host_subject_id_sam</th>\n",
       "      <td id=\"T_d0e83_row38_col0\" class=\"data row38 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row38_col1\" class=\"data row38 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row38_col2\" class=\"data row38 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row38_col3\" class=\"data row38 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row38_col4\" class=\"data row38 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row38_col5\" class=\"data row38 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row39\" class=\"row_heading level0 row39\" >insdc_center_alias_sam</th>\n",
       "      <td id=\"T_d0e83_row39_col0\" class=\"data row39 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row39_col1\" class=\"data row39 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row39_col2\" class=\"data row39 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row39_col3\" class=\"data row39 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row39_col4\" class=\"data row39 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row39_col5\" class=\"data row39 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row40\" class=\"row_heading level0 row40\" >isolate_sam_ss_dpl100</th>\n",
       "      <td id=\"T_d0e83_row40_col0\" class=\"data row40 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row40_col1\" class=\"data row40 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row40_col2\" class=\"data row40 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row40_col3\" class=\"data row40 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row40_col4\" class=\"data row40 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row40_col5\" class=\"data row40 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row41\" class=\"row_heading level0 row41\" >max_cycle_threshold_sam</th>\n",
       "      <td id=\"T_d0e83_row41_col0\" class=\"data row41 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row41_col1\" class=\"data row41 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row41_col2\" class=\"data row41 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row41_col3\" class=\"data row41 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row41_col4\" class=\"data row41 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row41_col5\" class=\"data row41 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row42\" class=\"row_heading level0 row42\" >min_cycle_threshold_sam</th>\n",
       "      <td id=\"T_d0e83_row42_col0\" class=\"data row42 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row42_col1\" class=\"data row42 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row42_col2\" class=\"data row42 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row42_col3\" class=\"data row42 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row42_col4\" class=\"data row42 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row42_col5\" class=\"data row42 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row43\" class=\"row_heading level0 row43\" >organism_sam</th>\n",
       "      <td id=\"T_d0e83_row43_col0\" class=\"data row43 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row43_col1\" class=\"data row43 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row43_col2\" class=\"data row43 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row43_col3\" class=\"data row43 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row43_col4\" class=\"data row43 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row43_col5\" class=\"data row43 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row44\" class=\"row_heading level0 row44\" >receipt_date_sam</th>\n",
       "      <td id=\"T_d0e83_row44_col0\" class=\"data row44 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row44_col1\" class=\"data row44 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row44_col2\" class=\"data row44 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row44_col3\" class=\"data row44 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row44_col4\" class=\"data row44 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row44_col5\" class=\"data row44 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row45\" class=\"row_heading level0 row45\" >sample_capture_status_sam</th>\n",
       "      <td id=\"T_d0e83_row45_col0\" class=\"data row45 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row45_col1\" class=\"data row45 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row45_col2\" class=\"data row45 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row45_col3\" class=\"data row45 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row45_col4\" class=\"data row45 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row45_col5\" class=\"data row45 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row46\" class=\"row_heading level0 row46\" >serotype__required_for_a_seropositive_sample__sam</th>\n",
       "      <td id=\"T_d0e83_row46_col0\" class=\"data row46 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row46_col1\" class=\"data row46 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row46_col2\" class=\"data row46 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row46_col3\" class=\"data row46 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row46_col4\" class=\"data row46 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row46_col5\" class=\"data row46 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row47\" class=\"row_heading level0 row47\" >virus_identifier_sam</th>\n",
       "      <td id=\"T_d0e83_row47_col0\" class=\"data row47 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row47_col1\" class=\"data row47 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row47_col2\" class=\"data row47 col2\" >1</td>\n",
       "      <td id=\"T_d0e83_row47_col3\" class=\"data row47 col3\" >0</td>\n",
       "      <td id=\"T_d0e83_row47_col4\" class=\"data row47 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row47_col5\" class=\"data row47 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row48\" class=\"row_heading level0 row48\" >alignment_software_exp</th>\n",
       "      <td id=\"T_d0e83_row48_col0\" class=\"data row48 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row48_col1\" class=\"data row48 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row48_col2\" class=\"data row48 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row48_col3\" class=\"data row48 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row48_col4\" class=\"data row48 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row48_col5\" class=\"data row48 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row49\" class=\"row_heading level0 row49\" >analyte_type_sam</th>\n",
       "      <td id=\"T_d0e83_row49_col0\" class=\"data row49 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row49_col1\" class=\"data row49 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row49_col2\" class=\"data row49 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row49_col3\" class=\"data row49 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row49_col4\" class=\"data row49 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row49_col5\" class=\"data row49 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row50\" class=\"row_heading level0 row50\" >avgreadlength_run</th>\n",
       "      <td id=\"T_d0e83_row50_col0\" class=\"data row50 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row50_col1\" class=\"data row50 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row50_col2\" class=\"data row50 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row50_col3\" class=\"data row50 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row50_col4\" class=\"data row50 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row50_col5\" class=\"data row50 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row51\" class=\"row_heading level0 row51\" >biospecimen_repository_sam</th>\n",
       "      <td id=\"T_d0e83_row51_col0\" class=\"data row51 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row51_col1\" class=\"data row51 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row51_col2\" class=\"data row51 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row51_col3\" class=\"data row51 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row51_col4\" class=\"data row51 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row51_col5\" class=\"data row51 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row52\" class=\"row_heading level0 row52\" >biospecimen_repository_sample_id_sam</th>\n",
       "      <td id=\"T_d0e83_row52_col0\" class=\"data row52 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row52_col1\" class=\"data row52 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row52_col2\" class=\"data row52 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row52_col3\" class=\"data row52 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row52_col4\" class=\"data row52 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row52_col5\" class=\"data row52 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row53\" class=\"row_heading level0 row53\" >body_site_sam</th>\n",
       "      <td id=\"T_d0e83_row53_col0\" class=\"data row53 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row53_col1\" class=\"data row53 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row53_col2\" class=\"data row53 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row53_col3\" class=\"data row53 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row53_col4\" class=\"data row53 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row53_col5\" class=\"data row53 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row54\" class=\"row_heading level0 row54\" >consent_code</th>\n",
       "      <td id=\"T_d0e83_row54_col0\" class=\"data row54 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row54_col1\" class=\"data row54 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row54_col2\" class=\"data row54 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row54_col3\" class=\"data row54 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row54_col4\" class=\"data row54 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row54_col5\" class=\"data row54 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row55\" class=\"row_heading level0 row55\" >coverage_run</th>\n",
       "      <td id=\"T_d0e83_row55_col0\" class=\"data row55 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row55_col1\" class=\"data row55 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row55_col2\" class=\"data row55 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row55_col3\" class=\"data row55 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row55_col4\" class=\"data row55 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row55_col5\" class=\"data row55 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row56\" class=\"row_heading level0 row56\" >gap_accession_sam</th>\n",
       "      <td id=\"T_d0e83_row56_col0\" class=\"data row56 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row56_col1\" class=\"data row56 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row56_col2\" class=\"data row56 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row56_col3\" class=\"data row56 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row56_col4\" class=\"data row56 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row56_col5\" class=\"data row56 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row57\" class=\"row_heading level0 row57\" >is_tumor_sam</th>\n",
       "      <td id=\"T_d0e83_row57_col0\" class=\"data row57 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row57_col1\" class=\"data row57 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row57_col2\" class=\"data row57 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row57_col3\" class=\"data row57 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row57_col4\" class=\"data row57 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row57_col5\" class=\"data row57 col5\" >0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row58\" class=\"row_heading level0 row58\" >molecular_data_type_sam</th>\n",
       "      <td id=\"T_d0e83_row58_col0\" class=\"data row58 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row58_col1\" class=\"data row58 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row58_col2\" class=\"data row58 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row58_col3\" class=\"data row58 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row58_col4\" class=\"data row58 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row58_col5\" class=\"data row58 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row59\" class=\"row_heading level0 row59\" >phs</th>\n",
       "      <td id=\"T_d0e83_row59_col0\" class=\"data row59 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row59_col1\" class=\"data row59 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row59_col2\" class=\"data row59 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row59_col3\" class=\"data row59 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row59_col4\" class=\"data row59 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row59_col5\" class=\"data row59 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row60\" class=\"row_heading level0 row60\" >sex_calc</th>\n",
       "      <td id=\"T_d0e83_row60_col0\" class=\"data row60 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row60_col1\" class=\"data row60 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row60_col2\" class=\"data row60 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row60_col3\" class=\"data row60 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row60_col4\" class=\"data row60 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row60_col5\" class=\"data row60 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row61\" class=\"row_heading level0 row61\" >study_design_sam</th>\n",
       "      <td id=\"T_d0e83_row61_col0\" class=\"data row61 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row61_col1\" class=\"data row61 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row61_col2\" class=\"data row61 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row61_col3\" class=\"data row61 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row61_col4\" class=\"data row61 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row61_col5\" class=\"data row61 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row62\" class=\"row_heading level0 row62\" >study_disease_sam</th>\n",
       "      <td id=\"T_d0e83_row62_col0\" class=\"data row62 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row62_col1\" class=\"data row62 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row62_col2\" class=\"data row62 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row62_col3\" class=\"data row62 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row62_col4\" class=\"data row62 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row62_col5\" class=\"data row62 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row63\" class=\"row_heading level0 row63\" >study_name_sam</th>\n",
       "      <td id=\"T_d0e83_row63_col0\" class=\"data row63 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row63_col1\" class=\"data row63 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row63_col2\" class=\"data row63 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row63_col3\" class=\"data row63 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row63_col4\" class=\"data row63 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row63_col5\" class=\"data row63 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row64\" class=\"row_heading level0 row64\" >subject_is_affected_sam</th>\n",
       "      <td id=\"T_d0e83_row64_col0\" class=\"data row64 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row64_col1\" class=\"data row64 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row64_col2\" class=\"data row64 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row64_col3\" class=\"data row64 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row64_col4\" class=\"data row64 col4\" >0</td>\n",
       "      <td id=\"T_d0e83_row64_col5\" class=\"data row64 col5\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d0e83_level0_row65\" class=\"row_heading level0 row65\" >submitted_subject_id_sam</th>\n",
       "      <td id=\"T_d0e83_row65_col0\" class=\"data row65 col0\" >0</td>\n",
       "      <td id=\"T_d0e83_row65_col1\" class=\"data row65 col1\" >0</td>\n",
       "      <td id=\"T_d0e83_row65_col2\" class=\"data row65 col2\" >0</td>\n",
       "      <td id=\"T_d0e83_row65_col3\" class=\"data row65 col3\" >1</td>\n",
       "      <td id=\"T_d0e83_row65_col4\" class=\"data row65 col4\" >1</td>\n",
       "      <td id=\"T_d0e83_row65_col5\" class=\"data row65 col5\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x17dfaad60>"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for scname in expt_schemas.keys():\n",
    "    attribute_summary_df.loc[attribute_summary_df[scname] > 0 , scname] = 1\n",
    "attribute_summary_df\n",
    "attribute_summary_df.style.background_gradient(cmap ='viridis').set_properties(**{'font-size': '1px'}) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Stable SRA metadata schema\n",
    "\n",
    "In addition to the custom set of attributes covered above there are a common set of columns/sttributes for every SRA sequencing run. \n",
    "\n",
    "The schema for the common columns is described at https://www.ncbi.nlm.nih.gov/sra/docs/sra-cloud-based-metadata-table/\n",
    "\n",
    "Note that all of the common attributes are populated for all studies. Investigating this would be a useful addition to identifying the experiment/run metadata for a given study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fetching schema\n",
      "{\n",
      "   \"$id\": \"\",\n",
      "   \"description\": \"Metadata Table (sra.metadata) contains information about the run and biological samples. Metadata Table (sra.metadata) contains information about the run and biological samples. The biological sample data is stored in two different columns.  See Record (array) column that you need to use the command UNNEST to query. See https://www.ncbi.nlm.nih.gov/sra/docs/sra-cloud-based-metadata-table/\",\n",
      "   \"$schema\": \"http://json-schema.org/draft-07/schema\",\n",
      "   \"properties\": {\n",
      "      \"acc\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"SRA Run accession in the form of SRR######## (ERR or DRR for INSDC partners)\"\n",
      "      },\n",
      "      \"assay_type\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Type of library (i.e. AMPLICON, RNA-Seq, WGS, etc)\",\n",
      "         \"oneOf\": [\n",
      "            {\n",
      "               \"title\": \"RNA-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"WGS\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"EST\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"SELEX\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"other\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"CTS\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"CLONE\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"Hi-C\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"POOLCLONE\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"GBS\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"MBD-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"miRNA-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"MeDIP-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"RIP-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"ChIP\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"MRE-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"FAIRE-seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"FL-cDNA\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"DNase-Hypersensitivity\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"ncRNA-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"WCS\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"MNase-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"AMPLICON\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"WXS\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"ChIP-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"Synthetic-Long-Read\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"WGA\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"FINISHING\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"Tethered Chromatin Conformation Capture\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"Targeted-Capture\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"ATAC-seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"RAD-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"ChIA-PET\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"OTHER\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"Bisulfite-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"Tn-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"CLONEEND\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"VALIDATION\"\n",
      "            }\n",
      "         ]\n",
      "      },\n",
      "      \"center_name\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Name of the sequencing center\"\n",
      "      },\n",
      "      \"consent\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Type of consent need to access the data (i.e. public is available to all, others are for dbGaP)\"\n",
      "      },\n",
      "      \"experiment\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"The accession in the form of SRX######## (ERX or DRX for INSDC partners)\"\n",
      "      },\n",
      "      \"sample_name\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Name of the sample\"\n",
      "      },\n",
      "      \"instrument\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Name of the sequencing instrument model\"\n",
      "      },\n",
      "      \"librarylayout\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Whether the data is SINGLE or PAIRED\",\n",
      "         \"oneOf\": [\n",
      "            {\n",
      "               \"title\": \"PAIRED\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"SINGLE\"\n",
      "            }\n",
      "         ]\n",
      "      },\n",
      "      \"libraryselection\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Library selection methodology (i.e. PCR, RANDOM, etc)\",\n",
      "         \"oneOf\": [\n",
      "            {\n",
      "               \"title\": \"MF\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"ChIP-Seq\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"unspecified\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"size fractionation\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"ChIP\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"CAGE\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"MSLL\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"other\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"DNase\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"Hybrid Selection\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"OTHER\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"RT-PCR\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"RANDOM\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"padlock probes capture method\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"PCR\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"Oligo-dT\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"Restriction Digest\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"Reduced Representation\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"Inverse rRNA\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"5-methylcytidine antibody\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"MNase\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"MBD2 protein methyl-CpG binding domain\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"HMPR\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"cDNA\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"PolyA\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"RANDOM PCR\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"MDA\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"repeat fractionation\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"RACE\"\n",
      "            }\n",
      "         ]\n",
      "      },\n",
      "      \"librarysource\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Source of the biological data (i.e. GENOMIC, METAGENOMIC, etc)\",\n",
      "         \"oneOf\": [\n",
      "            {\n",
      "               \"title\": \"SYNTHETIC\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"TRANSCRIPTOMIC\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"OTHER\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"VIRAL RNA\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"METATRANSCRIPTOMIC\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"METAGENOMIC\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"TRANSCRIPTOMIC SINGLE CELL\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"GENOMIC\"\n",
      "            },\n",
      "            {\n",
      "               \"title\": \"GENOMIC SINGLE CELL\"\n",
      "            }\n",
      "         ]\n",
      "      },\n",
      "      \"platform\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Name of the sequencing platform (i.e. ILLUMINA)\"\n",
      "      },\n",
      "      \"sample_acc\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"SRA Sample accession in the form of SRS######## (ERS or DRS for INSDC partners)\"\n",
      "      },\n",
      "      \"biosample\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"BioSample accession in the form of SAMN######## (SAMEA##### or SAMD##### for INSDC partners)\"\n",
      "      },\n",
      "      \"organism\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Scientific name of the organism that was sequenced (as found in the NCBI Taxonomy Browser)\"\n",
      "      },\n",
      "      \"sra_study\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"SRA Study accession in the form of SRP######## (ERP or DRP for INSDC partners)\"\n",
      "      },\n",
      "      \"releasedate\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"The date on which the data was released\"\n",
      "      },\n",
      "      \"bioproject\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"BioProject accession in the form of PRJNA######## (PRJEB####### or PRJDB###### for INSDC partners)\"\n",
      "      },\n",
      "      \"mbytes\": {\n",
      "         \"type\": \"integer\",\n",
      "         \"description\": \"Number of mega bytes of data in the SRA Run\"\n",
      "      },\n",
      "      \"loaddate\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"The date when the data was loaded into SRA\"\n",
      "      },\n",
      "      \"avgspotlen\": {\n",
      "         \"type\": \"integer\",\n",
      "         \"description\": \"Calculated average read length\"\n",
      "      },\n",
      "      \"mbases\": {\n",
      "         \"type\": \"integer\",\n",
      "         \"description\": \"Number of mega bases in the SRA Runs\"\n",
      "      },\n",
      "      \"insertsize\": {\n",
      "         \"type\": \"integer\",\n",
      "         \"description\": \"Submitter provided insert size\"\n",
      "      },\n",
      "      \"library_name\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"The name of the library\"\n",
      "      },\n",
      "      \"biosamplemodel_sam\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"The BioSample package/model that was picked\"\n",
      "      },\n",
      "      \"collection_date_sam\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"The collection date of the sample\"\n",
      "      },\n",
      "      \"geo_loc_name_country_calc\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Name of the country where the sample was collected\"\n",
      "      },\n",
      "      \"geo_loc_name_country_continent_calc\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Name of the continent where the sample was collected\"\n",
      "      },\n",
      "      \"geo_loc_name_sam\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Full location of collection\"\n",
      "      },\n",
      "      \"ena_first_public_run\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Date when INSDC partner record was public\"\n",
      "      },\n",
      "      \"ena_last_update_run\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Date when INSDC partner record was updated\"\n",
      "      },\n",
      "      \"sample_name_sam\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"INSDC sample name\"\n",
      "      },\n",
      "      \"datastore_filetype\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Type of files available to download from SRA\"\n",
      "      },\n",
      "      \"datastore_provider\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Locations of where the files are available to download from\"\n",
      "      },\n",
      "      \"datastore_region\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Regions of where the data is located\"\n",
      "      },\n",
      "      \"attributes\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Full list of sample attributes in a nested(array) structure\"\n",
      "      },\n",
      "      \"attributes.k\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Attribute's name\"\n",
      "      },\n",
      "      \"attributes.v\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"Attribute's value\"\n",
      "      },\n",
      "      \"jattr\": {\n",
      "         \"type\": \"string\",\n",
      "         \"description\": \"JSON based string of the sample attributes\"\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "namespace =\"dataconnect-demo\"        \n",
    "schema_name =\"nih_sra_datastore.sra.metadata\"        \n",
    "url = f\"{base}/schemas/{namespace}/{schema_name}/versions/v2\"\n",
    "print(\"fetching schema\")\n",
    "response = requests.get(url)\n",
    "response_schema = response.json()\n",
    "prettyprint(response_schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
