{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using a local GA4GH Data Connect server with dbGaP data\n",
    "\n",
    "This notebook illustrates how Data Connect is a practical and minimal approach to the following\n",
    "* Obtaining data in bulk\n",
    "* Making the data available in a Dataframe suitable for analysis\n",
    "* Selecting subsets of data according to attributes of subject and samples\n",
    "* Providing descriptions of the data via schema\n",
    "* Accessing related data elsewhere in BigQuery \n",
    "\n",
    "The examples work with a single dbGaP study - the Undiagnosed Disease Network study (phs001232). The Data Connect and FHIR servers used contain only simulated data.\n",
    "\n",
    "A local Data Connect instance is used. The examples will not run as shown without it, but the main purpose of this notebook is to show the examples to illustrate Data Connect.\n",
    "\n",
    "The examples in this notebook query on a single study. The important matter of querying across mutliple studies, within dbGaP or elsewhere, is not dealt with here. However, see related notebooks in this folder which explore how to use Data Connect with mappings in order to address the different codings or terms that may be used across different studies.\n",
    "\n",
    "The FASP library used below is at https://github.com/ga4gh/fasp-clients and may be installed as follows:\n",
    "```\n",
    "pip install git+https://github.com/ga4gh/fasp-clients\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table listing\n",
    "The following instantiates a Data Connect client and lists some of the tables available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['tutorial.onek_genomes.bdc_1000genomes',\n",
       " 'tutorial.onek_genomes.onek_drs',\n",
       " 'tutorial.onek_genomes.onek_recal_variants_drs',\n",
       " 'tutorial.onek_genomes.sra_drs_files',\n",
       " 'tutorial.onek_genomes.sra_onek_drs',\n",
       " 'tutorial.onek_genomes.ssd_drs',\n",
       " 'tutorial.onek_genomes.ssd_drs_copy',\n",
       " 'tutorial.onek_genomes.thousand_genomes_meta',\n",
       " 'tutorial.scr_gecco_susceptibility.gecco_sample_acc',\n",
       " 'tutorial.scr_gecco_susceptibility.phs001554_gecco_sddp_manifest',\n",
       " 'tutorial.scr_gecco_susceptibility.sample_attributes_multi',\n",
       " 'tutorial.scr_gecco_susceptibility.sample_multi',\n",
       " 'tutorial.scr_gecco_susceptibility.sample_run_file',\n",
       " 'tutorial.scr_gecco_susceptibility.sb_drs_index',\n",
       " 'tutorial.scr_gecco_susceptibility.subject_multi',\n",
       " 'tutorial.scr_gecco_susceptibility.subject_phenotypes_multi',\n",
       " 'tutorial.scr_icac.sample_multi',\n",
       " 'tutorial.scr_icac.sample_run_file',\n",
       " 'tutorial.scr_udn_v5.sample_multi',\n",
       " 'tutorial.scr_udn_v5.sample_run_file',\n",
       " 'tutorial.scr_udn_v5.udn-pedigree-multi',\n",
       " 'tutorial.scr_udn_v5.udn_phenotype_code',\n",
       " 'tutorial.scr_udn_v5.udn_phenotype_code2',\n",
       " 'tutorial.scr_udn_v5.udn_subject_phenotypes']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from fasp.search import DataConnectClient\n",
    "cl = DataConnectClient('http://localhost:8089/', debug=False, row_limit=100000)\n",
    "cl.list_tables('tutorial', verbose=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download a table\n",
    "If the aim is simply to provide access to and transfer of data then the Data Connect allows simple download by providing only the table name.\n",
    "\n",
    "This get_data method wraps the Data Connect /table/\\<tablename\\>/data endpoint See in [Data Connect Specification](https://github.com/ga4gh-discovery/data-connect/blob/develop/SPEC.md#table-discovery-and-browsing-examples).\n",
    "\n",
    "We use Jupyters ability to time this step.\n",
    "\n",
    "The whole table is retrieved and loaded into a dataframe in under 3 seconds. No transformation of the data is needed as it is returned in tabular json format which is easily loaded into a Dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n",
      "____Page8_______________\n",
      "____Page9_______________\n",
      "CPU times: user 141 ms, sys: 24.4 ms, total: 165 ms\n",
      "Wall time: 3.03 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbgap_subject_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>hpo_phenotypes</th>\n",
       "      <th>hpo_observed_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1827049</td>\n",
       "      <td>7ef1f8d7-8020-4781-8a4f-b9cb96913744</td>\n",
       "      <td>HP:0001962</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1827049</td>\n",
       "      <td>7ef1f8d7-8020-4781-8a4f-b9cb96913744</td>\n",
       "      <td>HP:0000765</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1827049</td>\n",
       "      <td>7ef1f8d7-8020-4781-8a4f-b9cb96913744</td>\n",
       "      <td>HP:0000925</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1827049</td>\n",
       "      <td>7ef1f8d7-8020-4781-8a4f-b9cb96913744</td>\n",
       "      <td>HP:0000927</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1827049</td>\n",
       "      <td>7ef1f8d7-8020-4781-8a4f-b9cb96913744</td>\n",
       "      <td>HP:0001155</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43658</th>\n",
       "      <td>4296808</td>\n",
       "      <td>4b5cc1f0-4bc8-4e9b-84b3-2b24cb1af42b</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43659</th>\n",
       "      <td>4296959</td>\n",
       "      <td>72c30970-392c-4dca-94f5-661dcbc21dc5</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43660</th>\n",
       "      <td>4296544</td>\n",
       "      <td>09d0c9f9-50ee-4c3c-89bf-f4510897c689</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43661</th>\n",
       "      <td>4297305</td>\n",
       "      <td>cc8e04b7-6293-4d4e-8cab-9d988a8fdf6d</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43662</th>\n",
       "      <td>4296730</td>\n",
       "      <td>379d5689-5663-4217-81b9-ab9869bdc030</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43663 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       dbgap_subject_id                            subject_id hpo_phenotypes  \\\n",
       "0               1827049  7ef1f8d7-8020-4781-8a4f-b9cb96913744     HP:0001962   \n",
       "1               1827049  7ef1f8d7-8020-4781-8a4f-b9cb96913744     HP:0000765   \n",
       "2               1827049  7ef1f8d7-8020-4781-8a4f-b9cb96913744     HP:0000925   \n",
       "3               1827049  7ef1f8d7-8020-4781-8a4f-b9cb96913744     HP:0000927   \n",
       "4               1827049  7ef1f8d7-8020-4781-8a4f-b9cb96913744     HP:0001155   \n",
       "...                 ...                                   ...            ...   \n",
       "43658           4296808  4b5cc1f0-4bc8-4e9b-84b3-2b24cb1af42b           None   \n",
       "43659           4296959  72c30970-392c-4dca-94f5-661dcbc21dc5           None   \n",
       "43660           4296544  09d0c9f9-50ee-4c3c-89bf-f4510897c689           None   \n",
       "43661           4297305  cc8e04b7-6293-4d4e-8cab-9d988a8fdf6d           None   \n",
       "43662           4296730  379d5689-5663-4217-81b9-ab9869bdc030           None   \n",
       "\n",
       "      hpo_observed_status  \n",
       "0                      no  \n",
       "1                      no  \n",
       "2                      no  \n",
       "3                      no  \n",
       "4                      no  \n",
       "...                   ...  \n",
       "43658                None  \n",
       "43659                None  \n",
       "43660                None  \n",
       "43661                None  \n",
       "43662                None  \n",
       "\n",
       "[43663 rows x 4 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "df = cl.get_data('tutorial.scr_udn_v5.udn_phenotype_code', return_type='dataframe')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The conversion to a dataframe is performed in the DataConnectClient class, but it can also be shown as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n",
      "____Page8_______________\n",
      "____Page9_______________\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "# return the data in json format\n",
    "json_data = cl.get_data('tutorial.scr_udn_v5.udn_phenotype_code', return_type='json')\n",
    "df2 = pd.DataFrame.from_dict(json_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The json data is as follows. As a python list of dictionaries it can be loaded into a dataframe via a standard method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'dbgap_subject_id': 1827049,\n",
       "  'subject_id': '7ef1f8d7-8020-4781-8a4f-b9cb96913744',\n",
       "  'hpo_phenotypes': 'HP:0001962',\n",
       "  'hpo_observed_status': 'no'},\n",
       " {'dbgap_subject_id': 1827049,\n",
       "  'subject_id': '7ef1f8d7-8020-4781-8a4f-b9cb96913744',\n",
       "  'hpo_phenotypes': 'HP:0000765',\n",
       "  'hpo_observed_status': 'no'},\n",
       " {'dbgap_subject_id': 1827049,\n",
       "  'subject_id': '7ef1f8d7-8020-4781-8a4f-b9cb96913744',\n",
       "  'hpo_phenotypes': 'HP:0000925',\n",
       "  'hpo_observed_status': 'no'}]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "json_data[0:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Access the same data via FHIR\n",
    "The dbgap_fhir module used here is as in https://github.com/ncbi/DbGaP-FHIR-API-Docs/tree/production/jupyter\n",
    "\n",
    "The module provides statistics about the amount of data retrieved and the time to do so."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dbgap_fhir import DbGapFHIR\n",
    "\n",
    "passport_path = '~/Downloads/task-specific-token.txt'\n",
    "mf = DbGapFHIR('https://dbgap-api.ncbi.nlm.nih.gov/fhir-jpa-pilot/x1'\n",
    "           ,passport=passport_path\n",
    "           ,debug=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total  Resources: 40320\n",
      "Total  Bytes: 39385045\n",
      "Total  Pages: 404\n",
      "Time elapsed 441.2335 seconds\n",
      "CPU times: user 2.6 s, sys: 549 ms, total: 3.15 s\n",
      "Wall time: 7min 21s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "conditions = mf.run_query(\"Condition\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Examine an example Condition resource"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'resourceType': 'Condition',\n",
       " 'id': 'cnd-2181175-278114-3551',\n",
       " 'meta': {'versionId': '1',\n",
       "  'lastUpdated': '2024-09-23T15:16:50.610-04:00',\n",
       "  'source': '#EDQU8nruHSXHeWea',\n",
       "  'security': [{'system': 'http://terminology.hl7.org/CodeSystem/v3-Confidentiality',\n",
       "    'code': 'U',\n",
       "    'display': 'unrestricted'}]},\n",
       " 'verificationStatus': {'coding': [{'system': 'http://hl7.org/fhir/ValueSet/condition-ver-status',\n",
       "    'code': 'confirmed'}]},\n",
       " 'code': {'coding': [{'system': 'http://human-phenotype-ontology.org',\n",
       "    'code': 'HP:0004322'}]},\n",
       " 'subject': {'reference': 'Patient/2181175'}}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conditions[4000]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transform the results to a dataframe\n",
    "The transformation to a dataframe is significantly more complicated. Specific nested items need to be identified and retrieved."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbgap_subject_id</th>\n",
       "      <th>hpo_phenotypes</th>\n",
       "      <th>hpo_observed_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2753986</td>\n",
       "      <td>HP:0002118</td>\n",
       "      <td>confirmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2753986</td>\n",
       "      <td>HP:0002194</td>\n",
       "      <td>confirmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2753986</td>\n",
       "      <td>HP:0002363</td>\n",
       "      <td>confirmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2753986</td>\n",
       "      <td>HP:0002538</td>\n",
       "      <td>confirmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2753986</td>\n",
       "      <td>HP:0010663</td>\n",
       "      <td>confirmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40315</th>\n",
       "      <td>4296515</td>\n",
       "      <td>HP:0000717</td>\n",
       "      <td>confirmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40316</th>\n",
       "      <td>4296515</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>confirmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40317</th>\n",
       "      <td>4296515</td>\n",
       "      <td>HP:0002500</td>\n",
       "      <td>confirmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40318</th>\n",
       "      <td>4296515</td>\n",
       "      <td>HP:0000545</td>\n",
       "      <td>confirmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40319</th>\n",
       "      <td>4296515</td>\n",
       "      <td>HP:0001508</td>\n",
       "      <td>confirmed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40320 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      dbgap_subject_id hpo_phenotypes hpo_observed_status\n",
       "0              2753986     HP:0002118           confirmed\n",
       "1              2753986     HP:0002194           confirmed\n",
       "2              2753986     HP:0002363           confirmed\n",
       "3              2753986     HP:0002538           confirmed\n",
       "4              2753986     HP:0010663           confirmed\n",
       "...                ...            ...                 ...\n",
       "40315          4296515     HP:0000717           confirmed\n",
       "40316          4296515     HP:0001263           confirmed\n",
       "40317          4296515     HP:0002500           confirmed\n",
       "40318          4296515     HP:0000545           confirmed\n",
       "40319          4296515     HP:0001508           confirmed\n",
       "\n",
       "[40320 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "conditions_dict = []\n",
    "for c in conditions:\n",
    "    conditions_dict.append( {\"dbgap_subject_id\" : c['subject']['reference'].split('/')[1],\n",
    "    \"hpo_phenotypes\" : c['code']['coding'][0]['code'],\n",
    "    \"hpo_observed_status\" : c['verificationStatus']['coding'][0]['code']\n",
    "        } )\n",
    "fhir_df = pd.DataFrame.from_dict(conditions_dict)\n",
    "fhir_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the two dataframes have essentially the same set of columns. The subject_id in the original udn_phenotype_code table is not in the Condition resource as it is stored elsewhere in the dbGaP FHIR representation as an alternate id for a Patient.\n",
    "\n",
    "The hpo_observed status values are different as a standard terminology was used in place of the yes/no in the original.\n",
    "\n",
    "There are fewer rows in the FHIR data. The difference is shown by the following counts by 'hpo_observed_status'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hpo_observed_status\n",
       "None     3343\n",
       "no       4512\n",
       "yes     35808\n",
       "dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby('hpo_observed_status').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hpo_observed_status\n",
       "confirmed    35808\n",
       "refuted       4512\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fhir_df.groupby('hpo_observed_status').size()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What are the missing rows?\n",
    "These data were not loaded to FHIR as Conditions because there was no specific condition (hpo_phenotype) identified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n",
      "____Page8_______________\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbgap_subject_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>hpo_phenotypes</th>\n",
       "      <th>hpo_observed_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1827062</td>\n",
       "      <td>960b9b8c-9f7a-4f41-be61-f3b603bb7076</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1827014</td>\n",
       "      <td>509dcfa7-9140-4ac6-b8f5-72b0ecdc29e0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1827023</td>\n",
       "      <td>64e45a23-3141-4069-8ee4-cabd374d35b1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1782246</td>\n",
       "      <td>5f6e9231-dcd2-4336-b620-16bcd88d1479</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1826979</td>\n",
       "      <td>08e47286-e4fe-4f16-acab-7dae368d1d67</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3338</th>\n",
       "      <td>4296808</td>\n",
       "      <td>4b5cc1f0-4bc8-4e9b-84b3-2b24cb1af42b</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3339</th>\n",
       "      <td>4296959</td>\n",
       "      <td>72c30970-392c-4dca-94f5-661dcbc21dc5</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3340</th>\n",
       "      <td>4296544</td>\n",
       "      <td>09d0c9f9-50ee-4c3c-89bf-f4510897c689</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3341</th>\n",
       "      <td>4297305</td>\n",
       "      <td>cc8e04b7-6293-4d4e-8cab-9d988a8fdf6d</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3342</th>\n",
       "      <td>4296730</td>\n",
       "      <td>379d5689-5663-4217-81b9-ab9869bdc030</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3343 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      dbgap_subject_id                            subject_id hpo_phenotypes  \\\n",
       "0              1827062  960b9b8c-9f7a-4f41-be61-f3b603bb7076           None   \n",
       "1              1827014  509dcfa7-9140-4ac6-b8f5-72b0ecdc29e0           None   \n",
       "2              1827023  64e45a23-3141-4069-8ee4-cabd374d35b1           None   \n",
       "3              1782246  5f6e9231-dcd2-4336-b620-16bcd88d1479           None   \n",
       "4              1826979  08e47286-e4fe-4f16-acab-7dae368d1d67           None   \n",
       "...                ...                                   ...            ...   \n",
       "3338           4296808  4b5cc1f0-4bc8-4e9b-84b3-2b24cb1af42b           None   \n",
       "3339           4296959  72c30970-392c-4dca-94f5-661dcbc21dc5           None   \n",
       "3340           4296544  09d0c9f9-50ee-4c3c-89bf-f4510897c689           None   \n",
       "3341           4297305  cc8e04b7-6293-4d4e-8cab-9d988a8fdf6d           None   \n",
       "3342           4296730  379d5689-5663-4217-81b9-ab9869bdc030           None   \n",
       "\n",
       "     hpo_observed_status  \n",
       "0                   None  \n",
       "1                   None  \n",
       "2                   None  \n",
       "3                   None  \n",
       "4                   None  \n",
       "...                  ...  \n",
       "3338                None  \n",
       "3339                None  \n",
       "3340                None  \n",
       "3341                None  \n",
       "3342                None  \n",
       "\n",
       "[3343 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.run_query('''select * from tutorial.scr_udn_v5.udn_phenotype_code\n",
    "where hpo_observed_status = 'None' ''', return_type='dataframe')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Do the same for the Subject Phenotypes table\n",
    "\n",
    "In this case the data is more complex\n",
    "\n",
    "First from Data Connect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n",
      "CPU times: user 59.8 ms, sys: 8.85 ms, total: 68.7 ms\n",
      "Wall time: 2.54 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbgap_subject_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>primary_symptom_category</th>\n",
       "      <th>age_onset</th>\n",
       "      <th>affection_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1827003</td>\n",
       "      <td>3c0558e0-c58d-490f-b9af-0cf426ff3e50</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>N/A</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1827081</td>\n",
       "      <td>c732713c-b08c-4ef7-b62b-4176bff922a8</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>N/A</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2181209</td>\n",
       "      <td>1dfd8fa6-0e97-48e3-b6b5-ae24cfed18ed</td>\n",
       "      <td>M</td>\n",
       "      <td>Asian</td>\n",
       "      <td>N/A</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2181199</td>\n",
       "      <td>186824e9-1b16-4af7-a40b-f71b656295a5</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>N/A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2181373</td>\n",
       "      <td>8fb2a9ef-550b-4c35-9f15-868e82c9f5d3</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>N/A</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5133</th>\n",
       "      <td>3509963</td>\n",
       "      <td>1a9e67b6-ad21-4415-90c0-71402cd7d446</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown/Not Reported Ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>7</td>\n",
       "      <td>UK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5134</th>\n",
       "      <td>3667228</td>\n",
       "      <td>4ddb2839-46be-4030-af49-bd7ff8d504df</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown/Not Reported Ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "      <td>UK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5135</th>\n",
       "      <td>4297390</td>\n",
       "      <td>e1a32592-709c-4f1e-b946-7c95fe91dfbb</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown/Not Reported Ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "      <td>UK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5136</th>\n",
       "      <td>4297308</td>\n",
       "      <td>cd1602af-4b89-4c17-9681-d313c5392ffa</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown/Not Reported Ethnicity</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>UK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5137</th>\n",
       "      <td>4297180</td>\n",
       "      <td>a9397853-b83d-49c5-9673-3c6c3e9a32b9</td>\n",
       "      <td>M</td>\n",
       "      <td>Other</td>\n",
       "      <td>Unknown/Not Reported Ethnicity</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>UK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5138 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      dbgap_subject_id                            subject_id sex   race  \\\n",
       "0              1827003  3c0558e0-c58d-490f-b9af-0cf426ff3e50   F  White   \n",
       "1              1827081  c732713c-b08c-4ef7-b62b-4176bff922a8   F  White   \n",
       "2              2181209  1dfd8fa6-0e97-48e3-b6b5-ae24cfed18ed   M  Asian   \n",
       "3              2181199  186824e9-1b16-4af7-a40b-f71b656295a5   F  White   \n",
       "4              2181373  8fb2a9ef-550b-4c35-9f15-868e82c9f5d3   F  White   \n",
       "...                ...                                   ...  ..    ...   \n",
       "5133           3509963  1a9e67b6-ad21-4415-90c0-71402cd7d446   F  White   \n",
       "5134           3667228  4ddb2839-46be-4030-af49-bd7ff8d504df   F  White   \n",
       "5135           4297390  e1a32592-709c-4f1e-b946-7c95fe91dfbb   F  White   \n",
       "5136           4297308  cd1602af-4b89-4c17-9681-d313c5392ffa   F  White   \n",
       "5137           4297180  a9397853-b83d-49c5-9673-3c6c3e9a32b9   M  Other   \n",
       "\n",
       "                           ethnicity primary_symptom_category  age_onset  \\\n",
       "0                                N/A                       17          0   \n",
       "1                                N/A                       22          0   \n",
       "2                                N/A                     None          0   \n",
       "3                                N/A                        1          1   \n",
       "4                                N/A                     None          0   \n",
       "...                              ...                      ...        ...   \n",
       "5133  Unknown/Not Reported Ethnicity                     None          7   \n",
       "5134  Unknown/Not Reported Ethnicity                     None         12   \n",
       "5135  Unknown/Not Reported Ethnicity                     None         12   \n",
       "5136  Unknown/Not Reported Ethnicity                       12          0   \n",
       "5137  Unknown/Not Reported Ethnicity                       12          0   \n",
       "\n",
       "     affection_status  \n",
       "0                   A  \n",
       "1                   A  \n",
       "2                   A  \n",
       "3                   A  \n",
       "4                   A  \n",
       "...               ...  \n",
       "5133               UK  \n",
       "5134               UK  \n",
       "5135               UK  \n",
       "5136               UK  \n",
       "5137               UK  \n",
       "\n",
       "[5138 rows x 8 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "subject_df = cl.run_query('select * from tutorial.scr_udn_v5.udn_subject_phenotypes', return_type='dataframe')\n",
    "subject_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### And via FHIR\n",
    "\n",
    "In this case each value in the original table is a separate FHIR observation resource. Each column of the original table is a variable identified by a phv variable id. We can query FHIR for observations with these ids.\n",
    "\n",
    "It's beyond the immediate scope of this example to show how to obtain the variable ids. There are various approaches. Note below that Data Connect provides a method to obtain variable ids as part of the schema for a table.\n",
    "\n",
    "Timings show that it takes significantly longer to retrieve the data via FHIR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Observation?code=phv00278118.v5.p2,phv00278119.v5.p2,phv00278120.v5.p2,phv00278121.v5.p2,phv00278122.v5.p2\n",
      "\n",
      "Total  Resources: 25690\n",
      "Total  Bytes: 24129385\n",
      "Total  Pages: 257\n",
      "Time elapsed 271.6424 seconds\n"
     ]
    }
   ],
   "source": [
    "obs_codes = ['phv00278118.v5.p2',\n",
    "             'phv00278119.v5.p2',\n",
    "             'phv00278120.v5.p2',\n",
    "             'phv00278121.v5.p2',\n",
    "             'phv00278122.v5.p2']\n",
    "query_string = f\"Observation?code={','.join(obs_codes)}\"\n",
    "print(query_string)\n",
    "print()\n",
    "observations = mf.run_query(query_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### An example observation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'resourceType': 'Observation',\n",
       " 'id': '278121-2753816-0',\n",
       " 'meta': {'versionId': '1',\n",
       "  'lastUpdated': '2024-09-23T14:50:10.931-04:00',\n",
       "  'source': '#rnrQ0dSn8j5N1MAH',\n",
       "  'security': [{'system': 'http://terminology.hl7.org/CodeSystem/v3-Confidentiality',\n",
       "    'code': 'U',\n",
       "    'display': 'unrestricted'}]},\n",
       " 'status': 'final',\n",
       " 'code': {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/CodeSystem/DbGaPConcept-VariableAccession',\n",
       "    'code': 'phv00278121.v5.p2',\n",
       "    'display': 'AGE_ONSET'}]},\n",
       " 'subject': {'reference': 'Patient/2753816'},\n",
       " 'valueString': '15'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "observations[4002]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transforming to a DataFrame\n",
    "\n",
    "In this case assembling the dataframe (table) from the individual Observation resources is more complex. Each observation will become a cell of the dataframe.\n",
    "\n",
    "Each row of the dataframe will be the observations for a given Patient.\n",
    "\n",
    "In this case, from the knowledge of the data obtained above we know that there is only one value of each Observation for each patient. This simplifies the transformation but that won't always be the case. For example, if the HPO phenotype data above were represented as observations the rows could not be assembled based on dbgap_subject_id as below because there multiple value pairs per subject. \n",
    "\n",
    "Covering all possible options for transforming FHIR data to a dataframe suitable for analysis is not a simple challenge."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AFFECTION_STATUS</th>\n",
       "      <th>ETHNICITY</th>\n",
       "      <th>RACE</th>\n",
       "      <th>PRIMARY_SYMPTOM_CATEGORY</th>\n",
       "      <th>AGE_ONSET</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient/2753407</th>\n",
       "      <td>UA</td>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>None</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient/2753413</th>\n",
       "      <td>UA</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>None</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient/1826993</th>\n",
       "      <td>A</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>10</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient/1827029</th>\n",
       "      <td>A</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>None</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient/1827049</th>\n",
       "      <td>UA</td>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>None</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient/4296808</th>\n",
       "      <td>A</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>None</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient/4296544</th>\n",
       "      <td>UA</td>\n",
       "      <td>Unknown/Not Reported Ethnicity</td>\n",
       "      <td>White</td>\n",
       "      <td>20</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient/4296730</th>\n",
       "      <td>A</td>\n",
       "      <td>None</td>\n",
       "      <td>White</td>\n",
       "      <td>None</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient/4297305</th>\n",
       "      <td>A</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>None</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient/4296515</th>\n",
       "      <td>UA</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>12</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5138 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                AFFECTION_STATUS                       ETHNICITY   RACE  \\\n",
       "Patient/2753407               UA              Hispanic or Latino  White   \n",
       "Patient/2753413               UA          Not Hispanic or Latino  White   \n",
       "Patient/1826993                A          Not Hispanic or Latino  White   \n",
       "Patient/1827029                A          Not Hispanic or Latino  White   \n",
       "Patient/1827049               UA              Hispanic or Latino  White   \n",
       "...                          ...                             ...    ...   \n",
       "Patient/4296808                A          Not Hispanic or Latino  White   \n",
       "Patient/4296544               UA  Unknown/Not Reported Ethnicity  White   \n",
       "Patient/4296730                A                            None  White   \n",
       "Patient/4297305                A          Not Hispanic or Latino  White   \n",
       "Patient/4296515               UA          Not Hispanic or Latino  White   \n",
       "\n",
       "                PRIMARY_SYMPTOM_CATEGORY AGE_ONSET  \n",
       "Patient/2753407                     None        76  \n",
       "Patient/2753413                     None        15  \n",
       "Patient/1826993                       10        34  \n",
       "Patient/1827029                     None        66  \n",
       "Patient/1827049                     None        78  \n",
       "...                                  ...       ...  \n",
       "Patient/4296808                     None        30  \n",
       "Patient/4296544                       20        88  \n",
       "Patient/4296730                     None        16  \n",
       "Patient/4297305                     None        29  \n",
       "Patient/4296515                       12        67  \n",
       "\n",
       "[5138 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from collections import Counter\n",
    "\n",
    "\n",
    "patient_observations_dict = {}\n",
    "for r in observations:\n",
    "\n",
    "    subject_id = r['subject']['reference']\n",
    "    obs_display_name = r['code']['coding'][0]['display']\n",
    "    obs_code = r['code']['coding'][0]['code']\n",
    "    if 'valueQuantity' in r:\n",
    "        value_text = r['valueQuantity']['value']\n",
    "    elif 'valueCodeableConcept' in r:\n",
    "         value_text = r['valueCodeableConcept']['coding'][0]['display']\n",
    "    elif 'valueString' in r:\n",
    "         value_text = r['valueString']\n",
    "    else:\n",
    "        value_text = 'unknown'\n",
    "        \n",
    "    if subject_id not in patient_observations_dict:\n",
    "        patient_observations_dict[subject_id] = {obs_display_name: value_text}\n",
    "    else:\n",
    "        patient_observations_dict[subject_id][obs_display_name] = value_text\n",
    "pd.set_option(\"display.max_rows\", 30, \"display.max_columns\", None)\n",
    "patient_df = pd.DataFrame.from_dict(patient_observations_dict, orient='index')\n",
    "display(patient_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Search/query capability\n",
    "The above considered retrieval of the data in bulk for a single table of data.\n",
    "\n",
    "Data Connect also provides the ability to run SQL queries against the tables.\n",
    "\n",
    "In this example subjects are selected in which a specific phenotype (HPO code) were observed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n",
      "____Page8_______________\n",
      "____Page9_______________\n",
      "____Page10_______________\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbgap_subject_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>hpo_phenotypes</th>\n",
       "      <th>hpo_observed_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1826993</td>\n",
       "      <td>274559bb-b938-471a-92bb-b39b773399c1</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1827029</td>\n",
       "      <td>6b537dfc-6ad8-45ec-bd16-b0d7b58284e4</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1827049</td>\n",
       "      <td>7ef1f8d7-8020-4781-8a4f-b9cb96913744</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1827065</td>\n",
       "      <td>9a0686ba-15ad-4c5a-bc26-86a4ed74e46d</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2181265</td>\n",
       "      <td>469290e0-d885-4b24-97eb-e2cf629a39f0</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1427</th>\n",
       "      <td>4297136</td>\n",
       "      <td>9fbf41e6-2530-48e0-97f2-7a13472ffac9</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>4296710</td>\n",
       "      <td>3221cedc-0980-444f-830d-e7c7b071fd8e</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>4297160</td>\n",
       "      <td>a4b716f1-3b15-4be2-a58a-af441ccaddda</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1430</th>\n",
       "      <td>4296911</td>\n",
       "      <td>65c733f2-d389-4467-ba83-cfd947cbe16e</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431</th>\n",
       "      <td>4296515</td>\n",
       "      <td>0203d316-85d2-4b1c-b5f1-908aed7ca379</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1432 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      dbgap_subject_id                            subject_id hpo_phenotypes  \\\n",
       "0              1826993  274559bb-b938-471a-92bb-b39b773399c1     HP:0001263   \n",
       "1              1827029  6b537dfc-6ad8-45ec-bd16-b0d7b58284e4     HP:0001263   \n",
       "2              1827049  7ef1f8d7-8020-4781-8a4f-b9cb96913744     HP:0001263   \n",
       "3              1827065  9a0686ba-15ad-4c5a-bc26-86a4ed74e46d     HP:0001263   \n",
       "4              2181265  469290e0-d885-4b24-97eb-e2cf629a39f0     HP:0001263   \n",
       "...                ...                                   ...            ...   \n",
       "1427           4297136  9fbf41e6-2530-48e0-97f2-7a13472ffac9     HP:0001263   \n",
       "1428           4296710  3221cedc-0980-444f-830d-e7c7b071fd8e     HP:0001263   \n",
       "1429           4297160  a4b716f1-3b15-4be2-a58a-af441ccaddda     HP:0001263   \n",
       "1430           4296911  65c733f2-d389-4467-ba83-cfd947cbe16e     HP:0001263   \n",
       "1431           4296515  0203d316-85d2-4b1c-b5f1-908aed7ca379     HP:0001263   \n",
       "\n",
       "     hpo_observed_status  \n",
       "0                    yes  \n",
       "1                    yes  \n",
       "2                    yes  \n",
       "3                    yes  \n",
       "4                    yes  \n",
       "...                  ...  \n",
       "1427                 yes  \n",
       "1428                 yes  \n",
       "1429                 yes  \n",
       "1430                 yes  \n",
       "1431                 yes  \n",
       "\n",
       "[1432 rows x 4 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.run_query('''select * \n",
    "             from tutorial.scr_udn_v5.udn_phenotype_code\n",
    "             where hpo_phenotypes = 'HP:0001263' and hpo_observed_status = 'yes' ''',\n",
    "             return_type='dataframe')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Queries can treat numbers as numbers\n",
    "\n",
    "This example selects subjects where the age of onset of the primary symptom was greater than 11."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbgap_subject_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>primary_symptom_category</th>\n",
       "      <th>age_onset</th>\n",
       "      <th>affection_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2753288</td>\n",
       "      <td>0a9cb7fe-9003-4b0b-9493-2eb155dd7c4d</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>N/A</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3040012</td>\n",
       "      <td>d3546678-c423-40f9-b84d-8d4f542f5e72</td>\n",
       "      <td>M</td>\n",
       "      <td>White</td>\n",
       "      <td>N/A</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3039265</td>\n",
       "      <td>52ce04d1-1dc9-4850-b7fe-c0df458cd379</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>N/A</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3039769</td>\n",
       "      <td>a75358a3-bfb7-4a36-847d-5f4d83f56a43</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>N/A</td>\n",
       "      <td>None</td>\n",
       "      <td>32</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3510369</td>\n",
       "      <td>6717dc25-bd18-42d6-8a95-9c2debf053ef</td>\n",
       "      <td>M</td>\n",
       "      <td>White</td>\n",
       "      <td>N/A</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>3039720</td>\n",
       "      <td>9e55b289-55ff-4fb4-b237-7b63fa957871</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown/Not Reported Ethnicity</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>3039867</td>\n",
       "      <td>b877251f-e074-45c7-8727-4451cc48224e</td>\n",
       "      <td>M</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown/Not Reported Ethnicity</td>\n",
       "      <td>12</td>\n",
       "      <td>17</td>\n",
       "      <td>UA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>3039629</td>\n",
       "      <td>8e89658e-5eaf-47b7-b528-d2b325b4be1e</td>\n",
       "      <td>M</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown/Not Reported Ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "      <td>UA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>3667228</td>\n",
       "      <td>4ddb2839-46be-4030-af49-bd7ff8d504df</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown/Not Reported Ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "      <td>UK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>4297390</td>\n",
       "      <td>e1a32592-709c-4f1e-b946-7c95fe91dfbb</td>\n",
       "      <td>F</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown/Not Reported Ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "      <td>UK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     dbgap_subject_id                            subject_id sex   race  \\\n",
       "0             2753288  0a9cb7fe-9003-4b0b-9493-2eb155dd7c4d   F  White   \n",
       "1             3040012  d3546678-c423-40f9-b84d-8d4f542f5e72   M  White   \n",
       "2             3039265  52ce04d1-1dc9-4850-b7fe-c0df458cd379   F  White   \n",
       "3             3039769  a75358a3-bfb7-4a36-847d-5f4d83f56a43   F  White   \n",
       "4             3510369  6717dc25-bd18-42d6-8a95-9c2debf053ef   M  White   \n",
       "..                ...                                   ...  ..    ...   \n",
       "97            3039720  9e55b289-55ff-4fb4-b237-7b63fa957871   F  White   \n",
       "98            3039867  b877251f-e074-45c7-8727-4451cc48224e   M  White   \n",
       "99            3039629  8e89658e-5eaf-47b7-b528-d2b325b4be1e   M  White   \n",
       "100           3667228  4ddb2839-46be-4030-af49-bd7ff8d504df   F  White   \n",
       "101           4297390  e1a32592-709c-4f1e-b946-7c95fe91dfbb   F  White   \n",
       "\n",
       "                          ethnicity primary_symptom_category  age_onset  \\\n",
       "0                               N/A                        5         12   \n",
       "1                               N/A                     None         12   \n",
       "2                               N/A                        1         12   \n",
       "3                               N/A                     None         32   \n",
       "4                               N/A                     None         12   \n",
       "..                              ...                      ...        ...   \n",
       "97   Unknown/Not Reported Ethnicity                        2         17   \n",
       "98   Unknown/Not Reported Ethnicity                       12         17   \n",
       "99   Unknown/Not Reported Ethnicity                     None         12   \n",
       "100  Unknown/Not Reported Ethnicity                     None         12   \n",
       "101  Unknown/Not Reported Ethnicity                     None         12   \n",
       "\n",
       "    affection_status  \n",
       "0                  A  \n",
       "1                  A  \n",
       "2                  A  \n",
       "3                  A  \n",
       "4                  A  \n",
       "..               ...  \n",
       "97                 A  \n",
       "98                UA  \n",
       "99                UA  \n",
       "100               UK  \n",
       "101               UK  \n",
       "\n",
       "[102 rows x 8 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "cl.run_query('''select * from tutorial.scr_udn_v5.udn_subject_phenotypes\n",
    "where age_onset > 11 ''', return_type='dataframe')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What is the unit for age? Using the table schema\n",
    "This might raise the question in the query above whether the value of 11 for the age of onset is 11 years, months, days etc.\n",
    "\n",
    "Data Connect allows a machine readable schema for each table to be provided as a JSON Schema. With an extension to provide the unit of measure we can see that the unit for age_onset is years.\n",
    "\n",
    "Codings for enumerated values are also provided.\n",
    "\n",
    "In this Data Connect implementation an extension was created to produce the schema automatically from information provided by dbGaP submitters. No additional curation was required. Data submitters are required to provide a data dictionary for each of their data tables which is then stored as XML. The extension transforms the XML data dictionary for each dbGaP table to a JSON schema.\n",
    "\n",
    "The list_table_info method of the DataConnectClient class is a simple wrapper around the /table/\\<tablename\\>/info endpoint for the Data Connect service."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tabletutorial.scr_udn_v5.udn_subject_phenotypes_\n",
      "{\n",
      "   \"name\": \"tutorial.scr_udn_v5.udn_subject_phenotypes\",\n",
      "   \"data_model\": {\n",
      "      \"description\": \"\",\n",
      "      \"$id\": \"dbgap:pht006079.v5\",\n",
      "      \"properties\": {\n",
      "         \"SUBJECT_ID\": {\n",
      "            \"$id\": \"dbgap:phv00278116.v5\",\n",
      "            \"description\": \"De-identified subject ID\",\n",
      "            \"type\": \"string\"\n",
      "         },\n",
      "         \"SEX\": {\n",
      "            \"$id\": \"dbgap:phv00278117.v5\",\n",
      "            \"description\": \"Biological sex\",\n",
      "            \"type\": \"encoded value\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"F\",\n",
      "                  \"title\": \"Female\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"M\",\n",
      "                  \"title\": \"Male\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"O\",\n",
      "                  \"title\": \"Other\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"RACE\": {\n",
      "            \"$id\": \"dbgap:phv00278118.v5\",\n",
      "            \"description\": \"Race of subject\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"American Indian or Alaska Native\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Asian\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Black or African American\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Native Hawaiian or Other Pacific Islander\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Other\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"White\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"AGE_ONSET\": {\n",
      "            \"$id\": \"dbgap:phv00278121.v5\",\n",
      "            \"description\": \"Onset of disease symptoms\",\n",
      "            \"type\": \"integer, encoded value\",\n",
      "            \"$unit\": \"years\"\n",
      "         },\n",
      "         \"ETHNICITY\": {\n",
      "            \"$id\": \"dbgap:phv00278119.v5\",\n",
      "            \"description\": \"Ethnicity of subject\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"Hispanic or Latino\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Not Hispanic or Latino\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Unknown/Not Reported Ethnicity\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"PRIMARY_SYMPTOM_CATEGORY\": {\n",
      "            \"$id\": \"dbgap:phv00278120.v5\",\n",
      "            \"description\": \"Primary category of symptoms\",\n",
      "            \"type\": \"encoded value\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"1\",\n",
      "                  \"title\": \"Allergies and Disorders of The Immune System\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"2\",\n",
      "                  \"title\": \"Cardiology and vascular conditions (heart, artery, vein, and lymph disorders)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"3\",\n",
      "                  \"title\": \"Dentistry and craniofacial (bones of head and face)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"4\",\n",
      "                  \"title\": \"Dermatology (skin diseases and disorders)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"5\",\n",
      "                  \"title\": \"Endocrinology (disorder of the endocrine glands and hormones)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"6\",\n",
      "                  \"title\": \"Gastroenterology (disorder of the stomach and intestines)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"7\",\n",
      "                  \"title\": \"Gynecology and reproductive medicine\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"8\",\n",
      "                  \"title\": \"Hematology (blood diseases and disorders)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"9\",\n",
      "                  \"title\": \"Infectious diseases\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"10\",\n",
      "                  \"title\": \"Musculoskeletal and orthopedics (structural and functional disorders of muscles, bones, and joints)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"11\",\n",
      "                  \"title\": \"Nephrology (kidney diseases and disorders)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"12\",\n",
      "                  \"title\": \"Neurology (disorders of the nervous system, including brain and spinal cord)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"13\",\n",
      "                  \"title\": \"Oncology (Tumors and cancer)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"14\",\n",
      "                  \"title\": \"Ophthalmology (Eye disorders and diseases)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"15\",\n",
      "                  \"title\": \"Psychiatry\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"16\",\n",
      "                  \"title\": \"Pulmonology (Lung disorders and diseases)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"17\",\n",
      "                  \"title\": \"Rheumatology (immune disorders of the joints, muscles, and ligaments)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"18\",\n",
      "                  \"title\": \"Urology\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"20\",\n",
      "                  \"title\": \"N/A\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"22\",\n",
      "                  \"title\": \"Other\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"AFFECTION_STATUS\": {\n",
      "            \"$id\": \"dbgap:phv00278122.v5\",\n",
      "            \"description\": \"Affection status of subject\",\n",
      "            \"type\": \"encoded value\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"A\",\n",
      "                  \"title\": \"Affected\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"UA\",\n",
      "                  \"title\": \"Unaffected\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"UK\",\n",
      "                  \"title\": \"Unknown\"\n",
      "               }\n",
      "            ]\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x168ce2400>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.list_table_info('tutorial.scr_udn_v5.udn_subject_phenotypes', verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is worth considering the FHIR syntax that would be necessary for equivalent queries. In some cases it will be realtively straightforward e.g. where both attributes of the Condition are in the same resource as in the example above. In other cases e.g. in the case where cells of the table are in separate cells the FHIR syntax will be relatively challenging. Even where the 'join' of Observations exists the query syntax is not straightforward.   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query SRA 'metadata' for a UDN sample\n",
    "\n",
    "Other data sources can also be made available through Data Connect from their original source.\n",
    "\n",
    "It was straightfoward to make available via Data Connect the SRA metadata table that NCBI provides in BigQuery."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the table list\n",
      "____Page1_______________\n",
      "nih_sra_datastore.sra.metadata\n",
      "nih_sra_datastore.sra.metadata_clustered\n",
      "nih_sra_datastore.sra_tax_analysis_tool.kmer\n",
      "nih_sra_datastore.sra_tax_analysis_tool.tax_analysis\n",
      "nih_sra_datastore.sra_tax_analysis_tool.tax_analysis_clustered\n",
      "nih_sra_datastore.sra_tax_analysis_tool.tax_analysis_info\n",
      "nih_sra_datastore.sra_tax_analysis_tool.taxonomy\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['nih_sra_datastore.sra.metadata',\n",
       " 'nih_sra_datastore.sra.metadata_clustered',\n",
       " 'nih_sra_datastore.sra_tax_analysis_tool.kmer',\n",
       " 'nih_sra_datastore.sra_tax_analysis_tool.tax_analysis',\n",
       " 'nih_sra_datastore.sra_tax_analysis_tool.tax_analysis_clustered',\n",
       " 'nih_sra_datastore.sra_tax_analysis_tool.tax_analysis_info',\n",
       " 'nih_sra_datastore.sra_tax_analysis_tool.taxonomy']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.list_tables('nih_sra_datastore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Query run data for a UDN sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n",
      "____Page8_______________\n",
      "____Page9_______________\n",
      "____Page10_______________\n",
      "____Page11_______________\n",
      "____Page12_______________\n",
      "____Page13_______________\n",
      "____Page14_______________\n",
      "____Page15_______________\n",
      "____Page16_______________\n",
      "____Page17_______________\n",
      "____Page18_______________\n",
      "____Page19_______________\n",
      "____Page20_______________\n",
      "____Page21_______________\n",
      "____Page22_______________\n",
      "____Page23_______________\n",
      "____Page24_______________\n",
      "____Page25_______________\n",
      "____Page26_______________\n",
      "____Page27_______________\n",
      "____Page28_______________\n",
      "____Page29_______________\n",
      "____Page30_______________\n",
      "____Page31_______________\n",
      "____Page32_______________\n",
      "____Page33_______________\n",
      "____Page34_______________\n",
      "____Page35_______________\n",
      "____Page36_______________\n",
      "____Page37_______________\n",
      "____Page38_______________\n",
      "____Page39_______________\n",
      "____Page40_______________\n",
      "____Page41_______________\n",
      "____Page42_______________\n",
      "____Page43_______________\n",
      "____Page44_______________\n",
      "____Page45_______________\n",
      "____Page46_______________\n",
      "____Page47_______________\n",
      "____Page48_______________\n",
      "____Page49_______________\n",
      "____Page50_______________\n",
      "____Page51_______________\n",
      "____Page52_______________\n",
      "____Page53_______________\n",
      "____Page54_______________\n",
      "____Page55_______________\n",
      "____Page56_______________\n",
      "____Page57_______________\n",
      "____Page58_______________\n",
      "____Page59_______________\n",
      "____Page60_______________\n",
      "____Page61_______________\n",
      "____Page62_______________\n",
      "____Page63_______________\n",
      "____Page64_______________\n",
      "____Page65_______________\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assay_type</th>\n",
       "      <th>experiment</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>avgspotlen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WGS</td>\n",
       "      <td>SRX4836919</td>\n",
       "      <td>HiSeq X Ten</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>302</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  assay_type  experiment   instrument librarysource  avgspotlen\n",
       "0        WGS  SRX4836919  HiSeq X Ten       GENOMIC         302"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query1 = '''SELECT assay_type, experiment, instrument, librarysource, avgspotlen\n",
    "FROM nih_sra_datastore.sra.metadata\n",
    "where acc='SRR8006118'\n",
    "'''\n",
    "\n",
    "res = cl.run_query(query1, return_type='dataframe')\n",
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Schema for SRA metadata\n",
    "\n",
    "While no machine readable schema was available for the SRA metadata table it was straightforward to create one based on the information made available at\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/sra/docs/sra-cloud-based-metadata-table/\n",
    "\n",
    "Data Connect provides a standard for providing schema for data (such as SRA) that is unlikely to be provided in FHIR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablenih_sra_datastore.sra.metadata_\n",
      "{\n",
      "   \"name\": \"nih_sra_datastore.sra.metadata\",\n",
      "   \"description\": \"Metadata Table (sra.metadata) contains information about the run and biological samples. Metadata Table (sra.metadata) contains information about the run and biological samples. The biological sample data is stored in two different columns.  See Record (array) column that you need to use the command UNNEST to query. See https://www.ncbi.nlm.nih.gov/sra/docs/sra-cloud-based-metadata-table/\",\n",
      "   \"data_model\": {\n",
      "      \"$id\": \"\",\n",
      "      \"description\": \"Metadata Table (sra.metadata) contains information about the run and biological samples. Metadata Table (sra.metadata) contains information about the run and biological samples. The biological sample data is stored in two different columns.  See Record (array) column that you need to use the command UNNEST to query. See https://www.ncbi.nlm.nih.gov/sra/docs/sra-cloud-based-metadata-table/\",\n",
      "      \"$schema\": \"http://json-schema.org/draft-07/schema\",\n",
      "      \"properties\": {\n",
      "         \"acc\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"SRA Run accession in the form of SRR######## (ERR or DRR for INSDC partners)\"\n",
      "         },\n",
      "         \"assay_type\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Type of library (i.e. AMPLICON, RNA-Seq, WGS, etc)\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"RNA-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"WGS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"EST\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"SELEX\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"other\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"CTS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"CLONE\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Hi-C\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"POOLCLONE\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"GBS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MBD-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"miRNA-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MeDIP-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RIP-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ChIP\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MRE-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"FAIRE-seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"FL-cDNA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"DNase-Hypersensitivity\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ncRNA-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"WCS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MNase-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"AMPLICON\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"WXS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ChIP-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Synthetic-Long-Read\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"WGA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"FINISHING\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Tethered Chromatin Conformation Capture\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Targeted-Capture\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ATAC-seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RAD-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ChIA-PET\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"OTHER\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Bisulfite-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Tn-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"CLONEEND\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"VALIDATION\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"center_name\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the sequencing center\"\n",
      "         },\n",
      "         \"consent\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Type of consent need to access the data (i.e. public is available to all, others are for dbGaP)\"\n",
      "         },\n",
      "         \"experiment\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The accession in the form of SRX######## (ERX or DRX for INSDC partners)\"\n",
      "         },\n",
      "         \"sample_name\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the sample\"\n",
      "         },\n",
      "         \"instrument\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the sequencing instrument model\"\n",
      "         },\n",
      "         \"librarylayout\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Whether the data is SINGLE or PAIRED\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"PAIRED\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"SINGLE\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"libraryselection\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Library selection methodology (i.e. PCR, RANDOM, etc)\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"MF\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ChIP-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"unspecified\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"size fractionation\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ChIP\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"CAGE\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MSLL\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"other\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"DNase\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Hybrid Selection\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"OTHER\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RT-PCR\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RANDOM\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"padlock probes capture method\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"PCR\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Oligo-dT\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Restriction Digest\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Reduced Representation\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Inverse rRNA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"5-methylcytidine antibody\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MNase\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MBD2 protein methyl-CpG binding domain\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"HMPR\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"cDNA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"PolyA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RANDOM PCR\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MDA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"repeat fractionation\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RACE\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"librarysource\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Source of the biological data (i.e. GENOMIC, METAGENOMIC, etc)\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"SYNTHETIC\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"TRANSCRIPTOMIC\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"OTHER\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"VIRAL RNA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"METATRANSCRIPTOMIC\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"METAGENOMIC\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"TRANSCRIPTOMIC SINGLE CELL\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"GENOMIC\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"GENOMIC SINGLE CELL\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"platform\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the sequencing platform (i.e. ILLUMINA)\"\n",
      "         },\n",
      "         \"sample_acc\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"SRA Sample accession in the form of SRS######## (ERS or DRS for INSDC partners)\"\n",
      "         },\n",
      "         \"biosample\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"BioSample accession in the form of SAMN######## (SAMEA##### or SAMD##### for INSDC partners)\"\n",
      "         },\n",
      "         \"organism\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Scientific name of the organism that was sequenced (as found in the NCBI Taxonomy Browser)\"\n",
      "         },\n",
      "         \"sra_study\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"SRA Study accession in the form of SRP######## (ERP or DRP for INSDC partners)\"\n",
      "         },\n",
      "         \"releasedate\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The date on which the data was released\"\n",
      "         },\n",
      "         \"bioproject\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"BioProject accession in the form of PRJNA######## (PRJEB####### or PRJDB###### for INSDC partners)\"\n",
      "         },\n",
      "         \"mbytes\": {\n",
      "            \"type\": \"integer\",\n",
      "            \"description\": \"Number of mega bytes of data in the SRA Run\"\n",
      "         },\n",
      "         \"loaddate\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The date when the data was loaded into SRA\"\n",
      "         },\n",
      "         \"avgspotlen\": {\n",
      "            \"type\": \"integer\",\n",
      "            \"description\": \"Calculated average read length\"\n",
      "         },\n",
      "         \"mbases\": {\n",
      "            \"type\": \"integer\",\n",
      "            \"description\": \"Number of mega bases in the SRA Runs\"\n",
      "         },\n",
      "         \"insertsize\": {\n",
      "            \"type\": \"integer\",\n",
      "            \"description\": \"Submitter provided insert size\"\n",
      "         },\n",
      "         \"library_name\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The name of the library\"\n",
      "         },\n",
      "         \"biosamplemodel_sam\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The BioSample package/model that was picked\"\n",
      "         },\n",
      "         \"collection_date_sam\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The collection date of the sample\"\n",
      "         },\n",
      "         \"geo_loc_name_country_calc\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the country where the sample was collected\"\n",
      "         },\n",
      "         \"geo_loc_name_country_continent_calc\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the continent where the sample was collected\"\n",
      "         },\n",
      "         \"geo_loc_name_sam\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Full location of collection\"\n",
      "         },\n",
      "         \"ena_first_public_run\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Date when INSDC partner record was public\"\n",
      "         },\n",
      "         \"ena_last_update_run\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Date when INSDC partner record was updated\"\n",
      "         },\n",
      "         \"sample_name_sam\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"INSDC sample name\"\n",
      "         },\n",
      "         \"datastore_filetype\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Type of files available to download from SRA\"\n",
      "         },\n",
      "         \"datastore_provider\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Locations of where the files are available to download from\"\n",
      "         },\n",
      "         \"datastore_region\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Regions of where the data is located\"\n",
      "         },\n",
      "         \"attributes\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Full list of sample attributes in a nested(array) structure\"\n",
      "         },\n",
      "         \"attributes.k\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Attribute's name\"\n",
      "         },\n",
      "         \"attributes.v\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Attribute's value\"\n",
      "         },\n",
      "         \"jattr\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"JSON based string of the sample attributes\"\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x175860310>"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.list_table_info('nih_sra_datastore.sra.metadata', verbose=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
