{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using a local GA4GH Data Connect server\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/forei/imfenvs/fasp/lib/python3.9/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.\n",
      "  warnings.warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Retrieving the table list_\n",
      "____Page1_______________\n",
      "bigquery.1000genomes.1000genomes_drs\n",
      "bigquery.1000genomes.aligned_read\n",
      "bigquery.1000genomes.aliquot\n",
      "bigquery.1000genomes.bdc_1000genomes\n",
      "bigquery.1000genomes.demographics\n",
      "bigquery.1000genomes.germline_variation_index\n",
      "bigquery.1000genomes.phenotype_drs\n",
      "bigquery.1000genomes.read_group\n",
      "bigquery.1000genomes.sample\n",
      "bigquery.1000genomes.simple_germline_variation\n",
      "bigquery.1000genomes.ssd_drs\n",
      "bigquery.1000genomes.ssd_drs_table\n",
      "bigquery.1000genomes.subject\n",
      "bigquery.100genomes.import2\n",
      "bigquery.100genomes.import3\n",
      "bigquery.100genomes.testimport\n",
      "bigquery.blog_unnest.firebase_raw\n",
      "bigquery.cda.cdav3\n",
      "bigquery.cidr_gecco_colorectal_cancer_gxe.sample\n",
      "bigquery.cidr_gecco_colorectal_cancer_gxe.sample_attributes\n",
      "bigquery.cidr_gecco_colorectal_cancer_gxe.subject\n",
      "bigquery.cidr_gecco_colorectal_cancer_gxe.subject_phenotypes\n",
      "bigquery.copdgene.aligned_read\n",
      "bigquery.copdgene.aliquot\n",
      "bigquery.copdgene.copd_drs\n",
      "bigquery.copdgene.copd_gene__hmb_vcf\n",
      "bigquery.copdgene.copdgene_hmb_female_90_94kg\n",
      "bigquery.copdgene.copdgene_pfb\n",
      "bigquery.copdgene.phenotype_drs\n",
      "bigquery.copdgene.read_group\n",
      "bigquery.copdgene.sample\n",
      "bigquery.copdgene.sample_multi\n",
      "bigquery.copdgene.subject\n",
      "bigquery.copdgene.subject_images_hmb\n",
      "bigquery.copdgene.subject_multi\n",
      "bigquery.copdgene.subject_phenotypes_hmb\n",
      "bigquery.cptac.breastsubjectclinicalphenotypes\n",
      "bigquery.cptac.sample_attributes\n",
      "bigquery.esp_lunggo_copdgene.sample_attributes_hmb\n",
      "bigquery.esp_lunggo_copdgene.subject_phenotypes_hmb\n",
      "bigquery.fasp_scrambled.scr_egapancreatic_sample\n",
      "bigquery.gecco_crc_susceptibility.public_sample\n",
      "bigquery.gecco_crc_susceptibility.sample_attributes\n",
      "bigquery.gecco_crc_susceptibility.sample_attributes_cadm\n",
      "bigquery.gecco_crc_susceptibility.sample_attributes_ds_cc_npu_gso\n",
      "bigquery.gecco_crc_susceptibility.sample_attributes_gru\n",
      "bigquery.gecco_crc_susceptibility.sample_attributes_hmb_irb\n",
      "bigquery.gecco_crc_susceptibility.sample_attributes_hmb_irb_npu\n",
      "bigquery.gecco_crc_susceptibility.sample_multi\n",
      "bigquery.gecco_crc_susceptibility.studies_1554\n",
      "bigquery.gecco_crc_susceptibility.subject_multi\n",
      "bigquery.gecco_crc_susceptibility.subject_phenotypes\n",
      "bigquery.gecco_crc_susceptibility.subject_phenotypes_cadm\n",
      "bigquery.gecco_crc_susceptibility.subject_phenotypes_ds_cc_npu_gso\n",
      "bigquery.gecco_crc_susceptibility.subject_phenotypes_gru\n",
      "bigquery.gecco_crc_susceptibility.subject_phenotypes_hmb_irb\n",
      "bigquery.gecco_crc_susceptibility.subject_phenotypes_hmb_irb_npu\n",
      "bigquery.gtex8.program\n",
      "bigquery.gtex8.project\n",
      "bigquery.gtex8.sample\n",
      "bigquery.gtex8.sequencing\n",
      "bigquery.gtex8.sraruntable_breast_cram\n",
      "bigquery.gtex8.sraruntable_nocram\n",
      "bigquery.gtex8.sraruntable_nocram_nobam\n",
      "bigquery.gtex8.subject\n",
      "bigquery.icdc.cotc_sample\n",
      "bigquery.kidsfirst.neuroblastoma_participant_family\n",
      "bigquery.kidsfirst.pediatric_brain_tumor_atlas\n",
      "bigquery.lccc_1108.lccc_drs_map\n",
      "bigquery.lccc_1108.sample\n",
      "bigquery.lccc_1108.sample_attributes\n",
      "bigquery.lccc_1108.subject\n",
      "bigquery.lccc_1108.subject_phenotypes\n",
      "bigquery.mapped.age_months_view\n",
      "bigquery.mapped.age_months_view2\n",
      "bigquery.mapped.age_view\n",
      "bigquery.mapped.age_years\n",
      "bigquery.mapped.subject_disease\n",
      "bigquery.mesa.airnrexammain_hmb\n",
      "bigquery.mesa.airnrspirometry_hmb\n",
      "bigquery.mesa.ancilmesaneighborfoodexam1_hmb\n",
      "bigquery.mesa.ancilmesaneighborfoodexam2_hmb\n",
      "bigquery.mesa.ancilmesaneighborfoodexam3_hmb\n",
      "bigquery.mesa.ancilmesaneighborfoodexam4_hmb\n",
      "bigquery.mesa.parent_mesa_hmb_npu\n",
      "bigquery.mesa.sample_multi\n",
      "bigquery.mesa.subject_multi\n",
      "bigquery.mesa_esp_heartgo.parent_mesa\n",
      "bigquery.mesa_esp_heartgo.sample_attributes_hmb\n",
      "bigquery.mesa_esp_heartgo.sample_attributes_hmb_npu\n",
      "bigquery.mesa_esp_heartgo.subject_phenotypes_hmb\n",
      "bigquery.mesa_esp_heartgo.subject_phenotypes_hmb_npu\n",
      "bigquery.metadata.dbgap_table\n",
      "bigquery.metadata.loinc2hpoannotation\n",
      "bigquery.metadata.mapping\n",
      "bigquery.metadata.mapping2\n",
      "bigquery.metadata.mapping2_term\n",
      "bigquery.metadata.merged\n",
      "bigquery.metadata.s_study\n",
      "bigquery.metadata.s_table\n",
      "bigquery.metadata.s_variable\n",
      "bigquery.metadata.study\n",
      "bigquery.metadata.value_map\n",
      "bigquery.metadata.vocab\n",
      "bigquery.metadata.vocab_term\n",
      "bigquery.onek_genomes.bdc_1000genomes\n",
      "bigquery.onek_genomes.onek_drs\n",
      "bigquery.onek_genomes.onek_recal_variants_drs\n",
      "bigquery.onek_genomes.sra_drs_files\n",
      "bigquery.onek_genomes.sra_onek_drs\n",
      "bigquery.onek_genomes.ssd_drs\n",
      "bigquery.onek_genomes.thousand_genomes_meta\n",
      "bigquery.organoid_profiling_pc.sample_attributes_gru\n",
      "bigquery.organoid_profiling_pc.sample_multi\n",
      "bigquery.organoid_profiling_pc.subject_multi\n",
      "bigquery.organoid_profiling_pc.subject_phenotypes_gru\n",
      "bigquery.pc_rare_variation.sample\n",
      "bigquery.pc_rare_variation.sample_attributes\n",
      "bigquery.pc_rare_variation.subject\n",
      "bigquery.pc_rare_variation.subject_phenotypes\n",
      "bigquery.pptc.manifest_small\n",
      "bigquery.pptc.sample\n",
      "bigquery.pptc.sample_attributes\n",
      "bigquery.pptc.subject\n",
      "bigquery.pptc.subject_phenotypes\n",
      "bigquery.rand_gmkf_cdh.phenotype\n",
      "bigquery.scr_ega.scr_egapancreatic_files\n",
      "bigquery.scr_ega.scr_egapancreatic_sample_multi\n",
      "bigquery.scr_egapancreatic.sample\n",
      "bigquery.scr_gecco_susceptibility.gecco_sample_acc\n",
      "bigquery.scr_gecco_susceptibility.phs001554_gecco_sddp_manifest\n",
      "bigquery.scr_gecco_susceptibility.sample_attributes_multi\n",
      "bigquery.scr_gecco_susceptibility.sample_multi\n",
      "bigquery.scr_gecco_susceptibility.sb_drs_index\n",
      "bigquery.scr_gecco_susceptibility.subject_multi\n",
      "bigquery.scr_gecco_susceptibility.subject_phenotypes_multi\n",
      "bigquery.scr_organoid.subject_phenotypes_gru\n",
      "bigquery.scripts.pipelinelog\n",
      "bigquery.scripts.scriptlist\n",
      "bigquery.sra.accession_summary\n",
      "bigquery.sra.gecco_join\n",
      "bigquery.sra.run\n",
      "bigquery.sra.run_file\n",
      "bigquery.tgtest.program\n",
      "bigquery.tgtest.project\n",
      "bigquery.tgtest.sample\n",
      "bigquery.tgtest.sequencing\n",
      "bigquery.tgtest.subject\n",
      "bigquery.topmed_wgs_mayo_vte.sample_attributes_gru\n",
      "bigquery.topmed_wgs_mayo_vte.subject_multi\n",
      "bigquery.topmed_wgs_mayo_vte.subject_phenotypes_gru\n",
      "bigquery.topmed_wgs_partners_afgen.sample_attributes_hmb\n",
      "bigquery.topmed_wgs_partners_afgen.sample_multi\n",
      "bigquery.topmed_wgs_partners_afgen.subject_multi\n",
      "bigquery.topmed_wgs_partners_afgen.subject_phenotypes_hmb\n",
      "bigquery.transfer.gdc_case_file_metadata_gdcfileid_to_gcsurl_current\n",
      "bigquery.transfer.tcga_clinical_gdc_current\n",
      "bigquery.transfer.tcga_somatic_mutation_hg38_gdc_current\n",
      "____Page2_______________\n",
      "jmx.current.com.sun.management:type=diagnosticcommand\n",
      "jmx.current.com.sun.management:type=hotspotdiagnostic\n",
      "jmx.current.io.airlift.discovery.client:name=announcer\n",
      "jmx.current.io.airlift.discovery.client:name=serviceinventory\n",
      "jmx.current.io.airlift.discovery.store:name=dynamic,type=distributedstore\n",
      "jmx.current.io.airlift.discovery.store:name=dynamic,type=httpremotestore\n",
      "jmx.current.io.airlift.discovery.store:name=dynamic,type=replicator\n",
      "jmx.current.io.airlift.http.client:name=fordiscoveryclient,type=httpclient\n",
      "jmx.current.io.airlift.http.client:name=fordynamicstore,type=httpclient\n",
      "jmx.current.io.airlift.http.client:name=forexchange,type=httpclient\n",
      "jmx.current.io.airlift.http.client:name=forfailuredetector,type=httpclient\n",
      "jmx.current.io.airlift.http.client:name=formemorymanager,type=httpclient\n",
      "jmx.current.io.airlift.http.client:name=fornodemanager,type=httpclient\n",
      "jmx.current.io.airlift.http.client:name=forscheduler,type=httpclient\n",
      "jmx.current.io.airlift.http.client:name=forworkerinfo,type=httpclient\n",
      "jmx.current.io.airlift.http.server:context=http_1_1_h2c@5543d800,id=0,type=httpserverchannellistener\n",
      "jmx.current.io.airlift.http.server:id=0,type=statsrecordinghandler\n",
      "jmx.current.io.airlift.http.server:name=httpserver\n",
      "jmx.current.io.airlift.http.server:name=requeststats\n",
      "jmx.current.io.airlift.jmx:name=stacktracembean\n",
      "jmx.current.io.airlift.log:name=logging\n",
      "jmx.current.io.airlift.node:name=nodeinfo\n",
      "jmx.current.io.airlift.stats:name=gcmonitor\n",
      "jmx.current.io.airlift.stats:name=pausemeter\n",
      "jmx.current.io.trino.plugin.jdbc:name=mysql,type=cachingjdbcclient\n",
      "jmx.current.io.trino.plugin.jdbc:name=mysql,type=connectionfactory\n",
      "jmx.current.io.trino.plugin.jdbc:name=mysql,type=jdbcclient\n",
      "jmx.current.java.lang:name=codecachemanager,type=memorymanager\n",
      "jmx.current.java.lang:name=codeheap 'non-nmethods',type=memorypool\n",
      "jmx.current.java.lang:name=codeheap 'non-profiled nmethods',type=memorypool\n",
      "jmx.current.java.lang:name=codeheap 'profiled nmethods',type=memorypool\n",
      "jmx.current.java.lang:name=compressed class space,type=memorypool\n",
      "jmx.current.java.lang:name=g1 eden space,type=memorypool\n",
      "jmx.current.java.lang:name=g1 old gen,type=memorypool\n",
      "jmx.current.java.lang:name=g1 old generation,type=garbagecollector\n",
      "jmx.current.java.lang:name=g1 survivor space,type=memorypool\n",
      "jmx.current.java.lang:name=g1 young generation,type=garbagecollector\n",
      "jmx.current.java.lang:name=metaspace manager,type=memorymanager\n",
      "jmx.current.java.lang:name=metaspace,type=memorypool\n",
      "jmx.current.java.lang:type=classloading\n",
      "jmx.current.java.lang:type=compilation\n",
      "jmx.current.java.lang:type=memory\n",
      "jmx.current.java.lang:type=operatingsystem\n",
      "jmx.current.java.lang:type=runtime\n",
      "jmx.current.java.lang:type=threading\n",
      "jmx.current.java.nio:name=direct,type=bufferpool\n",
      "jmx.current.java.nio:name=mapped,type=bufferpool\n",
      "jmx.current.java.util.logging:type=logging\n",
      "jmx.current.jdk.management.jfr:type=flightrecorder\n",
      "jmx.current.jmimplementation:type=mbeanserverdelegate\n",
      "jmx.current.org.eclipse.jetty.http2.server:context=http_1_1_h2c@5543d800,id=0,type=abstracthttp2serverconnectionfactory$http2sessioncontainer\n",
      "jmx.current.org.eclipse.jetty.http2.server:context=http_1_1_h2c@5543d800,id=0,type=http2cserverconnectionfactory\n",
      "jmx.current.org.eclipse.jetty.io:context=http_1_1_h2c@5543d800,id=0,type=arraybytebufferpool\n",
      "jmx.current.org.eclipse.jetty.io:context=http_1_1_h2c@5543d800,id=0,type=connectionstatistics\n",
      "jmx.current.org.eclipse.jetty.io:context=http_1_1_h2c@5543d800,id=0,type=managedselector\n",
      "jmx.current.org.eclipse.jetty.io:context=http_1_1_h2c@5543d800,id=0,type=managedselector$selectorproducer\n",
      "jmx.current.org.eclipse.jetty.io:context=http_1_1_h2c@5543d800,id=1,type=managedselector\n",
      "jmx.current.org.eclipse.jetty.io:context=http_1_1_h2c@5543d800,id=1,type=managedselector$selectorproducer\n",
      "jmx.current.org.eclipse.jetty.io:context=http_1_1_h2c@5543d800,id=2,type=managedselector\n",
      "jmx.current.org.eclipse.jetty.io:context=http_1_1_h2c@5543d800,id=2,type=managedselector$selectorproducer\n",
      "jmx.current.org.eclipse.jetty.io:context=http_1_1_h2c@5543d800,id=3,type=managedselector\n",
      "jmx.current.org.eclipse.jetty.io:context=http_1_1_h2c@5543d800,id=3,type=managedselector$selectorproducer\n",
      "jmx.current.org.eclipse.jetty.jmx:id=0,type=mbeancontainer\n",
      "jmx.current.org.eclipse.jetty.server.handler.gzip:context=_root@@http_,id=0,type=gziphandler\n",
      "jmx.current.org.eclipse.jetty.server.handler:id=0,type=errorhandler\n",
      "jmx.current.org.eclipse.jetty.server.handler:id=0,type=handlercollection\n",
      "jmx.current.org.eclipse.jetty.server.handler:id=0,type=handlerlist\n",
      "jmx.current.org.eclipse.jetty.server.handler:id=0,type=requestloghandler\n",
      "jmx.current.org.eclipse.jetty.server.handler:id=0,type=statisticshandler\n",
      "jmx.current.org.eclipse.jetty.server:context=http_1_1_h2c@5543d800,id=0,type=abstractconnector$acceptor\n",
      "jmx.current.org.eclipse.jetty.server:context=http_1_1_h2c@5543d800,id=0,type=httpconfiguration\n",
      "jmx.current.org.eclipse.jetty.server:context=http_1_1_h2c@5543d800,id=0,type=httpconnectionfactory\n",
      "jmx.current.org.eclipse.jetty.server:context=http_1_1_h2c@5543d800,id=0,type=serverconnector\n",
      "jmx.current.org.eclipse.jetty.server:context=http_1_1_h2c@5543d800,id=0,type=serverconnector$serverconnectormanager\n",
      "jmx.current.org.eclipse.jetty.server:id=0,type=server\n",
      "jmx.current.org.eclipse.jetty.servlet:context=_root@@http_,id=0,name=io_airlift_http_server_timingfilter-32f45e15,type=filterholder\n",
      "jmx.current.org.eclipse.jetty.servlet:context=_root@@http_,id=0,name=io_airlift_http_server_timingfilter-32f45e15,type=filtermapping\n",
      "jmx.current.org.eclipse.jetty.servlet:context=_root@@http_,id=0,name=io_airlift_http_server_tracetokenfilter-524dd373,type=filterholder\n",
      "jmx.current.org.eclipse.jetty.servlet:context=_root@@http_,id=0,name=io_airlift_http_server_tracetokenfilter-524dd373,type=filtermapping\n",
      "jmx.current.org.eclipse.jetty.servlet:context=_root@@http_,id=0,name=org_eclipse_jetty_servlet_servlethandler$default404servlet-78c23ade,type=servletholder\n",
      "jmx.current.org.eclipse.jetty.servlet:context=_root@@http_,id=0,name=org_eclipse_jetty_servlet_servlethandler$default404servlet-78c23ade,type=servletmapping\n",
      "jmx.current.org.eclipse.jetty.servlet:context=_root@@http_,id=0,name=org_glassfish_jersey_servlet_servletcontainer-ad585fb,type=servletholder\n",
      "jmx.current.org.eclipse.jetty.servlet:context=_root@@http_,id=0,name=org_glassfish_jersey_servlet_servletcontainer-ad585fb,type=servletmapping\n",
      "jmx.current.org.eclipse.jetty.servlet:context=_root@@http_,id=0,type=servletcontexthandler\n",
      "jmx.current.org.eclipse.jetty.servlet:context=_root@@http_,id=0,type=servlethandler\n",
      "jmx.current.org.eclipse.jetty.util.compression:context=_root@@http_,id=0,type=deflaterpool\n",
      "jmx.current.org.eclipse.jetty.util.thread.strategy:context=http_1_1_h2c@5543d800,id=0,type=eatwhatyoukill\n",
      "jmx.current.org.eclipse.jetty.util.thread.strategy:context=http_1_1_h2c@5543d800,id=1,type=eatwhatyoukill\n",
      "jmx.current.org.eclipse.jetty.util.thread.strategy:context=http_1_1_h2c@5543d800,id=2,type=eatwhatyoukill\n",
      "jmx.current.org.eclipse.jetty.util.thread.strategy:context=http_1_1_h2c@5543d800,id=3,type=eatwhatyoukill\n",
      "jmx.current.org.eclipse.jetty.util.thread:context=http_1_1_h2c@5543d800,id=0,type=scheduledexecutorscheduler\n",
      "jmx.current.org.eclipse.jetty.util.thread:id=0,type=queuedthreadpool\n",
      "jmx.current.org.eclipse.jetty.util.thread:id=0,type=reservedthreadexecutor\n",
      "jmx.current.org.eclipse.jetty.util.thread:id=0,type=threadpoolbudget\n",
      "jmx.current.sun.nio.ch:context=http_1_1_h2c@5543d800,id=0,type=serversocketchannelimpl\n",
      "jmx.current.trino.execution.executor:name=multilevelsplitqueue\n",
      "jmx.current.trino.execution.executor:name=taskexecutor\n",
      "jmx.current.trino.execution.resourcegroups:name=internalresourcegroupmanager\n",
      "jmx.current.trino.execution.scheduler:name=nodescheduler\n",
      "jmx.current.trino.execution.scheduler:name=splitschedulerstats\n",
      "jmx.current.trino.execution:name=clustersizemonitor\n",
      "jmx.current.trino.execution:name=queryexecution\n",
      "jmx.current.trino.execution:name=querymanager\n",
      "jmx.current.trino.execution:name=remotetaskfactory\n",
      "jmx.current.trino.execution:name=taskmanager\n",
      "jmx.current.trino.failuredetector:name=heartbeatfailuredetector\n",
      "jmx.current.trino.memory:name=clustermemorymanager\n",
      "jmx.current.trino.memory:name=general,type=clustermemorypool\n",
      "jmx.current.trino.memory:name=general,type=memorypool\n",
      "jmx.current.trino.metadata:name=discoverynodemanager\n",
      "jmx.current.trino.operator.index:name=indexjoinlookupstats\n",
      "jmx.current.trino.security:name=accesscontrolmanager\n",
      "jmx.current.trino.server.remotetask:name=remotetaskstats\n",
      "jmx.current.trino.server:name=asynchttpexecutionmbean\n",
      "jmx.current.trino.server:name=exchangeexecutionmbean\n",
      "jmx.current.trino.server:name=statementhttpexecutionmbean\n",
      "jmx.current.trino.server:name=taskexecutorresource\n",
      "jmx.current.trino.server:name=taskresource\n",
      "jmx.current.trino.spi.type:name=typeoperators\n",
      "jmx.current.trino.spiller:name=spillerfactory\n",
      "jmx.current.trino.sql.gen:name=expressioncompiler\n",
      "jmx.current.trino.sql.gen:name=joincompiler\n",
      "jmx.current.trino.sql.gen:name=joinfilterfunctioncompiler\n",
      "jmx.current.trino.sql.gen:name=orderingcompiler\n",
      "jmx.current.trino.sql.gen:name=pagefunctioncompiler\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=addexchangesbelowexchangepartialaggregationgroupid\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=addexchangesbelowprojectionpartialaggregationgroupid\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=addintermediateaggregations\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=aggregationexpressionrewrite\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=applypreferredtableexecutepartitioning\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=applypreferredtablewriterpartitioning\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=applytablescanredirection\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=createpartialtopn\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=decorrelateinnerunnestwithglobalaggregation\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=decorrelateleftunnestwithglobalaggregation\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=decorrelateunnest\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=determinejoindistributiontype\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=determinesemijoindistributiontype\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=determinetablescannodepartitioning\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=eliminatecrossjoins\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=evaluateemptyintersect\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=evaluatezerosample\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=extractdereferencesfromfilterabovescan\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=extractspatialinnerjoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=extractspatialleftjoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=filterexpressionrewrite\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=implementbernoullisampleasfilter\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=implementexceptall\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=implementexceptdistinctasunion\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=implementfilteredaggregations\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=implementintersectall\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=implementintersectdistinctasunion\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=implementlimitwithties\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=implementoffset\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=inlineprojectintofilter\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=inlineprojections\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=joinexpressionrewrite\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergeadjacentwindowsoverprojects\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergeexcept\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergefilters\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergeintersect\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergelimitoverprojectwithsort\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergelimits\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergelimitwithdistinct\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergelimitwithsort\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergelimitwithtopn\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergepatternrecognitionnodeswithoutproject\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergepatternrecognitionnodeswithproject\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergeprojectwithvalues\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=mergeunion\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=multipledistinctaggregationtomarkdistinct\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=optimizeduplicateinsensitivejoins\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=optimizerowpattern\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=patternrecognitionexpressionrewrite\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=projectexpressionrewrite\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneaggregationcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneaggregationsourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneapplycolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneapplycorrelation\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneapplysourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneassignuniqueidcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunecorrelatedjoincolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunecorrelatedjoincorrelation\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunecountaggregationoverscalar\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunedeletesourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunedistinctaggregation\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunedistinctlimitsourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneenforcesinglerowcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneexceptsourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneexchangecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneexchangesourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneexplainanalyzesourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunefiltercolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunegroupidcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunegroupidsourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneindexjoincolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneindexsourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneintersectsourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunejoinchildrencolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunejoincolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunelimitcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunemarkdistinctcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneoffsetcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneorderbyinaggregation\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneoutputsourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunepattenrecognitioncolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunepatternrecognitionsourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneprojectcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunerownumbercolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunesamplecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunesemijoincolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunesemijoinfilteringsourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunesortcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunespatialjoinchildrencolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunespatialjoincolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunetableexecutesourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunetablescancolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunetablewritersourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunetopncolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunetopnrankingcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneunioncolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneunionsourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneunnestcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneunnestsourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pruneupdatesourcecolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunevaluescolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=prunewindowcolumns\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushaggregationintotablescan\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushaggregationthroughouterjoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdeleteintoconnector\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdistinctlimitintotablescan\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencesthroughassignuniqueid\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencesthroughlimit\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencesthroughrownumber\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencesthroughsort\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencesthroughtopn\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencesthroughtopnranking\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencesthroughwindow\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencethroughfilter\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencethroughjoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencethroughproject\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencethroughsemijoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdowndereferencethroughunnest\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdownfilterintorownumber\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdownfilterintowindow\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdownlimitintorownumber\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdownlimitintowindow\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushdownprojectionsfrompatternrecognition\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushjoinintotablescan\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushlimitintotablescan\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushlimitthroughmarkdistinct\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushlimitthroughoffset\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushlimitthroughouterjoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushlimitthroughproject\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushlimitthroughsemijoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushlimitthroughunion\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushoffsetthroughproject\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushpartialaggregationthroughexchange\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushpartialaggregationthroughjoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushpredicateintotablescan\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushpredicatethroughprojectintorownumber\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushpredicatethroughprojectintowindow\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushprojectionintotablescan\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushprojectionthroughexchange\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushprojectionthroughunion\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushremoteexchangethroughassignuniqueid\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushsampleintotablescan\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushtablewritethroughunion\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushtopnintotablescan\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushtopnthroughouterjoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushtopnthroughproject\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=pushtopnthroughunion\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeaggregationinsemijoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeemptydelete\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeemptyexceptbranches\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeemptytableexecute\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeemptyunionbranches\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removefullsample\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeredundantdistinctlimit\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeredundantenforcesinglerownode\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeredundantexists\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeredundantidentityprojections\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeredundantjoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeredundantlimit\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeredundantoffset\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeredundantsort\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeredundantsortbelowlimitwithties\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeredundanttablescanpredicate\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeredundanttopn\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removetrivialfilters\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeunreferencedscalarapplynodes\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=removeunreferencedscalarsubqueries\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=reorderjoins\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=replacejoinoverconstantwithproject\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=replaceredundantjoinwithproject\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=replaceredundantjoinwithsource\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=replacewindowwithrownumber\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=rewritespatialpartitioningaggregation\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=simplifycountoverconstant\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=simplifyfilterpredicate\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=singledistinctaggregationtogroupby\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=swapadjacentwindowsbyspecifications\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformcorrelateddistinctaggregationwithoutprojection\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformcorrelateddistinctaggregationwithprojection\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformcorrelatedglobalaggregationwithoutprojection\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformcorrelatedglobalaggregationwithprojection\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformcorrelatedgroupedaggregationwithoutprojection\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformcorrelatedgroupedaggregationwithprojection\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformcorrelatedinpredicatetojoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformcorrelatedjointojoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformcorrelatedscalarsubquery\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformcorrelatedsinglerowsubquerytoproject\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformexistsapplytocorrelatedjoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformfilteringsemijointoinnerjoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformuncorrelatedinpredicatesubquerytosemijoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=transformuncorrelatedsubquerytojoin\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=unwrapsinglecolumnrowinapply\n",
      "jmx.current.trino.sql.planner.iterative:name=iterativeoptimizer,rule=valuesexpressionrewrite\n",
      "jmx.current.trino.sql.planner.optimizations:name=planoptimizer,optimizer=addexchanges\n",
      "jmx.current.trino.sql.planner.optimizations:name=planoptimizer,optimizer=predicatepushdown\n",
      "jmx.current.trino.type:name=typeoperatorscache\n",
      "____Page3_______________\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'tables'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/wz/jjbxsnr13v7dkw2jkbpmqd_dly65gq/T/ipykernel_14947/2318540970.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mfasp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msearch\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mDataConnectClient\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mcl\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDataConnectClient\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'http://localhost:8089/'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mcl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlistTables\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/imfenvs/fasp/lib/python3.9/site-packages/fasp/search/data_connect_client.py\u001b[0m in \u001b[0;36mlistTables\u001b[0;34m(self, requestedCatalog, verbose)\u001b[0m\n\u001b[1;32m     58\u001b[0m                         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m                                 \u001b[0mnext_url\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 60\u001b[0;31m                         \u001b[0;32mfor\u001b[0m \u001b[0mt\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'tables'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     61\u001b[0m                                 \u001b[0;32mif\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m                                         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'tables'"
     ]
    }
   ],
   "source": [
    "from fasp.search import DataConnectClient\n",
    "cl = DataConnectClient('http://localhost:8089/')\n",
    "cl.listTables(verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The listTableColumns function prints a more concise listing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# init search client\n",
    "from search_python_client.search import DrsClient, SearchClient\n",
    "base_url = 'http://localhost:8089/'\n",
    "search_client = SearchClient(base_url=base_url)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'name': 'bigquery.blog_unnest.firebase_raw', 'data_model': {'$ref': 'http://localhost:8089/table/bigquery.blog_unnest.firebase_raw/info'}}, {'name': 'bigquery.cda.cdav3', 'data_model': {'$ref': 'http://localhost:8089/table/bigquery.cda.cdav3/info'}}, {'name': 'bigquery.fasp_scrambled.scr_egapancreatic_sample', 'data_model': {'$ref': 'http://localhost:8089/table/bigquery.fasp_scrambled.scr_egapancreatic_sample/info'}}, {'name': 'bigquery.gtex8.program', 'data_model': {'$ref': 'http://localhost:8089/table/bigquery.gtex8.program/info'}}, {'name': 'bigquery.gtex8.project', 'data_model': {'$ref': 'http://localhost:8089/table/bigquery.gtex8.project/info'}}, {'name': 'bigquery.gtex8.sample', 'data_model': {'$ref': 'http://localhost:8089/table/bigquery.gtex8.sample/info'}}, {'name': 'bigquery.gtex8.sequencing', 'data_model': {'$ref': 'http://localhost:8089/table/bigquery.gtex8.sequencing/info'}}, {'name': 'bigquery.gtex8.sraruntable_breast_cram', 'data_model': {'$ref': 'http://localhost:8089/table/bigquery.gtex8.sraruntable_breast_cram/info'}}, {'name': 'bigquery.gtex8.sraruntable_nocram', 'data_model': {'$ref': 'http://localhost:8089/table/bigquery.gtex8.sraruntable_nocram/info'}}, {'name': 'bigquery.gtex8.sraruntable_nocram_nobam', 'data_model': {'$ref': 'http://localhost:8089/table/bigquery.gtex8.sraruntable_nocram_nobam/info'}}]\n"
     ]
    }
   ],
   "source": [
    "# get tables\n",
    "tables_iterator = search_client.get_table_list()\n",
    "tables = [next(tables_iterator, None) for i in range(10)]\n",
    "tables = list(filter(None, tables))\n",
    "print(tables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Retrieving the query_\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n",
      "____Page8_______________\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[['Female', 46, 'NHS'],\n",
       " ['Female', 46, 'NHS'],\n",
       " ['Female', 47, 'NHS'],\n",
       " ['Female', 47, 'NHS'],\n",
       " ['Female', 48, 'NHS'],\n",
       " ['Female', 49, 'NHS'],\n",
       " ['Male', 51, 'NHS'],\n",
       " ['Male', 51, 'NHS'],\n",
       " ['Female', 51, 'NHS'],\n",
       " ['Female', 51, 'NHS'],\n",
       " ['Male', 52, 'NHS'],\n",
       " ['Female', 52, 'NHS'],\n",
       " ['Female', 52, 'NHS'],\n",
       " ['Female', 53, 'NHS'],\n",
       " ['Female', 53, 'NHS'],\n",
       " ['Female', 54, 'NHS'],\n",
       " ['Male', 54, 'NHS'],\n",
       " ['Male', 54, 'NHS'],\n",
       " ['Female', 54, 'NHS'],\n",
       " ['Male', 55, 'NHS'],\n",
       " ['Male', 55, 'NHS'],\n",
       " ['Female', 55, 'NHS'],\n",
       " ['Female', 55, 'NHS'],\n",
       " ['Female', 55, 'NHS'],\n",
       " ['Female', 55, 'NHS'],\n",
       " ['Female', 55, 'NHS'],\n",
       " ['Female', 55, 'NHS'],\n",
       " ['Male', 55, 'NHS'],\n",
       " ['Female', 56, 'NHS'],\n",
       " ['Female', 56, 'NHS'],\n",
       " ['Female', 56, 'NHS'],\n",
       " ['Male', 56, 'NHS'],\n",
       " ['Female', 56, 'NHS'],\n",
       " ['Female', 56, 'NHS'],\n",
       " ['Male', 56, 'NHS'],\n",
       " ['Female', 56, 'NHS'],\n",
       " ['Female', 57, 'NHS'],\n",
       " ['Female', 57, 'NHS'],\n",
       " ['Female', 57, 'NHS'],\n",
       " ['Male', 58, 'NHS'],\n",
       " ['Female', 58, 'NHS'],\n",
       " ['Female', 58, 'NHS'],\n",
       " ['Male', 58, 'NHS'],\n",
       " ['Female', 59, 'NHS'],\n",
       " ['Female', 59, 'NHS'],\n",
       " ['Male', 59, 'NHS'],\n",
       " ['Male', 59, 'NHS'],\n",
       " ['Male', 59, 'NHS'],\n",
       " ['Male', 59, 'NHS'],\n",
       " ['Female', 60, 'NHS'],\n",
       " ['Female', 60, 'NHS'],\n",
       " ['Female', 60, 'NHS'],\n",
       " ['Female', 60, 'NHS'],\n",
       " ['Female', 60, 'NHS'],\n",
       " ['Male', 60, 'NHS'],\n",
       " ['Female', 61, 'NHS'],\n",
       " ['Male', 61, 'NHS'],\n",
       " ['Male', 62, 'NHS'],\n",
       " ['Female', 62, 'NHS'],\n",
       " ['Female', 62, 'NHS'],\n",
       " ['Male', 62, 'NHS'],\n",
       " ['Female', 62, 'NHS'],\n",
       " ['Female', 62, 'NHS'],\n",
       " ['Female', 62, 'NHS'],\n",
       " ['Female', 62, 'NHS'],\n",
       " ['Male', 63, 'NHS'],\n",
       " ['Female', 63, 'NHS'],\n",
       " ['Female', 63, 'NHS'],\n",
       " ['Male', 63, 'NHS'],\n",
       " ['Female', 63, 'NHS'],\n",
       " ['Female', 63, 'NHS'],\n",
       " ['Female', 63, 'NHS'],\n",
       " ['Male', 63, 'NHS'],\n",
       " ['Male', 63, 'NHS'],\n",
       " ['Male', 63, 'NHS'],\n",
       " ['Male', 63, 'NHS'],\n",
       " ['Female', 63, 'NHS'],\n",
       " ['Male', 63, 'NHS'],\n",
       " ['Female', 63, 'NHS'],\n",
       " ['Female', 63, 'NHS'],\n",
       " ['Female', 63, 'NHS'],\n",
       " ['Female', 64, 'NHS'],\n",
       " ['Male', 64, 'NHS'],\n",
       " ['Female', 64, 'NHS'],\n",
       " ['Female', 64, 'NHS'],\n",
       " ['Male', 64, 'NHS'],\n",
       " ['Female', 64, 'NHS'],\n",
       " ['Female', 64, 'NHS'],\n",
       " ['Male', 64, 'NHS'],\n",
       " ['Female', 64, 'NHS'],\n",
       " ['Male', 64, 'NHS'],\n",
       " ['Female', 64, 'NHS'],\n",
       " ['Female', 64, 'NHS'],\n",
       " ['Male', 64, 'NHS'],\n",
       " ['Female', 64, 'NHS'],\n",
       " ['Male', 65, 'NHS'],\n",
       " ['Male', 65, 'NHS'],\n",
       " ['Female', 65, 'NHS'],\n",
       " ['Female', 65, 'NHS'],\n",
       " ['Female', 65, 'NHS'],\n",
       " ['Male', 65, 'NHS'],\n",
       " ['Male', 65, 'NHS'],\n",
       " ['Female', 65, 'NHS'],\n",
       " ['Female', 65, 'NHS'],\n",
       " ['Female', 65, 'NHS'],\n",
       " ['Female', 65, 'NHS'],\n",
       " ['Female', 66, 'NHS'],\n",
       " ['Female', 66, 'NHS'],\n",
       " ['Female', 66, 'NHS'],\n",
       " ['Male', 66, 'NHS'],\n",
       " ['Male', 66, 'NHS'],\n",
       " ['Female', 66, 'NHS'],\n",
       " ['Male', 66, 'NHS'],\n",
       " ['Female', 66, 'NHS'],\n",
       " ['Male', 66, 'NHS'],\n",
       " ['Female', 66, 'NHS'],\n",
       " ['Female', 66, 'NHS'],\n",
       " ['Female', 67, 'NHS'],\n",
       " ['Female', 67, 'NHS'],\n",
       " ['Female', 67, 'NHS'],\n",
       " ['Female', 67, 'NHS'],\n",
       " ['Female', 67, 'NHS'],\n",
       " ['Female', 67, 'NHS'],\n",
       " ['Male', 67, 'NHS'],\n",
       " ['Male', 67, 'NHS'],\n",
       " ['Female', 68, 'NHS'],\n",
       " ['Female', 68, 'NHS'],\n",
       " ['Female', 68, 'NHS'],\n",
       " ['Female', 68, 'NHS'],\n",
       " ['Male', 68, 'NHS'],\n",
       " ['Female', 68, 'NHS'],\n",
       " ['Female', 68, 'NHS'],\n",
       " ['Female', 68, 'NHS'],\n",
       " ['Female', 68, 'NHS'],\n",
       " ['Female', 68, 'NHS'],\n",
       " ['Male', 68, 'NHS'],\n",
       " ['Male', 68, 'NHS'],\n",
       " ['Female', 68, 'NHS'],\n",
       " ['Male', 69, 'NHS'],\n",
       " ['Female', 69, 'NHS'],\n",
       " ['Female', 69, 'NHS'],\n",
       " ['Female', 69, 'NHS'],\n",
       " ['Female', 69, 'NHS'],\n",
       " ['Female', 69, 'NHS'],\n",
       " ['Female', 69, 'NHS'],\n",
       " ['Female', 69, 'NHS'],\n",
       " ['Male', 70, 'NHS'],\n",
       " ['Female', 70, 'NHS'],\n",
       " ['Female', 70, 'NHS'],\n",
       " ['Male', 70, 'NHS'],\n",
       " ['Male', 70, 'NHS'],\n",
       " ['Female', 70, 'NHS'],\n",
       " ['Male', 70, 'NHS'],\n",
       " ['Female', 70, 'NHS'],\n",
       " ['Female', 71, 'NHS'],\n",
       " ['Female', 71, 'NHS'],\n",
       " ['Female', 71, 'NHS'],\n",
       " ['Female', 71, 'NHS'],\n",
       " ['Male', 71, 'NHS'],\n",
       " ['Male', 71, 'NHS'],\n",
       " ['Female', 71, 'NHS'],\n",
       " ['Female', 71, 'NHS'],\n",
       " ['Male', 72, 'NHS'],\n",
       " ['Female', 72, 'NHS'],\n",
       " ['Male', 72, 'NHS'],\n",
       " ['Male', 72, 'NHS'],\n",
       " ['Male', 72, 'NHS'],\n",
       " ['Male', 72, 'NHS'],\n",
       " ['Female', 72, 'NHS'],\n",
       " ['Male', 72, 'NHS'],\n",
       " ['Female', 72, 'NHS'],\n",
       " ['Male', 72, 'NHS'],\n",
       " ['Female', 72, 'NHS'],\n",
       " ['Female', 72, 'NHS'],\n",
       " ['Female', 73, 'NHS'],\n",
       " ['Female', 73, 'NHS'],\n",
       " ['Male', 73, 'NHS'],\n",
       " ['Male', 74, 'NHS'],\n",
       " ['Female', 74, 'NHS'],\n",
       " ['Male', 74, 'NHS'],\n",
       " ['Female', 74, 'NHS'],\n",
       " ['Male', 74, 'NHS'],\n",
       " ['Male', 74, 'NHS'],\n",
       " ['Male', 74, 'NHS'],\n",
       " ['Female', 74, 'NHS'],\n",
       " ['Female', 74, 'NHS'],\n",
       " ['Male', 74, 'NHS'],\n",
       " ['Male', 74, 'NHS'],\n",
       " ['Female', 74, 'NHS'],\n",
       " ['Male', 74, 'NHS'],\n",
       " ['Female', 75, 'NHS'],\n",
       " ['Male', 75, 'NHS'],\n",
       " ['Female', 75, 'NHS'],\n",
       " ['Male', 75, 'NHS'],\n",
       " ['Female', 75, 'NHS'],\n",
       " ['Male', 76, 'NHS'],\n",
       " ['Male', 76, 'NHS'],\n",
       " ['Female', 76, 'NHS'],\n",
       " ['Female', 76, 'NHS'],\n",
       " ['Female', 76, 'NHS'],\n",
       " ['Female', 77, 'NHS'],\n",
       " ['Male', 77, 'NHS'],\n",
       " ['Male', 77, 'NHS'],\n",
       " ['Female', 77, 'NHS'],\n",
       " ['Female', 77, 'NHS'],\n",
       " ['Female', 78, 'NHS'],\n",
       " ['Male', 78, 'NHS'],\n",
       " ['Male', 78, 'NHS'],\n",
       " ['Male', 79, 'NHS'],\n",
       " ['Female', 79, 'NHS'],\n",
       " ['Male', 80, 'NHS'],\n",
       " ['Female', 80, 'NHS'],\n",
       " ['Male', 81, 'NHS'],\n",
       " ['Female', 81, 'NHS'],\n",
       " ['Male', 82, 'NHS'],\n",
       " ['Male', 84, 'NHS'],\n",
       " ['Male', 85, 'NHS'],\n",
       " ['Female', 87, 'NHS'],\n",
       " ['Female', 87, 'NHS'],\n",
       " ['Female', 88, 'NHS'],\n",
       " ['Male', 37, 'WHI'],\n",
       " ['Male', 37, 'WHI'],\n",
       " ['Female', 38, 'WHI'],\n",
       " ['Male', 44, 'WHI'],\n",
       " ['Female', 44, 'WHI'],\n",
       " ['Female', 45, 'WHI'],\n",
       " ['Male', 45, 'WHI'],\n",
       " ['Female', 46, 'WHI'],\n",
       " ['Female', 47, 'WHI'],\n",
       " ['Female', 47, 'WHI'],\n",
       " ['Female', 47, 'WHI'],\n",
       " ['Female', 48, 'WHI'],\n",
       " ['Female', 48, 'WHI'],\n",
       " ['Female', 48, 'WHI'],\n",
       " ['Male', 48, 'WHI'],\n",
       " ['Male', 48, 'WHI'],\n",
       " ['Male', 49, 'WHI'],\n",
       " ['Female', 49, 'WHI'],\n",
       " ['Male', 49, 'WHI'],\n",
       " ['Male', 49, 'WHI'],\n",
       " ['Female', 49, 'WHI'],\n",
       " ['Male', 50, 'WHI'],\n",
       " ['Female', 50, 'WHI'],\n",
       " ['Female', 50, 'WHI'],\n",
       " ['Male', 50, 'WHI'],\n",
       " ['Female', 50, 'WHI'],\n",
       " ['Male', 50, 'WHI'],\n",
       " ['Male', 51, 'WHI'],\n",
       " ['Female', 51, 'WHI'],\n",
       " ['Female', 51, 'WHI'],\n",
       " ['Female', 51, 'WHI'],\n",
       " ['Female', 51, 'WHI'],\n",
       " ['Female', 51, 'WHI'],\n",
       " ['Male', 51, 'WHI'],\n",
       " ['Female', 51, 'WHI'],\n",
       " ['Female', 52, 'WHI'],\n",
       " ['Male', 52, 'WHI'],\n",
       " ['Male', 52, 'WHI'],\n",
       " ['Female', 52, 'WHI'],\n",
       " ['Female', 52, 'WHI'],\n",
       " ['Male', 52, 'WHI'],\n",
       " ['Male', 52, 'WHI'],\n",
       " ['Female', 52, 'WHI'],\n",
       " ['Female', 52, 'WHI'],\n",
       " ['Female', 52, 'WHI'],\n",
       " ['Female', 53, 'WHI'],\n",
       " ['Female', 53, 'WHI'],\n",
       " ['Female', 53, 'WHI'],\n",
       " ['Female', 53, 'WHI'],\n",
       " ['Male', 53, 'WHI'],\n",
       " ['Female', 53, 'WHI'],\n",
       " ['Male', 53, 'WHI'],\n",
       " ['Female', 53, 'WHI'],\n",
       " ['Female', 53, 'WHI'],\n",
       " ['Male', 53, 'WHI'],\n",
       " ['Female', 53, 'WHI'],\n",
       " ['Male', 54, 'WHI'],\n",
       " ['Female', 54, 'WHI'],\n",
       " ['Female', 54, 'WHI'],\n",
       " ['Female', 54, 'WHI'],\n",
       " ['Male', 54, 'WHI'],\n",
       " ['Male', 54, 'WHI'],\n",
       " ['Female', 54, 'WHI'],\n",
       " ['Female', 54, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Male', 55, 'WHI'],\n",
       " ['Male', 55, 'WHI'],\n",
       " ['Male', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Male', 55, 'WHI'],\n",
       " ['Male', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Male', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Male', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Female', 55, 'WHI'],\n",
       " ['Male', 55, 'WHI'],\n",
       " ['Male', 55, 'WHI'],\n",
       " ['Male', 55, 'WHI'],\n",
       " ['Male', 55, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Male', 56, 'WHI'],\n",
       " ['Male', 56, 'WHI'],\n",
       " ['Male', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Male', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Male', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Male', 56, 'WHI'],\n",
       " ['Male', 56, 'WHI'],\n",
       " ['Male', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Male', 56, 'WHI'],\n",
       " ['Male', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 56, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Male', 57, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Male', 57, 'WHI'],\n",
       " ['Male', 57, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Male', 57, 'WHI'],\n",
       " ['Male', 57, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Male', 57, 'WHI'],\n",
       " ['Female', 57, 'WHI'],\n",
       " ['Male', 57, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Male', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Male', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Male', 58, 'WHI'],\n",
       " ['Male', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Male', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Male', 58, 'WHI'],\n",
       " ['Female', 58, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Male', 59, 'WHI'],\n",
       " ['Male', 59, 'WHI'],\n",
       " ['Male', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Male', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Male', 59, 'WHI'],\n",
       " ['Male', 59, 'WHI'],\n",
       " ['Male', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Male', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Male', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Female', 59, 'WHI'],\n",
       " ['Male', 59, 'WHI'],\n",
       " ['Male', 59, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Male', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 60, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Female', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Male', 61, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Female', 62, 'WHI'],\n",
       " ['Male', 62, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Male', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Female', 63, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Female', 64, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Male', 65, 'WHI'],\n",
       " ['Female', 65, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Male', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Male', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Male', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Male', 66, 'WHI'],\n",
       " ['Male', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Male', 66, 'WHI'],\n",
       " ['Male', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Male', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 66, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Male', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Female', 67, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Female', 68, 'WHI'],\n",
       " ['Male', 68, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Male', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 69, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Female', 70, 'WHI'],\n",
       " ['Male', 70, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 71, 'WHI'],\n",
       " ['Male', 71, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Male', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 72, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Male', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Male', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Male', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Male', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Male', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Male', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Male', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Male', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Male', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Male', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Female', 73, 'WHI'],\n",
       " ['Male', 73, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Male', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Male', 74, 'WHI'],\n",
       " ['Male', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Male', 74, 'WHI'],\n",
       " ['Male', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ['Female', 74, 'WHI'],\n",
       " ...]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query1 = '''SELECT sex, age, study\n",
    "FROM bigquery.scr_gecco_susceptibility.subject_phenotypes_multi\n",
    "'''\n",
    "\n",
    "res = cl.runQuery(query1)\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Retrieving the table list_\n",
      "____Page1_______________\n",
      "collections.public_datasets.dbgap_scr_ega_scr_egapancreatic_files\n",
      "collections.public_datasets.dbgap_scr_ega_scr_egapancreatic_sample_multi\n",
      "collections.public_datasets.dbgap_scr_gecco_susceptibility_gecco_sample_acc\n",
      "collections.public_datasets.dbgap_scr_gecco_susceptibility_phs001554_gecco_sddp_manifest\n",
      "collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi\n",
      "collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_multi\n",
      "collections.public_datasets.dbgap_scr_gecco_susceptibility_sb_drs_index\n",
      "collections.public_datasets.dbgap_scr_gecco_susceptibility_subject_multi\n",
      "collections.public_datasets.dbgap_scr_gecco_susceptibility_subject_phenotypes_multi\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['collections.public_datasets.dbgap_scr_ega_scr_egapancreatic_files',\n",
       " 'collections.public_datasets.dbgap_scr_ega_scr_egapancreatic_sample_multi',\n",
       " 'collections.public_datasets.dbgap_scr_gecco_susceptibility_gecco_sample_acc',\n",
       " 'collections.public_datasets.dbgap_scr_gecco_susceptibility_phs001554_gecco_sddp_manifest',\n",
       " 'collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi',\n",
       " 'collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_multi',\n",
       " 'collections.public_datasets.dbgap_scr_gecco_susceptibility_sb_drs_index',\n",
       " 'collections.public_datasets.dbgap_scr_gecco_susceptibility_subject_multi',\n",
       " 'collections.public_datasets.dbgap_scr_gecco_susceptibility_subject_phenotypes_multi']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTables('dbgap_demo')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This provides the full schema for sample attributes from the same study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablecollections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi_\n",
      "{\n",
      "   \"name\": \"collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi\",\n",
      "   \"description\": \"Automatically generated schema\",\n",
      "   \"data_model\": {\n",
      "      \"$id\": \"http://localhost:8089/table/collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi/info\",\n",
      "      \"description\": \"Automatically generated schema\",\n",
      "      \"$schema\": \"http://json-schema.org/draft-07/schema#\",\n",
      "      \"properties\": {\n",
      "         \"dbgap_sample_id\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"sample_id\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"body_site\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"analyte_type\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"is_tumor\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"dna_source\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"batch\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"read_depth\": {\n",
      "            \"format\": \"double\",\n",
      "            \"type\": \"number\",\n",
      "            \"$comment\": \"double\"\n",
      "         },\n",
      "         \"sequencing_center\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x12a3ee9d0>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTableInfo('collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi', verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dbgap_sample_id\n",
      "varchar\n",
      "_______________________________________\n",
      "sample_id\n",
      "varchar\n",
      "_______________________________________\n",
      "body_site\n",
      "varchar\n",
      "_______________________________________\n",
      "analyte_type\n",
      "varchar\n",
      "_______________________________________\n",
      "is_tumor\n",
      "varchar\n",
      "_______________________________________\n",
      "dna_source\n",
      "varchar\n",
      "_______________________________________\n",
      "batch\n",
      "varchar\n",
      "_______________________________________\n",
      "read_depth\n",
      "double\n",
      "_______________________________________\n",
      "sequencing_center\n",
      "varchar\n",
      "_______________________________________\n"
     ]
    }
   ],
   "source": [
    "cl.listTableColumns('collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi',descriptions=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### An EGA Table\n",
    "The same format of data dictionary was created for an EGA table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablecollections.public_datasets.dbgap_scr_ega_scr_egapancreatic_sample_multi_\n",
      "{\n",
      "   \"name\": \"collections.public_datasets.dbgap_scr_ega_scr_egapancreatic_sample_multi\",\n",
      "   \"description\": \"Automatically generated schema\",\n",
      "   \"data_model\": {\n",
      "      \"$id\": \"http://localhost:8089/table/collections.public_datasets.dbgap_scr_ega_scr_egapancreatic_sample_multi/info\",\n",
      "      \"description\": \"Automatically generated schema\",\n",
      "      \"$schema\": \"http://json-schema.org/draft-07/schema#\",\n",
      "      \"properties\": {\n",
      "         \"sample_primary_id\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"sample_submitter_id\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"disease_site\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"gender\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"phenotype\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"subject_id\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"ena_checklist\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"bam_file\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x12a3eea60>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTableInfo('collections.public_datasets.dbgap_scr_ega_scr_egapancreatic_sample_multi',verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dbgap_subject_id\n",
      "varchar\n",
      "varchar\n",
      "_______________________________________\n",
      "subject_id\n",
      "varchar\n",
      "varchar\n",
      "_______________________________________\n",
      "affection_status\n",
      "varchar\n",
      "varchar\n",
      "_______________________________________\n",
      "sex\n",
      "varchar\n",
      "varchar\n",
      "_______________________________________\n",
      "age\n",
      "bigint\n",
      "bigint\n",
      "_______________________________________\n",
      "race\n",
      "varchar\n",
      "varchar\n",
      "_______________________________________\n",
      "ethnicity\n",
      "varchar\n",
      "varchar\n",
      "_______________________________________\n",
      "study\n",
      "varchar\n",
      "varchar\n",
      "_______________________________________\n"
     ]
    }
   ],
   "source": [
    "cl.listTableColumns('bigquery.scr_gecco_susceptibility.subject_phenotypes_multi', descriptions=True, enums=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablecollections.public_datasets.dbgap_scr_gecco_susceptibility_subject_phenotypes_multi_\n",
      "{\n",
      "   \"name\": \"collections.public_datasets.dbgap_scr_gecco_susceptibility_subject_phenotypes_multi\",\n",
      "   \"data_model\": {\n",
      "      \"description\": \"\",\n",
      "      \"$id\": \"dbgap:pht007609.v1\",\n",
      "      \"properties\": {\n",
      "         \"STUDY\": {\n",
      "            \"$id\": \"dbgap:phv00357188.v1\",\n",
      "            \"description\": \"Study acronym\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"CPS-II\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"DACHS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"HPFS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"NHS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"PLCO\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"WHI\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"SUBJECT_ID\": {\n",
      "            \"$id\": \"dbgap:phv00357182.v1\",\n",
      "            \"description\": \"De-identified subject ID\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": []\n",
      "         },\n",
      "         \"SEX\": {\n",
      "            \"$id\": \"dbgap:phv00357184.v1\",\n",
      "            \"description\": \"Sex of participant\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"Female\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Male\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"RACE\": {\n",
      "            \"$id\": \"dbgap:phv00357186.v1\",\n",
      "            \"description\": \"Race of participant\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"White\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"ETHNICITY\": {\n",
      "            \"$id\": \"dbgap:phv00357187.v1\",\n",
      "            \"description\": \"Ethnicity of participant\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"Unknown\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"AFFECTION_STATUS\": {\n",
      "            \"$id\": \"dbgap:phv00357183.v1\",\n",
      "            \"description\": \"Case control status of the subject for colorectal cancer\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"Case\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Control\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"AGE\": {\n",
      "            \"$id\": \"dbgap:phv00357185.v1\",\n",
      "            \"description\": \"Participant reference age\",\n",
      "            \"type\": \"integer\",\n",
      "            \"$unit\": \"Years\",\n",
      "            \"oneOf\": []\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x12f7c5820>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTableInfo('collections.public_datasets.dbgap_scr_gecco_susceptibility_subject_phenotypes_multi', verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablecollections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi_\n",
      "{\n",
      "   \"name\": \"collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi\",\n",
      "   \"data_model\": {\n",
      "      \"description\": \"\",\n",
      "      \"$id\": \"dbgap:pht007610.v1\",\n",
      "      \"properties\": {\n",
      "         \"SEQUENCING_CENTER\": {\n",
      "            \"$id\": \"dbgap:phv00357196.v1\",\n",
      "            \"description\": \"Name of the center that conducted sequencing\",\n",
      "            \"type\": \"encoded value\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"1\",\n",
      "                  \"title\": \"University of Washington Sequencing Center (Seattle, WA, USA)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"2\",\n",
      "                  \"title\": \"Macrogen USA\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"ANALYTE_TYPE\": {\n",
      "            \"$id\": \"dbgap:phv00357191.v1\",\n",
      "            \"description\": \"Analyte Type\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"DNA\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"IS_TUMOR\": {\n",
      "            \"$id\": \"dbgap:phv00357192.v1\",\n",
      "            \"description\": \"Tumor status\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"No\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"DNA_SOURCE\": {\n",
      "            \"$id\": \"dbgap:phv00357193.v1\",\n",
      "            \"description\": \"DNA origin\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"Blood\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Buffy coat\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Unknown\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Whole blood\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"BATCH\": {\n",
      "            \"$id\": \"dbgap:phv00357194.v1\",\n",
      "            \"description\": \"Sequencing batch\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": []\n",
      "         },\n",
      "         \"READ_DEPTH\": {\n",
      "            \"$id\": \"dbgap:phv00357195.v1\",\n",
      "            \"description\": \"Sequencing coverage\",\n",
      "            \"type\": \"decimal\",\n",
      "            \"$unit\": \"fold coverage\",\n",
      "            \"oneOf\": []\n",
      "         },\n",
      "         \"BODY_SITE\": {\n",
      "            \"$id\": \"dbgap:phv00357190.v1\",\n",
      "            \"description\": \"Body site where sample was collected\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"Blood\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"SAMPLE_ID\": {\n",
      "            \"$id\": \"dbgap:phv00357189.v1\",\n",
      "            \"description\": \"De-identified Sample ID\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": []\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x12f77c7f0>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTableInfo('collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi', verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablecollections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi_\n",
      "{\n",
      "   \"name\": \"collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi\",\n",
      "   \"data_model\": {\n",
      "      \"description\": \"\",\n",
      "      \"$id\": \"dbgap:pht007610.v1\",\n",
      "      \"properties\": {\n",
      "         \"SEQUENCING_CENTER\": {\n",
      "            \"$id\": \"dbgap:phv00357196.v1\",\n",
      "            \"description\": \"Name of the center that conducted sequencing\",\n",
      "            \"type\": \"encoded value\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"1\",\n",
      "                  \"title\": \"University of Washington Sequencing Center (Seattle, WA, USA)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"2\",\n",
      "                  \"title\": \"Macrogen USA\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"ANALYTE_TYPE\": {\n",
      "            \"$id\": \"dbgap:phv00357191.v1\",\n",
      "            \"description\": \"Analyte Type\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"DNA\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"IS_TUMOR\": {\n",
      "            \"$id\": \"dbgap:phv00357192.v1\",\n",
      "            \"description\": \"Tumor status\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"No\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"DNA_SOURCE\": {\n",
      "            \"$id\": \"dbgap:phv00357193.v1\",\n",
      "            \"description\": \"DNA origin\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"Blood\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Buffy coat\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Unknown\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Whole blood\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"BATCH\": {\n",
      "            \"$id\": \"dbgap:phv00357194.v1\",\n",
      "            \"description\": \"Sequencing batch\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": []\n",
      "         },\n",
      "         \"READ_DEPTH\": {\n",
      "            \"$id\": \"dbgap:phv00357195.v1\",\n",
      "            \"description\": \"Sequencing coverage\",\n",
      "            \"type\": \"decimal\",\n",
      "            \"$unit\": \"fold coverage\",\n",
      "            \"oneOf\": []\n",
      "         },\n",
      "         \"BODY_SITE\": {\n",
      "            \"$id\": \"dbgap:phv00357190.v1\",\n",
      "            \"description\": \"Body site where sample was collected\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"Blood\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"SAMPLE_ID\": {\n",
      "            \"$id\": \"dbgap:phv00357189.v1\",\n",
      "            \"description\": \"De-identified Sample ID\",\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": []\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x12f8709a0>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTableInfo('collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_attributes_multi', verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablebigquery.scr_gecco_susceptibility.subject_phenotypes_multi_\n",
      "{\n",
      "   \"name\": \"bigquery.scr_gecco_susceptibility.subject_phenotypes_multi\",\n",
      "   \"data_model\": {\n",
      "      \"description\": \"\",\n",
      "      \"$id\": \"dbgap:pht007609.v1\",\n",
      "      \"properties\": {\n",
      "         \"STUDY\": {\n",
      "            \"$id\": \"dbgap:phv00357188.v1\",\n",
      "            \"description\": \"Study acronym\",\n",
      "            \"type\": \"string\"\n",
      "         },\n",
      "         \"SUBJECT_ID\": {\n",
      "            \"$id\": \"dbgap:phv00357182.v1\",\n",
      "            \"description\": \"De-identified subject ID\",\n",
      "            \"type\": \"string\"\n",
      "         },\n",
      "         \"SEX\": {\n",
      "            \"$id\": \"dbgap:phv00357184.v1\",\n",
      "            \"description\": \"Sex of participant\",\n",
      "            \"type\": \"string\"\n",
      "         },\n",
      "         \"RACE\": {\n",
      "            \"$id\": \"dbgap:phv00357186.v1\",\n",
      "            \"description\": \"Race of participant\",\n",
      "            \"type\": \"string\"\n",
      "         },\n",
      "         \"ETHNICITY\": {\n",
      "            \"$id\": \"dbgap:phv00357187.v1\",\n",
      "            \"description\": \"Ethnicity of participant\",\n",
      "            \"type\": \"string\"\n",
      "         },\n",
      "         \"AFFECTION_STATUS\": {\n",
      "            \"$id\": \"dbgap:phv00357183.v1\",\n",
      "            \"description\": \"Case control status of the subject for colorectal cancer\",\n",
      "            \"type\": \"string\"\n",
      "         },\n",
      "         \"AGE\": {\n",
      "            \"$id\": \"dbgap:phv00357185.v1\",\n",
      "            \"description\": \"Participant reference age\",\n",
      "            \"type\": \"integer\",\n",
      "            \"$unit\": \"Years\"\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x10f286bb0>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTableInfo('bigquery.scr_gecco_susceptibility.subject_phenotypes_multi', verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablecollections.public_datasets.dbgap_scr_ega_scr_egapancreatic_sample_multi_\n",
      "{\n",
      "   \"name\": \"collections.public_datasets.dbgap_scr_ega_scr_egapancreatic_sample_multi\",\n",
      "   \"data_model\": {\n",
      "      \"$id\": \"scr_EGApancreatic_sample.data_dict.xml.scr_EGApancreatic_sample\",\n",
      "      \"$schema\": \"http://json-schema.org/draft-07/schema\",\n",
      "      \"properties\": {\n",
      "         \"gender\": {\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"female\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"male\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Sex of participant\"\n",
      "         },\n",
      "         \"bam_file\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Relative path for alignment file\"\n",
      "         },\n",
      "         \"phenotype\": {\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"lung tissue\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"muscle tissue\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"pancreatic adenocarcinoma\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"skin tissue\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"spleen tissue\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Site independent tissue type\"\n",
      "         },\n",
      "         \"subject_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Subject Id used locally by submitter\"\n",
      "         },\n",
      "         \"disease_site\": {\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"Liver_Metastasis\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"Lung_Metastasis\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"Lung_Normal\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"Lymph_Node_Metastasis\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"Muscle_Normal\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"Normal Pancreas\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"Pancreas_Primary_Tumor\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"Peritoneal_Metastasis\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"Skin_Normal\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"Spleen_Normal\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"body site sample was taken from\"\n",
      "         },\n",
      "         \"ena-checklist\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Minimal information checklist used in this study\"\n",
      "         },\n",
      "         \"sample_primary_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"De-identified sample ID\"\n",
      "         },\n",
      "         \"sample_submitter_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Sample id used locally by submitter\"\n",
      "         }\n",
      "      },\n",
      "      \"type\": \"object\"\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x12fff4cd0>"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTableInfo('collections.public_datasets.dbgap_scr_ega_scr_egapancreatic_sample_multi', verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablecollections.public_datasets.dbgap_scr_gecco_susceptibility_sample_multi_\n",
      "{\n",
      "   \"name\": \"collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_multi\",\n",
      "   \"description\": \"Scrambled version of sample crossreference data for phs001554 Colorectal cancer susceptibility study.\",\n",
      "   \"data_model\": {\n",
      "      \"$id\": \"phs001554.v1.pht007608.v1.GECCO_CRC_Susceptibility_Sample\",\n",
      "      \"description\": \"Scrambled version of sample crossreference data for phs001554 Colorectal cancer susceptibility study.\",\n",
      "      \"$schema\": \"http://json-schema.org/draft-07/schema\",\n",
      "      \"properties\": {\n",
      "         \"sample_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Sample ID\"\n",
      "         },\n",
      "         \"sample_use\": {\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"Seq_DNA_SNP\",\n",
      "                  \"title\": \"SNP genotypes derived from sequence data\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Sample Use\"\n",
      "         },\n",
      "         \"subject_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Subject ID\"\n",
      "         },\n",
      "         \"sample_source\": {\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"WHI\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Source repository (study) where samples originate\"\n",
      "         },\n",
      "         \"source_sample_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Subject ID used in the Source Repository\"\n",
      "         }\n",
      "      },\n",
      "      \"type\": \"object\"\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x130c12190>"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTableInfo('collections.public_datasets.dbgap_scr_gecco_susceptibility_sample_multi', verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablecollections.public_datasets.dbgap_scr_gecco_susceptibility_subject_multi_\n",
      "{\n",
      "   \"name\": \"collections.public_datasets.dbgap_scr_gecco_susceptibility_subject_multi\",\n",
      "   \"description\": \"Scrambled version of subject crossreference data for phs001554 Colorectal cancer susceptibility study.\",\n",
      "   \"data_model\": {\n",
      "      \"$id\": \"phs001554.v1.pht007607.v1.GECCO_CRC_Susceptibility_Subject\",\n",
      "      \"description\": \"Scrambled version of subject crossreference data for phs001554 Colorectal cancer susceptibility study.\",\n",
      "      \"$schema\": \"http://json-schema.org/draft-07/schema\",\n",
      "      \"properties\": {\n",
      "         \"consent\": {\n",
      "            \"type\": \"string\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"1\",\n",
      "                  \"title\": \"General Research Use (GRU)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"2\",\n",
      "                  \"title\": \"Health/Medical/Biomedical (IRB) (HMB-IRB)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"3\",\n",
      "                  \"title\": \"Health/Medical/Biomedical (IRB, NPU) (HMB-IRB-NPU)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"4\",\n",
      "                  \"title\": \"Disease-Specific (Colorectal Cancer, NPU, GSO) (DS-CC-NPU-GSO)\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"5\",\n",
      "                  \"title\": \"Cancer in all age groups, other diseases in adults only, and methods (CADM)\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Consent group as determined by DAC\"\n",
      "         },\n",
      "         \"subject_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Subject ID\"\n",
      "         },\n",
      "         \"subject_source\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Source repository where subjects originate\"\n",
      "         },\n",
      "         \"source_subject_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Subject ID used in the Source Repository\"\n",
      "         }\n",
      "      },\n",
      "      \"type\": \"object\"\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x1309d7a60>"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTableInfo('collections.public_datasets.dbgap_scr_gecco_susceptibility_subject_multi', verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
