{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Retrieving the catalog list_\n",
      "cshcodeathon\n",
      "data_lake\n",
      "dbgap_demo\n",
      "kidsfirst\n",
      "sample_phenopackets\n",
      "search_cloud\n",
      "search_postgres\n",
      "search_postgres_pgpc\n",
      "thousand_genomes\n"
     ]
    }
   ],
   "source": [
    "from fasp.search import DiscoverySearchClient\n",
    "cl = DiscoverySearchClient('https://ga4gh-search-adapter-presto-public.prod.dnastack.com')\n",
    "cl.listCatalogs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "program_name\n",
      "_______________________________________\n",
      "case_barcode\n",
      "_______________________________________\n",
      "case_gdc_id\n",
      "_______________________________________\n",
      "program_dbgap_accession_number\n",
      "_______________________________________\n",
      "project_short_name\n",
      "_______________________________________\n",
      "project_name\n",
      "_______________________________________\n",
      "disease_code\n",
      "_______________________________________\n",
      "gender\n",
      "_______________________________________\n",
      "vital_status\n",
      "_______________________________________\n",
      "race\n",
      "_______________________________________\n",
      "ethnicity\n",
      "_______________________________________\n",
      "age_at_diagnosis\n",
      "_______________________________________\n",
      "days_to_birth\n",
      "_______________________________________\n",
      "days_to_death\n",
      "_______________________________________\n",
      "days_to_initial_pathologic_diagnosis\n",
      "_______________________________________\n",
      "days_to_last_followup\n",
      "_______________________________________\n",
      "days_to_last_known_alive\n",
      "_______________________________________\n",
      "days_to_submitted_specimen_dx\n",
      "_______________________________________\n",
      "clinical_stage\n",
      "_______________________________________\n",
      "clinical_t\n",
      "_______________________________________\n",
      "clinical_n\n",
      "_______________________________________\n",
      "clinical_m\n",
      "_______________________________________\n",
      "pathologic_stage\n",
      "_______________________________________\n",
      "pathologic_t\n",
      "_______________________________________\n",
      "pathologic_n\n",
      "_______________________________________\n",
      "pathologic_m\n",
      "_______________________________________\n",
      "year_of_initial_pathologic_diagnosis\n",
      "_______________________________________\n",
      "tumor_tissue_site\n",
      "_______________________________________\n",
      "primary_neoplasm_melanoma_dx\n",
      "_______________________________________\n",
      "anatomic_neoplasm_subdivision\n",
      "_______________________________________\n",
      "country\n",
      "_______________________________________\n",
      "other_dx\n",
      "_______________________________________\n",
      "other_malignancy_anatomic_site\n",
      "_______________________________________\n",
      "other_malignancy_type\n",
      "_______________________________________\n",
      "other_malignancy_histological_type\n",
      "_______________________________________\n",
      "history_of_neoadjuvant_treatment\n",
      "_______________________________________\n",
      "primary_therapy_outcome_success\n",
      "_______________________________________\n",
      "histological_type\n",
      "_______________________________________\n",
      "neoplasm_histologic_grade\n",
      "_______________________________________\n",
      "icd_10\n",
      "_______________________________________\n",
      "icd_o_3_site\n",
      "_______________________________________\n",
      "icd_o_3_histology\n",
      "_______________________________________\n",
      "person_neoplasm_cancer_status\n",
      "_______________________________________\n",
      "residual_tumor\n",
      "_______________________________________\n",
      "tumor_type\n",
      "_______________________________________\n",
      "new_tumor_event_after_initial_treatment\n",
      "_______________________________________\n",
      "lymphnodes_examined\n",
      "_______________________________________\n",
      "number_of_lymphnodes_examined\n",
      "_______________________________________\n",
      "number_of_lymphnodes_positive_by_he\n",
      "_______________________________________\n",
      "lymphatic_invasion\n",
      "_______________________________________\n",
      "venous_invasion\n",
      "_______________________________________\n",
      "lymphovascular_invasion_present\n",
      "_______________________________________\n",
      "bcr\n",
      "_______________________________________\n",
      "batch_number\n",
      "_______________________________________\n",
      "tss_code\n",
      "_______________________________________\n",
      "age_began_smoking_in_years\n",
      "_______________________________________\n",
      "year_of_tobacco_smoking_onset\n",
      "_______________________________________\n",
      "stopped_smoking_year\n",
      "_______________________________________\n",
      "tobacco_smoking_history\n",
      "_______________________________________\n",
      "number_pack_years_smoked\n",
      "_______________________________________\n",
      "height\n",
      "_______________________________________\n",
      "weight\n",
      "_______________________________________\n",
      "bmi\n",
      "_______________________________________\n",
      "mononucleotide_and_dinucleotide_marker_panel_analysis_status\n",
      "_______________________________________\n",
      "menopause_status\n",
      "_______________________________________\n",
      "pregnancies\n",
      "_______________________________________\n",
      "hpv_status\n",
      "_______________________________________\n",
      "hpv_calls\n",
      "_______________________________________\n",
      "h_pylori_infection\n",
      "_______________________________________\n",
      "gleason_score_combined\n",
      "_______________________________________\n",
      "psa_value\n",
      "_______________________________________\n",
      "colorectal_cancer\n",
      "_______________________________________\n",
      "history_of_colon_polyps\n",
      "_______________________________________\n"
     ]
    }
   ],
   "source": [
    "cl.listTableColumns('search_cloud.cshcodeathon.tcga_bioclin_v0_clinical')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Retrieving the table list_\n",
      "____Page1_______________\n",
      "search_cloud.brca_exchange.v32\n",
      "search_cloud.clinvar.allele_gene\n",
      "search_cloud.clinvar.cross_references\n",
      "search_cloud.clinvar.gene_specific_summary\n",
      "search_cloud.clinvar.hgvs4variation\n",
      "search_cloud.clinvar.organization_summary\n",
      "search_cloud.clinvar.submission_summary\n",
      "search_cloud.clinvar.summary_of_conflicting_interpretations\n",
      "search_cloud.clinvar.var_citations\n",
      "search_cloud.clinvar.variant_summary\n",
      "search_cloud.clinvar.variation_allele\n",
      "search_cloud.covid19_jhu_csse.confirmed_cases\n",
      "search_cloud.covid19_jhu_csse.deaths\n",
      "search_cloud.covid19_jhu_csse.recovered_cases\n",
      "search_cloud.covid19_jhu_csse.sandbox_main\n",
      "search_cloud.covid19_jhu_csse.summary\n",
      "search_cloud.cshcodeathon.gdc_case_file_metadata_filedata_active_current\n",
      "search_cloud.cshcodeathon.gdc_case_file_metadata_gdcfileid_to_gcsurl_current\n",
      "search_cloud.cshcodeathon.gdc_rel24_filedata_active\n",
      "search_cloud.cshcodeathon.gecco_gen3_drs_index\n",
      "search_cloud.cshcodeathon.gecco_sra_drs_index\n",
      "search_cloud.cshcodeathon.isb-cgc_gdc_metadata_rel24_filedata_active\n",
      "search_cloud.cshcodeathon.md_mapping\n",
      "search_cloud.cshcodeathon.md_mapping_term\n",
      "search_cloud.cshcodeathon.md_value_map\n",
      "search_cloud.cshcodeathon.organoid_profiling_pc_subject_phenotypes_gru\n",
      "search_cloud.cshcodeathon.tcga_bioclin_v0_clinical\n",
      "search_cloud.cshcodeathon.tcga_clinical_gdc_current\n",
      "search_cloud.cshcodeathon.tcga_somatic_mutation_hg38_gdc_current\n",
      "search_cloud.cshcodeathon.test\n",
      "search_cloud.genomics_england.activities\n",
      "search_cloud.genomics_england.genes\n",
      "search_cloud.genomics_england.panels\n",
      "search_cloud.genomics_england.regions\n",
      "search_cloud.genomics_england.strs\n",
      "search_cloud.human_cell_atlas.hca\n",
      "search_cloud.refseq.gencode_v32\n",
      "search_cloud.refseq.refseq_genes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['search_cloud.brca_exchange.v32',\n",
       " 'search_cloud.clinvar.allele_gene',\n",
       " 'search_cloud.clinvar.cross_references',\n",
       " 'search_cloud.clinvar.gene_specific_summary',\n",
       " 'search_cloud.clinvar.hgvs4variation',\n",
       " 'search_cloud.clinvar.organization_summary',\n",
       " 'search_cloud.clinvar.submission_summary',\n",
       " 'search_cloud.clinvar.summary_of_conflicting_interpretations',\n",
       " 'search_cloud.clinvar.var_citations',\n",
       " 'search_cloud.clinvar.variant_summary',\n",
       " 'search_cloud.clinvar.variation_allele',\n",
       " 'search_cloud.covid19_jhu_csse.confirmed_cases',\n",
       " 'search_cloud.covid19_jhu_csse.deaths',\n",
       " 'search_cloud.covid19_jhu_csse.recovered_cases',\n",
       " 'search_cloud.covid19_jhu_csse.sandbox_main',\n",
       " 'search_cloud.covid19_jhu_csse.summary',\n",
       " 'search_cloud.cshcodeathon.gdc_case_file_metadata_filedata_active_current',\n",
       " 'search_cloud.cshcodeathon.gdc_case_file_metadata_gdcfileid_to_gcsurl_current',\n",
       " 'search_cloud.cshcodeathon.gdc_rel24_filedata_active',\n",
       " 'search_cloud.cshcodeathon.gecco_gen3_drs_index',\n",
       " 'search_cloud.cshcodeathon.gecco_sra_drs_index',\n",
       " 'search_cloud.cshcodeathon.isb-cgc_gdc_metadata_rel24_filedata_active',\n",
       " 'search_cloud.cshcodeathon.md_mapping',\n",
       " 'search_cloud.cshcodeathon.md_mapping_term',\n",
       " 'search_cloud.cshcodeathon.md_value_map',\n",
       " 'search_cloud.cshcodeathon.organoid_profiling_pc_subject_phenotypes_gru',\n",
       " 'search_cloud.cshcodeathon.tcga_bioclin_v0_clinical',\n",
       " 'search_cloud.cshcodeathon.tcga_clinical_gdc_current',\n",
       " 'search_cloud.cshcodeathon.tcga_somatic_mutation_hg38_gdc_current',\n",
       " 'search_cloud.cshcodeathon.test',\n",
       " 'search_cloud.genomics_england.activities',\n",
       " 'search_cloud.genomics_england.genes',\n",
       " 'search_cloud.genomics_england.panels',\n",
       " 'search_cloud.genomics_england.regions',\n",
       " 'search_cloud.genomics_england.strs',\n",
       " 'search_cloud.human_cell_atlas.hca',\n",
       " 'search_cloud.refseq.gencode_v32',\n",
       " 'search_cloud.refseq.refseq_genes']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTables('search_cloud')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'search_cloud.cshcodeathon.tcga_bioclin_v0_clinical',\n",
       " 'description': 'Automatically generated schema',\n",
       " 'data_model': {'$id': 'https://ga4gh-search-adapter-presto-public.prod.dnastack.com/table/search_cloud.cshcodeathon.tcga_bioclin_v0_clinical/info',\n",
       "  'description': 'Automatically generated schema',\n",
       "  '$schema': 'http://json-schema.org/draft-07/schema#',\n",
       "  'properties': {'program_name': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Program Name, always TCGA (The Cancer Genome Atlas) in this table.'},\n",
       "   'case_barcode': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'TCGA patient/case barcode; e.g. TCGA-06-0119'},\n",
       "   'case_gdc_id': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'GDC unique identifer for this case; e.g. 7b4ce492-ff40-4bf1-b3e8-75ff83e8746d. This identifier can be used to find more information at the GDC Data Portal ( https://portal.gdc.cancer.gov/cases/7b4ce492-ff40-4bf1-b3e8-75ff83e8746d )'},\n",
       "   'program_dbgap_accession_number': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Program dbGaP accession number, eg phs000178'},\n",
       "   'project_short_name': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Project name abbreviation; the program name appended with a project name abbreviation; e.g. TCGA-OV, etc.'},\n",
       "   'project_name': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Project Name; frequently the same as or similar to the project disease type eg   Acute Lymphoblastic Leukemia - Phase II   or   Breast Invasive Carcinoma   etc'},\n",
       "   'disease_code': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'A code representing the type of cancer. Values can be found at https://gdc.cancer.gov/resources-TCGA-users/TCGA-code-tables/TCGA-study-abbreviations; e.g. OV, GBM, LUAD'},\n",
       "   'gender': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Participant gender; e.g. Male, Female'},\n",
       "   'vital_status': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Survival state of the participant; e.g. dead, alive'},\n",
       "   'race': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The race of a participant, independent from ethnicity; e.g. White, Black or African American, Asian, Native Hawaiian or other Pacific Islander, American Indian or Alaska Native'},\n",
       "   'ethnicity': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The ethnic origination of a participant, independent from race; e.g. Not Hispanic or Latino, Hispanic or Latino'},\n",
       "   'age_at_diagnosis': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Age, in years, at which a condition or disease was first diagnosed'},\n",
       "   'days_to_birth': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Days since birth (a negative value, relative to the date of diagnosis which is considered to be t=0)'},\n",
       "   'days_to_death': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': \"If vital_status indicates 'dead', then this is the number of days to death (a positive value) relative to the date of diagnosis.\"},\n",
       "   'days_to_initial_pathologic_diagnosis': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'The number of days to initial pathologic diagnosis'},\n",
       "   'days_to_last_followup': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': \"If vital_status indicates 'alive', then this is the number of days from the date of diagnosis to the last followup (communication contact).\"},\n",
       "   'days_to_last_known_alive': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': \"Number of days from the date of diagnosis to the last day on which a person was known to be alive. This field is a combination of the days_to_last_followup and days_to death fields. Therefore, this field can be used whether the vital_status is 'alive' or 'dead'.\"},\n",
       "   'days_to_submitted_specimen_dx': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Numeric value to represent the day of the diagnosis for the submitted sample.'},\n",
       "   'clinical_stage': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The stage of the cancer, based on results of tests done before surgery, such as physical exams and imaging scans; e.g. Stage I, Stage II, Stage III, Stage IV'},\n",
       "   'clinical_t': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The clinical T score is a rating of the extent of the primary tumor, based on results of tests done before surgery, such as physical exams and imaging scans; e.g. T1, T2, T3, T4'},\n",
       "   'clinical_n': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The clinical N score is a rating of the extent of cancer within nearby lymph nodes, based on results of tests done before surgery, such as physical exams and imaging scans; e.g. N1, N2, N3'},\n",
       "   'clinical_m': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The clinical M score is a rating of metastasis, that is, how much cancer has spread to distant sites. It is based on results of tests done before surgery, such as physical exams and imaging scans; e.g. MX, M0, M1'},\n",
       "   'pathologic_stage': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The stage of the cancer, based on surgery or biopsy; e.g. Stage I, Stage II, Stage III, Stage IV'},\n",
       "   'pathologic_t': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The pathologic T score is a rating of the extent of the primary tumor, based on surgery or biopsy; e.g. T1, T2, T3, T4'},\n",
       "   'pathologic_n': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The pathologic N score is a rating of the extent of cancer within nearby lymph nodes, based on surgery or biopsy; e.g. N1, N2, N3'},\n",
       "   'pathologic_m': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The pathologic M score is a rating of metastasis, that is, how much cancer has spread to distant sites. It is based on surgery or biopsy; e.g. MX, M0, M1'},\n",
       "   'year_of_initial_pathologic_diagnosis': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Four digit year of the initial pathologic diagnosis'},\n",
       "   'tumor_tissue_site': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The location on the body where the tumor exists; e.g. Brain, Ovary, Lung, etc.'},\n",
       "   'primary_neoplasm_melanoma_dx': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicates whether a person had a primary diagnosis of melanoma; e.g. Yes, No'},\n",
       "   'anatomic_neoplasm_subdivision': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The spatial location, subdivisions and/or atomic site name of a tumor; e.g. Rectum, Larynx, Tonsil, etc.'},\n",
       "   'country': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Country of the participant; e.g. Germany, United States, Canada, etc.'},\n",
       "   'other_dx': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicator and type of other diagnosis; e.g. No; Yes, History of Prior Malignancy; Yes, History of Synchronous/Bilateral Malignancy; Both History of Synchronous/Bilateral and Prior Malignancy'},\n",
       "   'other_malignancy_anatomic_site': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The site on the body where the other malignancy occurred; e.g. Bladder, Uterus, etc.'},\n",
       "   'other_malignancy_type': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicates whether another malignancy occurred before or at the same time; e.g. Prior Malignancy, Synchronous Malignancy'},\n",
       "   'other_malignancy_histological_type': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The histological type of cancer of the other malignancy; e.g. Adenosquamous, Astrocytoma, etc.'},\n",
       "   'history_of_neoadjuvant_treatment': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"Describes the patient's history of neoadjuvant treatment and the kind of treament given prior to resection of the tumor; e.g. Yes, Pharmaceutical Treatment Prior to Resection, etc.\"},\n",
       "   'primary_therapy_outcome_success': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Kind of outcome achieved after the completion of the initial course of treatment; e.g. Complete Remission/Response, Progressive Disease, etc.'},\n",
       "   'histological_type': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The histological type of cancer; e.g. Adenosquamous, Astrocytoma, etc.'},\n",
       "   'neoplasm_histologic_grade': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Rating which expresses the degree of abnormality of cancer cells, a measure of differentiation and aggressiveness; e.g. G1, G2, G3, G4'},\n",
       "   'icd_10': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Alphanumeric value used to describe the disease code from the tenth version of the International Classification of Disease (ICD-10)'},\n",
       "   'icd_o_3_site': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Code that represents the site of the disease using the third edition of the International Classification of Diseases for Oncology, published in 2000, used principally in tumor and cancer registries for coding the site (topography) and the histology (morphology) of neoplasms'},\n",
       "   'icd_o_3_histology': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Code that represents the histology of the disease using the third edition of the International Classification of Diseases for Oncology, published in 2000, used principally in tumor and cancer registries for coding the site (topography) and the histology (morphology) of neoplasms'},\n",
       "   'person_neoplasm_cancer_status': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"The participant's cancer status; e.g. With Tumor, Tumor Free\"},\n",
       "   'residual_tumor': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Status of the tissue margin following surgical resection; e.g. R0, R1, R2, RX'},\n",
       "   'tumor_type': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Tumor Type; e.g. Primary, Type 1, Type 2'},\n",
       "   'new_tumor_event_after_initial_treatment': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicates whether a patient has had a new tumor event after initial treatment; Yes, No'},\n",
       "   'lymphnodes_examined': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicator whether the lymph nodes were examined; e.g. Yes, No'},\n",
       "   'number_of_lymphnodes_examined': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'The number of lymph nodes examined'},\n",
       "   'number_of_lymphnodes_positive_by_he': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'The number of lymph nodes involved with disease as determined by histological examination'},\n",
       "   'lymphatic_invasion': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicator if small or thin-walled vessel invasion is present, indicating lymphatic involvement; e.g. Yes, No'},\n",
       "   'venous_invasion': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicator if large vessel or venous invasion was detected by surgery or presence in a tumor specimen; e.g. Yes, No'},\n",
       "   'lymphovascular_invasion_present': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicator if  invasion of a cancer to the blood vessels and/or lymphatics invasion is present; e.g. Yes, No'},\n",
       "   'bcr': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"Biospecimen Core Resource; e.g. Nationwide Children's Hospital, Washington University\"},\n",
       "   'batch_number': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'The BCR Batch Code; see https://gdc.cancer.gov/resources-TCGA-users/TCGA-code-tables/bcr-batch-codes for explanation of the codes'},\n",
       "   'tss_code': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Tissue Source Site Code; see https://gdc.cancer.gov/resources-TCGA-users/TCGA-code-tables/tissue-source-site-codes for explanation of codes'},\n",
       "   'age_began_smoking_in_years': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Age, in years, that the participant began smoking'},\n",
       "   'year_of_tobacco_smoking_onset': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Four-digit year that the participant started smoking; e.g. 2010'},\n",
       "   'stopped_smoking_year': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Four-digit year that the participant stopped smoking; e.g. 2010'},\n",
       "   'tobacco_smoking_history': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Category describing current smoking status and smoking history as self-reported by a participant.; e.g. 1, 2, etc.'},\n",
       "   'number_pack_years_smoked': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': 'The number of packs smoked per day times the number of years smoked by the participant'},\n",
       "   'height': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': 'Height of participant in centimeters'},\n",
       "   'weight': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': 'Weight of participant in kilograms'},\n",
       "   'bmi': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': 'Body Mass Index of participant'},\n",
       "   'mononucleotide_and_dinucleotide_marker_panel_analysis_status': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Status of the mononucleotide and dinucleotide marker panel analysis; microsatellite-stable (MSS), low level MSI (MSI-L) if less than 40% of markers were altered and high level MSI (MSI-H) if greater than 40% of markers were altered'},\n",
       "   'menopause_status': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"A female participant's menopausal status; e.g. Post, Pre, Indeterminate, Peri\"},\n",
       "   'pregnancies': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The number of times (including currently if it applies) that a female participant has been pregnant, regardless of outcome.'},\n",
       "   'hpv_status': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicates whether the participant has a record of a positive finding for human papillomavirus; e.g. Positive, Negative, Indeterminate'},\n",
       "   'hpv_calls': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'HPV strains; e.g. HPV18, HPV45, etc.'},\n",
       "   'h_pylori_infection': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicator of previous or current diagnosis of Helicobactor pylori infection; e.g. Yes, No, Never, Current'},\n",
       "   'gleason_score_combined': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'The Gleason Score (ranging from 1 -5) is the grade of the aggressiveness of prostate cancer. The combined Gleason Score is the primary Gleason score  (for the largest area of the tumor) added to the secondary Gleason Score (for the next largest section of the tumor).'},\n",
       "   'psa_value': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': \"A male participant's Prostrate-Specific Antigen (PSA) value\"},\n",
       "   'colorectal_cancer': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicates whether the participant has colorectal cancer; e.g. Yes, No'},\n",
       "   'history_of_colon_polyps': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicator if the participant had a previous history of colon polyps as noted in the history/physical or previous endoscopic report(s); e.g. Yes, No'}}}}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.listTableInfo('search_cloud.cshcodeathon.tcga_bioclin_v0_clinical')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
