{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run workflow on Kid First DRS files\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "kf_drs_ids = ['7c286da0-5245-415b-b5aa-b4b23e9cffdc',\n",
    "'0024c033-2942-40aa-8a45-a76504db3dd1',\n",
    "'c9772e5d-01ab-4d40-b42d-1721a4e39a6e']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'access_methods': [{'access_id': 's3',\n",
       "   'access_url': {'url': 's3://kf-seq-data-hudsonalpha/haib17JM4544/SL264200/SL264200.bam'},\n",
       "   'region': '',\n",
       "   'type': 's3'}],\n",
       " 'aliases': [],\n",
       " 'checksums': [{'checksum': '6c74d06c9b90326cdb12a58a0d4702a302b5a7379f1cf1ae5314865f656a8a76',\n",
       "   'type': 'sha256'},\n",
       "  {'checksum': 'd7bbad4382263bb697fc61925e474d27', 'type': 'md5'}],\n",
       " 'contents': [],\n",
       " 'created_time': '2021-07-13T18:55:52.059551',\n",
       " 'description': None,\n",
       " 'form': 'object',\n",
       " 'id': '0024c033-2942-40aa-8a45-a76504db3dd1',\n",
       " 'mime_type': 'application/json',\n",
       " 'name': 'SL264200.bam',\n",
       " 'self_uri': 'drs://data.kidsfirstdrc.org/0024c033-2942-40aa-8a45-a76504db3dd1',\n",
       " 'size': 65185924261,\n",
       " 'updated_time': '2021-07-13T18:55:52.059558',\n",
       " 'version': '675158e7'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from fasp.loc import kfDRSClient\n",
    "drsClient = kfDRSClient('~/.keys/kf_credentials.json','s3', debug=False)\n",
    "drsClient.getObject(kf_drs_ids[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = drsClient.getAccessURL(kf_drs_ids[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://kf-seq-data-hudsonalpha.s3.amazonaws.com/haib17JM4544/SL264200/SL264200.bam?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIZAHWHQWDB7XV4RA%2F20210930%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210930T020946Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&user_id=136&username=forei&X-Amz-Signature=718824cb9f89f0dd90f18d2de4d9ebe67c7f0ce84b3fed89425ca2aa1c5f9cf6\n"
     ]
    }
   ],
   "source": [
    "print(url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = drsClient.getAccessURL('16712090-50f7-4cd1-bf2d-90ce989c2139')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://kf-study-us-east-1-prd-sd-m3dbxd12.s3.amazonaws.com/harmonized/simple-variants/77af1324-3754-4e34-a208-d1342a2f2ca6.mutect2_somatic.vep.maf?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIZAHWHQWDB7XV4RA%2F20210930%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210930T020953Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&user_id=136&username=forei&X-Amz-Signature=9cb8c257e03ef8cb41376c23222bd73767e76df76e5bbcdc0e4a8c8ae488b0d7\n"
     ]
    }
   ],
   "source": [
    "print(url)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kids First BigQuery search (Neuroblastoma) - SB WES through to results retrieval and aggregation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/forei/imfenvs/fasp/lib/python3.9/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.\n",
      "  warnings.warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running /var/folders/wz/jjbxsnr13v7dkw2jkbpmqd_dly65gq/T/ipykernel_79294/1484740848.py\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id</th>\n",
       "      <th>latest_did</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GF_034F7RMB</td>\n",
       "      <td>795c9e2d-9f25-4996-ae7b-97da70419e33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GF_04D094R4</td>\n",
       "      <td>18b108ed-fad6-44b7-a410-9ed2c1fc821f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       file_id                            latest_did\n",
       "0  GF_034F7RMB  795c9e2d-9f25-4996-ae7b-97da70419e33\n",
       "1  GF_04D094R4  18b108ed-fad6-44b7-a410-9ed2c1fc821f"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from fasp.search import BigQuerySearchClient\n",
    "from fasp.workflow import sbcgcWESClient\n",
    "from fasp.runner import FASPRunner\n",
    "\n",
    "faspRunner = FASPRunner()\n",
    "runNote = 'KF Neuroblastoma cram files retrieved from Gen3 DRS and computed on SB'\n",
    "\n",
    "searchClient = BigQuerySearchClient()\n",
    "query = '''SELECT file_id, latest_did\n",
    " FROM `isbcgc-216220.kidsfirst.neuroblastoma_participant_family` \n",
    "where file_format = 'bam' \n",
    "and file_id in ('GF_034F7RMB','GF_04D094R4')'''\n",
    "df = searchClient.runQuery(query, returnType='dataframe')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running query\n",
      "SELECT file_id, latest_did\n",
      " FROM `isbcgc-216220.kidsfirst.neuroblastoma_participant_family` \n",
      "where file_format = 'bam' \n",
      "and file_id in ('GF_034F7RMB','GF_04D094R4')\n",
      "subject=GF_034F7RMB, drsID=795c9e2d-9f25-4996-ae7b-97da70419e33\n",
      "https://cgc-ga4gh-api.sbgenomics.com/ga4gh/wes/v1\n",
      "workflow submitted, run:b4687245-5202-4e2c-ba8a-0e8dfb64db9a\n",
      "____________________________________________________________\n",
      "subject=GF_04D094R4, drsID=18b108ed-fad6-44b7-a410-9ed2c1fc821f\n",
      "https://cgc-ga4gh-api.sbgenomics.com/ga4gh/wes/v1\n",
      "workflow submitted, run:7e77c0f4-803c-4508-85d9-efd20c3f00b4\n",
      "____________________________________________________________\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>run_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GF_034F7RMB</td>\n",
       "      <td>b4687245-5202-4e2c-ba8a-0e8dfb64db9a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GF_04D094R4</td>\n",
       "      <td>7e77c0f4-803c-4508-85d9-efd20c3f00b4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       subject                                run_id\n",
       "0  GF_034F7RMB  b4687245-5202-4e2c-ba8a-0e8dfb64db9a\n",
       "1  GF_04D094R4  7e77c0f4-803c-4508-85d9-efd20c3f00b4"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "settings = faspRunner.settings\n",
    "wesClient = sbcgcWESClient(settings['SevenBridgesProject'])\n",
    "\n",
    "faspRunner.configure(searchClient, drsClient, wesClient)\n",
    "\n",
    "\n",
    "run_ids = faspRunner.runQuery(query, runNote)\n",
    "import pandas as pd\n",
    "runs_df = pd.DataFrame(run_ids)\n",
    "runs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id</th>\n",
       "      <th>latest_did</th>\n",
       "      <th>run_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GF_034F7RMB</td>\n",
       "      <td>795c9e2d-9f25-4996-ae7b-97da70419e33</td>\n",
       "      <td>b4687245-5202-4e2c-ba8a-0e8dfb64db9a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GF_04D094R4</td>\n",
       "      <td>18b108ed-fad6-44b7-a410-9ed2c1fc821f</td>\n",
       "      <td>7e77c0f4-803c-4508-85d9-efd20c3f00b4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       file_id                            latest_did  \\\n",
       "0  GF_034F7RMB  795c9e2d-9f25-4996-ae7b-97da70419e33   \n",
       "1  GF_04D094R4  18b108ed-fad6-44b7-a410-9ed2c1fc821f   \n",
       "\n",
       "                                 run_id  \n",
       "0  b4687245-5202-4e2c-ba8a-0e8dfb64db9a  \n",
       "1  7e77c0f4-803c-4508-85d9-efd20c3f00b4  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df = pd.merge(\n",
    "    df,\n",
    "    runs_df,\n",
    "    how=\"inner\",\n",
    "    left_on='file_id',\n",
    "    right_on='subject'\n",
    ")\n",
    "merged_df.drop('subject', axis=1, inplace=True)\n",
    "#merged_df.to_csv('~/mywork.csv')\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id</th>\n",
       "      <th>latest_did</th>\n",
       "      <th>run_id</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GF_034F7RMB</td>\n",
       "      <td>795c9e2d-9f25-4996-ae7b-97da70419e33</td>\n",
       "      <td>b4687245-5202-4e2c-ba8a-0e8dfb64db9a</td>\n",
       "      <td>COMPLETE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GF_04D094R4</td>\n",
       "      <td>18b108ed-fad6-44b7-a410-9ed2c1fc821f</td>\n",
       "      <td>7e77c0f4-803c-4508-85d9-efd20c3f00b4</td>\n",
       "      <td>COMPLETE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       file_id                            latest_did  \\\n",
       "0  GF_034F7RMB  795c9e2d-9f25-4996-ae7b-97da70419e33   \n",
       "1  GF_04D094R4  18b108ed-fad6-44b7-a410-9ed2c1fc821f   \n",
       "\n",
       "                                 run_id    status  \n",
       "0  b4687245-5202-4e2c-ba8a-0e8dfb64db9a  COMPLETE  \n",
       "1  7e77c0f4-803c-4508-85d9-efd20c3f00b4  COMPLETE  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for index, row in merged_df.iterrows():\n",
    "    status = wesClient.getTaskStatus(row['run_id'])\n",
    "    merged_df.at[index, \"status\"] = status\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import os\n",
    "def download(url, file_path):\n",
    "    '''Download a file from a URL to a local file path'''\n",
    "    with open(os.path.expanduser(file_path), \"wb\") as file:\n",
    "        response = requests.get(url)\n",
    "        file.write(response.content)\n",
    "        \n",
    "def getStats(filePath, statsList):\n",
    "    ''' Extract values from a SAMTools Stats results file'''\n",
    "    ### \n",
    "    \n",
    "    retDict = {}\n",
    "    f = open(filePath, \"r\")\n",
    "    for x in f:\n",
    "        if x.startswith('SN'):   \n",
    "            parts = x.split('\\t')\n",
    "            statName = parts[1].split(':')[0]\n",
    "            if statName in statsList:\n",
    "                retDict[statName] = parts[2].rstrip()\n",
    "    return retDict\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Invalid access token in /Users/forei/.keys/anvil_credentials.json\n",
      "Searching the GA4GH registry for org.ga4gh:drs services\n",
      "drs://cgc-ga4gh-api.sbgenomics.com/61551f68e9a3f642ca56b41e\n",
      "drs://cgc-ga4gh-api.sbgenomics.com/61552c79e9a3f642ca56b46b\n"
     ]
    }
   ],
   "source": [
    "from fasp.loc import DRSMetaResolver\n",
    "from fasp.loc import sbcgcDRSClient\n",
    "mr = DRSMetaResolver()\n",
    "statsList = []\n",
    "for index, row in merged_df.iterrows():\n",
    "    log = wesClient.GetRunLog(row['run_id'])\n",
    "    resultsDRSID = log['outputs']['statistics']['path']\n",
    "    print(resultsDRSID)\n",
    "    url = mr.getAccessURL2(resultsDRSID,'s3')\n",
    "    fileName = log['outputs']['statistics']['name']\n",
    "    download(url, fileName)\n",
    "    statsRequired = ['insert size average','insert size standard deviation']\n",
    "    stats = getStats(fileName, statsRequired)\n",
    "    stats['file_id'] = row['file_id']\n",
    "    statsList.append(stats)\n",
    "    os.remove(fileName)\n",
    "stats_df =  pd.DataFrame(statsList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>insert size average</th>\n",
       "      <th>insert size standard deviation</th>\n",
       "      <th>file_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>236.9</td>\n",
       "      <td>82.0</td>\n",
       "      <td>GF_034F7RMB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>301.4</td>\n",
       "      <td>130.9</td>\n",
       "      <td>GF_04D094R4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  insert size average insert size standard deviation      file_id\n",
       "0               236.9                           82.0  GF_034F7RMB\n",
       "1               301.4                          130.9  GF_04D094R4"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
