{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data that DRS might provide about a file\n",
    "\n",
    "Use case: Provide data* about a file identified by a given DRS id.\n",
    "\n",
    "The realistic scenarios for this use case should be identified. Notably, when does this scenario occur, compared with the scenario where file ids are obtained from some other data driven query which reflects user interest?\n",
    "\n",
    "Key points are that\n",
    "\n",
    "* Search and other APIs already provide data relevant to answering that question\n",
    "* Would expect to use the same schemas that those APIs use \n",
    "* Reusing those schema provides rich functionality\n",
    "\n",
    "i.e. rather than DRS determining what metadata should be provided about a file we should tap into the existing schema that are available.\n",
    "\n",
    "*Note: there is a conscious intent here to avoid the term metadata. For one thing we need to be more specific about what would be useful."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Retrieving the query_\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n"
     ]
    }
   ],
   "source": [
    "from fasp.search import DiscoverySearchClient\n",
    "from fasp.loc import crdcDRSClient\n",
    "\n",
    "import json\n",
    "# TCGA Query - CRDC\n",
    "searchClient = DiscoverySearchClient('https://ga4gh-search-adapter-presto-public.prod.dnastack.com/')\n",
    "\n",
    "query = \"\"\"\n",
    "    SELECT  file_id drs_id\n",
    "    FROM search_cloud.cshcodeathon.gdc_rel24_filedata_active \n",
    "    where data_format = 'BAM'\n",
    "    and project_disease_type = 'Breast Invasive Carcinoma'\n",
    "    limit 1\"\"\"\n",
    "res = searchClient.runQuery(query)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### An example of basic data about a file.\n",
    "\n",
    "The data returned is the same data used in the original search. e.g. the data_format 'BAM' returned below is exactly that which is in the query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "   \"id\": \"030e5e74-6461-4f05-a399-de8e470bc056\",\n",
      "   \"data_format\": \"BAM\",\n",
      "   \"access\": \"controlled\",\n",
      "   \"file_name\": \"46db33a7f2003837e88d0a81b8ebec2c_gdc_realn.bam\",\n",
      "   \"data_type\": \"Aligned Reads\",\n",
      "   \"data_category\": \"Sequencing Reads\",\n",
      "   \"type\": \"aligned_reads\",\n",
      "   \"experimental_strategy\": \"WXS\",\n",
      "   \"platform\": \"Illumina\",\n",
      "   \"created_datetime\": \"2016-05-03T00:35:52.946132-05:00\",\n",
      "   \"file_size\": 23894757370\n",
      "}\n",
      "________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "drsClient = crdcDRSClient()\n",
    "for id in res:\n",
    "    print(json.dumps(drsClient.getFileData(id[0]), indent=3))\n",
    "    print('_'*80)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The data can be retrieved through Search API itself\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Retrieving the query_\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "[\n",
      "   {\n",
      "      \"dbname\": \"active\",\n",
      "      \"file_gdc_id\": \"030e5e74-6461-4f05-a399-de8e470bc056\",\n",
      "      \"access\": \"controlled\",\n",
      "      \"acl\": \"phs000178\",\n",
      "      \"analysis_input_file_gdc_ids\": null,\n",
      "      \"analysis_workflow_link\": \"https://github.com/NCI-GDC/cocleaning-cwl\",\n",
      "      \"analysis_workflow_type\": \"BWA with Mark Duplicates and Cocleaning\",\n",
      "      \"archive_gdc_id\": null,\n",
      "      \"archive_revision\": null,\n",
      "      \"archive_state\": null,\n",
      "      \"archive_submitter_id\": null,\n",
      "      \"associated_entities__case_gdc_id\": \"1b703058-e596-45bc-80fe-8b98d545c2e2\",\n",
      "      \"associated_entities__entity_gdc_id\": \"7b0b60c7-5fa0-440e-937f-8d82119330d6\",\n",
      "      \"associated_entities__entity_submitter_id\": \"TCGA-AR-A2LK-01A-11D-A17W-09\",\n",
      "      \"associated_entities__entity_type\": \"aliquot\",\n",
      "      \"case_gdc_id\": \"1b703058-e596-45bc-80fe-8b98d545c2e2\",\n",
      "      \"project_dbgap_accession_number\": null,\n",
      "      \"project_disease_type\": \"Breast Invasive Carcinoma\",\n",
      "      \"project_name\": \"Breast Invasive Carcinoma\",\n",
      "      \"program_dbgap_accession_number\": \"phs000178\",\n",
      "      \"program_name\": \"TCGA\",\n",
      "      \"project_short_name\": \"TCGA-BRCA\",\n",
      "      \"created_datetime\": \"2016-05-03T00:35:52.946132-05:00\",\n",
      "      \"data_category\": \"Sequencing Reads\",\n",
      "      \"data_format\": \"BAM\",\n",
      "      \"data_type\": \"Aligned Reads\",\n",
      "      \"downstream_analyses__output_file_gdc_ids\": \"b68a22c3-110e-492f-866d-062518d296e5;d1ef3b6c-cce1-408c-85e6-9ca60c93ae28;c29294b1-ce40-4eaa-88e0-ed10c1848834;f0382a48-0f82-494a-9224-7b41103d519b\",\n",
      "      \"downstream_analyses__workflow_link\": \"https://github.com/NCI-GDC/somaticsniper-cwl;https://github.com/NCI-GDC/varscan-cwl;https://github.com/NCI-GDC/mutect-cwl;https://github.com/NCI-GDC/muse-cwl\",\n",
      "      \"downstream_analyses__workflow_type\": \"SomaticSniper;VarScan2;MuTect2;MuSE\",\n",
      "      \"experimental_strategy\": \"WXS\",\n",
      "      \"file_name\": \"46db33a7f2003837e88d0a81b8ebec2c_gdc_realn.bam\",\n",
      "      \"file_size\": 23894757370,\n",
      "      \"file_id\": \"030e5e74-6461-4f05-a399-de8e470bc056\",\n",
      "      \"index_file_gdc_id\": \"55192166-9f11-4271-97d9-ec135a9635e8\",\n",
      "      \"index_file_name\": \"46db33a7f2003837e88d0a81b8ebec2c_gdc_realn.bai\",\n",
      "      \"index_file_size\": 8560432,\n",
      "      \"md5sum\": \"f0cf564932ed0418c08b828a556d6478\",\n",
      "      \"platform\": \"Illumina\",\n",
      "      \"file_state\": \"released\",\n",
      "      \"file_submitter_id\": \"13794022-fe14-45d2-a0cc-21622ef537bf\",\n",
      "      \"file_type\": \"aligned_reads\",\n",
      "      \"updated_datetime\": \"2018-11-15T17:21:45.437138-06:00\"\n",
      "   }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "    SELECT * \n",
    "    FROM search_cloud.cshcodeathon.gdc_rel24_filedata_active \n",
    "    where file_id = '030e5e74-6461-4f05-a399-de8e470bc056' \"\"\"\n",
    "res2 = searchClient.runQuery(query, returnType='json')\n",
    "print(json.dumps(res2,indent=3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting linked objects - Case, Indexes, Downstream Analyses, ...\n",
    "\n",
    "More detailed information is also available.\n",
    "\n",
    "Making use of existing schema means that a DRS id can be a jumping off point into the graph. A user could fulfill any use case that the graph allows. Tapping into the schemas provided by Search provides rich functionality with little additional effort."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "   \"id\": \"030e5e74-6461-4f05-a399-de8e470bc056\",\n",
      "   \"data_format\": \"BAM\",\n",
      "   \"access\": \"controlled\",\n",
      "   \"cases\": [\n",
      "      {\n",
      "         \"case_id\": \"1b703058-e596-45bc-80fe-8b98d545c2e2\"\n",
      "      }\n",
      "   ],\n",
      "   \"associated_entities\": [\n",
      "      {\n",
      "         \"entity_type\": \"aliquot\",\n",
      "         \"entity_id\": \"7b0b60c7-5fa0-440e-937f-8d82119330d6\"\n",
      "      }\n",
      "   ],\n",
      "   \"file_name\": \"46db33a7f2003837e88d0a81b8ebec2c_gdc_realn.bam\",\n",
      "   \"data_category\": \"Sequencing Reads\",\n",
      "   \"downstream_analyses\": [\n",
      "      {\n",
      "         \"analysis_id\": \"e05199a1-d6ad-4d31-a616-dcc6fb057216\"\n",
      "      },\n",
      "      {\n",
      "         \"analysis_id\": \"d0cc7c6d-675c-401e-b751-60c68b4436e5\"\n",
      "      },\n",
      "      {\n",
      "         \"analysis_id\": \"90e83350-f2cf-49bc-910e-5331a2cea795\"\n",
      "      },\n",
      "      {\n",
      "         \"analysis_id\": \"702c782f-ce48-41c4-b7f2-c97ba26bc8a0\"\n",
      "      }\n",
      "   ],\n",
      "   \"type\": \"aligned_reads\",\n",
      "   \"analysis\": {\n",
      "      \"analysis_id\": \"35eb6a6c-76d7-4568-ae0a-45734676c43e\"\n",
      "   },\n",
      "   \"platform\": \"Illumina\",\n",
      "   \"created_datetime\": \"2016-05-03T00:35:52.946132-05:00\",\n",
      "   \"file_size\": 23894757370,\n",
      "   \"index_files\": [\n",
      "      {\n",
      "         \"file_id\": \"55192166-9f11-4271-97d9-ec135a9635e8\"\n",
      "      }\n",
      "   ],\n",
      "   \"data_type\": \"Aligned Reads\",\n",
      "   \"experimental_strategy\": \"WXS\"\n",
      "}\n",
      "________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "for id in res:\n",
    "    fileData = drsClient.getFileData(id[0], expanded=True)\n",
    "    print(json.dumps(fileData, indent=3))\n",
    "    print('_'*80)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting attributes of related objects\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "   \"id\": \"030e5e74-6461-4f05-a399-de8e470bc056\",\n",
      "   \"data_format\": \"BAM\",\n",
      "   \"access\": \"controlled\",\n",
      "   \"cases\": [\n",
      "      {\n",
      "         \"case_id\": \"1b703058-e596-45bc-80fe-8b98d545c2e2\",\n",
      "         \"project\": {\n",
      "            \"disease_type\": \"Breast Invasive Carcinoma\"\n",
      "         },\n",
      "         \"diagnoses\": [\n",
      "            {\n",
      "               \"days_to_recurrence\": null,\n",
      "               \"morphology\": \"8520/3\",\n",
      "               \"tumor_stage\": \"stage iii\",\n",
      "               \"created_datetime\": null,\n",
      "               \"tissue_or_organ_of_origin\": \"Breast, NOS\",\n",
      "               \"primary_diagnosis\": \"Lobular carcinoma, NOS\",\n",
      "               \"age_at_diagnosis\": 22800,\n",
      "               \"classification_of_tumor\": \"not reported\",\n",
      "               \"prior_malignancy\": \"no\",\n",
      "               \"site_of_resection_or_biopsy\": \"Breast, NOS\",\n",
      "               \"days_to_last_known_disease_status\": null,\n",
      "               \"tumor_grade\": \"not reported\",\n",
      "               \"progression_or_recurrence\": \"not reported\"\n",
      "            }\n",
      "         ],\n",
      "         \"demographic\": {\n",
      "            \"race\": \"white\",\n",
      "            \"updated_datetime\": \"2019-07-31T21:33:13.355468-05:00\",\n",
      "            \"submitter_id\": \"TCGA-AR-A2LK_demographic\",\n",
      "            \"state\": \"released\",\n",
      "            \"year_of_death\": null,\n",
      "            \"year_of_birth\": 1945\n",
      "         }\n",
      "      }\n",
      "   ],\n",
      "   \"associated_entities\": [\n",
      "      {\n",
      "         \"entity_type\": \"aliquot\",\n",
      "         \"entity_id\": \"7b0b60c7-5fa0-440e-937f-8d82119330d6\"\n",
      "      }\n",
      "   ],\n",
      "   \"file_name\": \"46db33a7f2003837e88d0a81b8ebec2c_gdc_realn.bam\",\n",
      "   \"data_category\": \"Sequencing Reads\",\n",
      "   \"downstream_analyses\": [\n",
      "      {\n",
      "         \"analysis_id\": \"e05199a1-d6ad-4d31-a616-dcc6fb057216\"\n",
      "      },\n",
      "      {\n",
      "         \"analysis_id\": \"d0cc7c6d-675c-401e-b751-60c68b4436e5\"\n",
      "      },\n",
      "      {\n",
      "         \"analysis_id\": \"90e83350-f2cf-49bc-910e-5331a2cea795\"\n",
      "      },\n",
      "      {\n",
      "         \"analysis_id\": \"702c782f-ce48-41c4-b7f2-c97ba26bc8a0\"\n",
      "      }\n",
      "   ],\n",
      "   \"type\": \"aligned_reads\",\n",
      "   \"analysis\": {\n",
      "      \"analysis_id\": \"35eb6a6c-76d7-4568-ae0a-45734676c43e\"\n",
      "   },\n",
      "   \"platform\": \"Illumina\",\n",
      "   \"created_datetime\": \"2016-05-03T00:35:52.946132-05:00\",\n",
      "   \"file_size\": 23894757370,\n",
      "   \"index_files\": [\n",
      "      {\n",
      "         \"file_id\": \"55192166-9f11-4271-97d9-ec135a9635e8\"\n",
      "      }\n",
      "   ],\n",
      "   \"data_type\": \"Aligned Reads\",\n",
      "   \"experimental_strategy\": \"WXS\"\n",
      "}\n",
      "________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "for id in res:\n",
    "    fileData = drsClient.getFileData(id[0], linked=True)\n",
    "    print(json.dumps(fileData, indent=3))\n",
    "    print('_'*80)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using the related data\n",
    "\n",
    "In this example the file id for the related Index file can be retrieved through DRS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'access_methods': [{'access_id': 'gs',\n",
       "   'access_url': {'url': 'gs://gdc-tcga-phs000178-controlled/18fb79bb-7259-41da-bd76-9dd9f8f84bfc/TCGA-A1-A0SD-01A-11D-A10Y-09_IlluminaGA-DNASeq_exome_gdc_realn.bai'},\n",
       "   'region': '',\n",
       "   'type': 'gs'},\n",
       "  {'access_id': 's3',\n",
       "   'access_url': {'url': 's3://tcga-2-controlled/18fb79bb-7259-41da-bd76-9dd9f8f84bfc/TCGA-A1-A0SD-01A-11D-A10Y-09_IlluminaGA-DNASeq_exome_gdc_realn.bai'},\n",
       "   'region': '',\n",
       "   'type': 's3'}],\n",
       " 'aliases': [],\n",
       " 'checksums': [{'checksum': 'dd3b9e4fa8a85cc18c413e8b5b58e252',\n",
       "   'type': 'md5'}],\n",
       " 'contents': [],\n",
       " 'created_time': '2018-08-08T17:11:24.583780',\n",
       " 'description': None,\n",
       " 'form': 'object',\n",
       " 'id': '18fb79bb-7259-41da-bd76-9dd9f8f84bfc',\n",
       " 'mime_type': 'application/json',\n",
       " 'name': None,\n",
       " 'self_uri': 'drs://nci-crdc.datacommons.io/18fb79bb-7259-41da-bd76-9dd9f8f84bfc',\n",
       " 'size': 6700896,\n",
       " 'updated_time': '2018-08-08T17:11:24.583791',\n",
       " 'version': '3c6bf46f'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drsClient.getObject(\"18fb79bb-7259-41da-bd76-9dd9f8f84bfc\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fileData = drsClient.getFileData(\"18fb79bb-7259-41da-bd76-9dd9f8f84bfc\")\n",
    "#print(json.dumps(fileData, indent=3))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using table info to find out about Age at Diagnosis\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'search_cloud.cshcodeathon.tcga_clinical_gdc_current',\n",
       " 'description': 'Automatically generated schema',\n",
       " 'data_model': {'$id': 'https://ga4gh-search-adapter-presto-public.prod.dnastack.com/table/search_cloud.cshcodeathon.tcga_clinical_gdc_current/info',\n",
       "  'description': 'Automatically generated schema',\n",
       "  '$schema': 'http://json-schema.org/draft-07/schema#',\n",
       "  'properties': {'submitter_id': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The submitter_id of a case entity corresponds to the submitted_subject_id of the study participant in dbGaP records for the project.'},\n",
       "   'case_id': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'GDC unique identifier for this case (corresponds to the case_barcode) -- this can be used to access more information from the GDC data portal in the following way: https://portal.gdc.cancer.gov/files/c21b332c-06c6-4403-9032-f91c8f407ba9.'},\n",
       "   'diag__treat__count': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Total child record count (located in cases table).'},\n",
       "   'primary_site': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"The text term used to describe the primary site of disease, as categorized by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O). This categorization groups cases into general categories. Reference tissue_or_organ_of_origin on the diagnosis node for more specific primary sites of disease.\"},\n",
       "   'disease_type': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"The text term used to describe the type of malignant disease, as categorized by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).\"},\n",
       "   'proj__name': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Display name for the project to which a case belongs.'},\n",
       "   'proj__project_id': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'GDC-assigned identifier for the project to which a case belongs.'},\n",
       "   'demo__demographic_id': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"GDC-assigned identifier for a case's demographic data.\"},\n",
       "   'demo__gender': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Text designations that identify gender. Gender is described as the assemblage of properties that distinguish people on the basis of their societal roles. [Explanatory Comment 1: Identification of gender is based upon self-report and may come from a form, questionnaire, interview, etc.]'},\n",
       "   'demo__race': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'An arbitrary classification of a taxonomic group that is a division of a species. It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.'},\n",
       "   'demo__ethnicity': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"An individual's self-described social and cultural grouping, specifically whether an individual describes themselves as Hispanic or Latino. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.\"},\n",
       "   'demo__vital_status': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The survival state of the person registered on the protocol.'},\n",
       "   'demo__days_to_birth': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': \"Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days.\"},\n",
       "   'demo__year_of_birth': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Numeric value to represent the calendar year in which an individual was born.'},\n",
       "   'demo__age_at_index': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': \"The patient's age (in years) on the reference or anchor date date used during date obfuscation.\"},\n",
       "   'demo__year_of_death': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Numeric value to represent the year of the death of an individual.'},\n",
       "   'demo__days_to_death': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': \"Number of days between the date used for index and the date from a person's date of death represented as a calculated number of days.\"},\n",
       "   'demo__state': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The current state of the object.'},\n",
       "   'demo__created_datetime': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]'},\n",
       "   'demo__updated_datetime': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]'},\n",
       "   'diag__diagnosis_id': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"GDC-assigned identifier for a case's diagnosis.\"},\n",
       "   'diag__ajcc_clinical_n': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Extent of the regional lymph node involvement for the cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.'},\n",
       "   'diag__masaoka_stage': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The text term used to describe the Masaoka staging system, a classification that defines prognostic indicators for thymic malignancies and predicts tumor recurrence.'},\n",
       "   'diag__ajcc_clinical_m': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Extent of the distant metastasis for the cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.'},\n",
       "   'diag__primary_diagnosis': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"Text term used to describe the patient's histologic diagnosis, as described by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).\"},\n",
       "   'diag__primary_gleason_grade': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"The text term used to describe the primary Gleason score, which describes the pattern of cells making up the largest area of the tumor. The primary and secondary Gleason pattern grades are combined to determine the patient's Gleason grade group, which is used to determine the aggresiveness of prostate cancer. Note that this grade describes the entire prostatectomy specimen and is not specific to the sample used for sequencing.\"},\n",
       "   'diag__year_of_diagnosis': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': \"Numeric value to represent the year of an individual's initial pathologic diagnosis of cancer.\"},\n",
       "   'diag__figo_stage': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The extent of a cervical or endometrial cancer within the body, especially whether the disease has spread from the original site to other parts of the body, as described by the International Federation of Gynecology and Obstetrics (FIGO) stages.'},\n",
       "   'diag__progression_or_recurrence': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Yes/No/Unknown indicator to identify whether a patient has had a new tumor event after initial treatment.'},\n",
       "   'diag__ajcc_pathologic_m': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Code to represent the defined absence or presence of distant spread or metastases (M) to locations via vascular channels or lymphatics beyond the regional lymph nodes, using criteria established by the American Joint Committee on Cancer (AJCC).'},\n",
       "   'diag__site_of_resection_or_biopsy': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"The text term used to describe the anatomic site of origin, of the patient's malignant disease, as described by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).\"},\n",
       "   'diag__ajcc_staging_system_edition': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The text term used to describe the version or edition of the American Joint Committee on Cancer Staging Handbooks, a publication by the group formed for the purpose of developing a system of staging for cancer that is acceptable to the American medical profession and is compatible with other accepted classifications.'},\n",
       "   'diag__icd_10_code': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Alphanumeric value used to describe the  disease code from the tenth version of the International Classification of Disease (ICD-10).'},\n",
       "   'diag__age_at_diagnosis': {'format': 'bigint',\n",
       "    'type': 'int',\n",
       "    '$comment': 'Age at the time of diagnosis expressed in number of days since birth.'},\n",
       "   'diag__ajcc_clinical_t': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Extent of the primary cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.'},\n",
       "   'diag__days_to_last_follow_up': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': 'Time interval from the date of last follow up to the date of initial pathologic diagnosis, represented as a calculated number of days.'},\n",
       "   'diag__ajcc_pathologic_stage': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The extent of a cancer, especially whether the disease has spread from the original site to other parts of the body based on AJCC staging criteria.'},\n",
       "   'diag__tumor_grade': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Numeric value to express the degree of abnormality of cancer cells, a measure of differentiation and aggressiveness.'},\n",
       "   'diag__last_known_disease_status': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"Text term that describes the last known state or condition of an individual's neoplasm.\"},\n",
       "   'diag__ann_arbor_extranodal_involvement': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Indicator that identifies whether a patient with malignant lymphoma has lymphomatous involvement of an extranodal site.'},\n",
       "   'diag__ajcc_clinical_stage': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Stage group determined from clinical information on the tumor (T), regional node (N) and metastases (M) and by grouping cases with similar prognosis for cancer.'},\n",
       "   'diag__secondary_gleason_grade': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"The text term used to describe the secondary Gleason score, which describes the pattern of cells making up the second largest area of the tumor. The primary and secondary Gleason pattern grades are combined to determine the patient's Gleason grade group, which is used to determine the aggresiveness of prostate cancer. Note that this grade describes the entire prostatectomy specimen and is not specific to the sample used for sequencing.\"},\n",
       "   'diag__synchronous_malignancy': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'A yes/no/unknown indicator used to describe whether the patient had an additional malignant diagnosis at the same time the tumor used for sequencing was diagnosed. If both tumors were sequenced, both tumors would have synchronous malignancies.'},\n",
       "   'diag__morphology': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The third edition of the International Classification of Diseases for Oncology, published in 2000 used principally in tumor and cancer registries for coding the site (topography) and the histology (morphology) of neoplasms. The study of the structure of the cells and their arrangement to constitute tissues and, finally, the association among these to form organs. In pathology, the microscopic process of identifying normal and abnormal morphologic characteristics in tissues, by employing various cytochemical and immunocytochemical stains. A system of numbered categories for representation of data.'},\n",
       "   'diag__ajcc_pathologic_t': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Code of pathological T (primary tumor) to define the size or contiguous extension of the primary tumor (T), using staging criteria from the American Joint Committee on Cancer (AJCC).'},\n",
       "   'diag__igcccg_stage': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The text term used to describe the International Germ Cell Cancer Collaborative Group (IGCCCG), a grouping used to further classify metastatic testicular tumors.'},\n",
       "   'diag__classification_of_tumor': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Text that describes the kind of disease present in the tumor specimen as related to a specific timepoint.'},\n",
       "   'diag__ann_arbor_clinical_stage': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The text term used to describe the clinical classification of lymphoma, as defined by the Ann Arbor Lymphoma Staging System.'},\n",
       "   'diag__ann_arbor_b_symptoms': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"Text term to signify whether lymphoma B-symptoms are present as noted in the patient's medical record.\"},\n",
       "   'diag__tumor_stage': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The extent of a cancer in the body. Staging is usually based on the size of the tumor, whether lymph nodes contain cancer, and whether the cancer has spread from the original site to other parts of the body. The accepted values for tumor_stage depend on the tumor site, type, and accepted staging system. These items should accompany the tumor_stage value as associated metadata.'},\n",
       "   'diag__prior_treatment': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'A yes/no/unknown/not applicable indicator related to the administration of therapeutic agents received before the body specimen was collected.'},\n",
       "   'diag__ajcc_pathologic_n': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"The codes that represent the stage of cancer based on the nodes present (N stage) according to criteria based on multiple editions of the AJCC's Cancer Staging Manual.\"},\n",
       "   'diag__tissue_or_organ_of_origin': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"The text term used to describe the anatomic site of origin, of the patient's malignant disease, as described by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).\"},\n",
       "   'diag__prior_malignancy': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"The yes/no/unknown indicator used to describe the patient's history of prior cancer diagnosis.\"},\n",
       "   'diag__state': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The current state of the object.'},\n",
       "   'diag__created_datetime': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]'},\n",
       "   'diag__updated_datetime': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]'},\n",
       "   'exp__exposure_id': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': \"GDC-assigned identifier for a case's exposure data.\"},\n",
       "   'exp__height': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': 'The height of the patient in centimeters.'},\n",
       "   'exp__weight': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': 'The weight of the patient measured in kilograms.'},\n",
       "   'exp__bmi': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': \"A calculated numerical quantity that represents an individual's weight to height ratio.\"},\n",
       "   'exp__years_smoked': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': 'Numeric value (or unknown) to represent the number of years a person has been smoking.'},\n",
       "   'exp__pack_years_smoked': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': 'Numeric computed value to represent lifetime tobacco exposure defined as number of cigarettes smoked per day x number of years smoked divided by 20.'},\n",
       "   'exp__cigarettes_per_day': {'format': 'double',\n",
       "    'type': 'number',\n",
       "    '$comment': 'The average number of cigarettes smoked per day.'},\n",
       "   'exp__alcohol_history': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'A response to a question that asks whether the participant has consumed at least 12 drinks of any kind of alcoholic beverage in their lifetime.'},\n",
       "   'exp__state': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The current state of the object.'},\n",
       "   'exp__created_datetime': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]'},\n",
       "   'exp__updated_datetime': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]'},\n",
       "   'state': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'The current state of the object.'},\n",
       "   'updated_datetime': {'format': 'varchar',\n",
       "    'type': 'string',\n",
       "    '$comment': 'Timestamp of last update in GDC; formatted as a combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm].'}}}}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "searchClient.listTableInfo('search_cloud.cshcodeathon.tcga_clinical_gdc_current')\n",
    "#searchClient.listTables('search_cloud')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
