{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "82f5654d",
   "metadata": {},
   "source": [
    "### Retrieve example from Gen3 Metadata Service\n",
    "\n",
    "The following is an example response from the Gen3 Metadata service for a specific file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4af28020",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "service_url = 'https://gen3.biodatacatalyst.nhlbi.nih.gov/mds/metadata/'\n",
    "example_id = 'dg.4503/00000027-a1c4-4587-a6c9-d35f5818a724'\n",
    "call_url = f'{service_url}{example_id}'\n",
    "\n",
    "response = requests.get(call_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a776547c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sex': 'female',\n",
       " 'body_site': 'peripheral blood',\n",
       " 'repository': 'GeneSTAR',\n",
       " 'sample_use': ['Seq_DNA_SNP_CNV'],\n",
       " 'analyte_type': 'DNA',\n",
       " 'biosample_id': 'SAMN06062404',\n",
       " 'consent_code': 2,\n",
       " 'dbgap_status': 'Loaded',\n",
       " 'sra_sample_id': 'SAMN06062404',\n",
       " 'dbgap_sample_id': 2170791,\n",
       " 'study_accession': 'phs001218.v2.p1',\n",
       " 'dbgap_subject_id': 1797782,\n",
       " 'sra_data_details': {'runs': '2',\n",
       "  'bases': '260957674200',\n",
       "  'center': 'Macrogen',\n",
       "  'status': 'public',\n",
       "  'size_Gb': '49',\n",
       "  'platform': 'ILLUMINA',\n",
       "  'experiments': '2',\n",
       "  'experiment_type': 'WGS'},\n",
       " 'study_subject_id': 'phs001218.v2_GS08900974',\n",
       " 'consent_short_name': 'DS-CVD-IRB-NPU-MDS',\n",
       " 'study_with_consent': 'phs001218.c2',\n",
       " 'submitted_sample_id': 'NWD403016',\n",
       " 'submitted_subject_id': 'GS08900974',\n",
       " 'study_accession_with_consent': 'phs001218.v2.p1.c2'}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.json()['dbgap']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "452f6f23",
   "metadata": {},
   "source": [
    "### Querying SRA for the same data via Data Connect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "3b4fcfdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from fasp.search import DataConnectClient\n",
    "searchClient = DataConnectClient('http://localhost:8089',\n",
    "                                passport='~/Downloads/task-specific-token-20.txt')\n",
    "#searchClient = DataConnectClient('https://collection-service.staging.dnastack.com/data-connect')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "74752371",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablenih_sra_datastore.sra.metadata_\n",
      "{\n",
      "   \"name\": \"nih_sra_datastore.sra.metadata\",\n",
      "   \"description\": \"Metadata Table (sra.metadata) contains information about the run and biological samples. Metadata Table (sra.metadata) contains information about the run and biological samples. The biological sample data is stored in two different columns.  See Record (array) column that you need to use the command UNNEST to query. See https://www.ncbi.nlm.nih.gov/sra/docs/sra-cloud-based-metadata-table/\",\n",
      "   \"data_model\": {\n",
      "      \"$id\": \"\",\n",
      "      \"description\": \"Metadata Table (sra.metadata) contains information about the run and biological samples. Metadata Table (sra.metadata) contains information about the run and biological samples. The biological sample data is stored in two different columns.  See Record (array) column that you need to use the command UNNEST to query. See https://www.ncbi.nlm.nih.gov/sra/docs/sra-cloud-based-metadata-table/\",\n",
      "      \"$schema\": \"http://json-schema.org/draft-07/schema\",\n",
      "      \"properties\": {\n",
      "         \"acc\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"SRA Run accession in the form of SRR######## (ERR or DRR for INSDC partners)\"\n",
      "         },\n",
      "         \"assay_type\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Type of library (i.e. AMPLICON, RNA-Seq, WGS, etc)\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"RNA-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"WGS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"EST\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"SELEX\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"other\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"CTS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"CLONE\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Hi-C\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"POOLCLONE\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"GBS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MBD-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"miRNA-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MeDIP-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RIP-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ChIP\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MRE-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"FAIRE-seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"FL-cDNA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"DNase-Hypersensitivity\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ncRNA-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"WCS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MNase-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"AMPLICON\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"WXS\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ChIP-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Synthetic-Long-Read\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"WGA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"FINISHING\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Tethered Chromatin Conformation Capture\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Targeted-Capture\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ATAC-seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RAD-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ChIA-PET\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"OTHER\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Bisulfite-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Tn-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"CLONEEND\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"VALIDATION\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"center_name\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the sequencing center\"\n",
      "         },\n",
      "         \"consent\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Type of consent need to access the data (i.e. public is available to all, others are for dbGaP)\"\n",
      "         },\n",
      "         \"experiment\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The accession in the form of SRX######## (ERX or DRX for INSDC partners)\"\n",
      "         },\n",
      "         \"sample_name\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the sample\"\n",
      "         },\n",
      "         \"instrument\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the sequencing instrument model\"\n",
      "         },\n",
      "         \"librarylayout\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Whether the data is SINGLE or PAIRED\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"PAIRED\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"SINGLE\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"libraryselection\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Library selection methodology (i.e. PCR, RANDOM, etc)\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"MF\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ChIP-Seq\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"unspecified\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"size fractionation\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"ChIP\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"CAGE\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MSLL\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"other\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"DNase\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Hybrid Selection\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"OTHER\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RT-PCR\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RANDOM\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"padlock probes capture method\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"PCR\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Oligo-dT\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Restriction Digest\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Reduced Representation\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"Inverse rRNA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"5-methylcytidine antibody\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MNase\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MBD2 protein methyl-CpG binding domain\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"HMPR\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"cDNA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"PolyA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RANDOM PCR\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"MDA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"repeat fractionation\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"RACE\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"librarysource\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Source of the biological data (i.e. GENOMIC, METAGENOMIC, etc)\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"title\": \"SYNTHETIC\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"TRANSCRIPTOMIC\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"OTHER\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"VIRAL RNA\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"METATRANSCRIPTOMIC\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"METAGENOMIC\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"TRANSCRIPTOMIC SINGLE CELL\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"GENOMIC\"\n",
      "               },\n",
      "               {\n",
      "                  \"title\": \"GENOMIC SINGLE CELL\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"platform\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the sequencing platform (i.e. ILLUMINA)\"\n",
      "         },\n",
      "         \"sample_acc\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"SRA Sample accession in the form of SRS######## (ERS or DRS for INSDC partners)\"\n",
      "         },\n",
      "         \"biosample\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"BioSample accession in the form of SAMN######## (SAMEA##### or SAMD##### for INSDC partners)\"\n",
      "         },\n",
      "         \"organism\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Scientific name of the organism that was sequenced (as found in the NCBI Taxonomy Browser)\"\n",
      "         },\n",
      "         \"sra_study\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"SRA Study accession in the form of SRP######## (ERP or DRP for INSDC partners)\"\n",
      "         },\n",
      "         \"releasedate\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The date on which the data was released\"\n",
      "         },\n",
      "         \"bioproject\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"BioProject accession in the form of PRJNA######## (PRJEB####### or PRJDB###### for INSDC partners)\"\n",
      "         },\n",
      "         \"mbytes\": {\n",
      "            \"type\": \"integer\",\n",
      "            \"description\": \"Number of mega bytes of data in the SRA Run\"\n",
      "         },\n",
      "         \"loaddate\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The date when the data was loaded into SRA\"\n",
      "         },\n",
      "         \"avgspotlen\": {\n",
      "            \"type\": \"integer\",\n",
      "            \"description\": \"Calculated average read length\"\n",
      "         },\n",
      "         \"mbases\": {\n",
      "            \"type\": \"integer\",\n",
      "            \"description\": \"Number of mega bases in the SRA Runs\"\n",
      "         },\n",
      "         \"insertsize\": {\n",
      "            \"type\": \"integer\",\n",
      "            \"description\": \"Submitter provided insert size\"\n",
      "         },\n",
      "         \"library_name\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The name of the library\"\n",
      "         },\n",
      "         \"biosamplemodel_sam\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The BioSample package/model that was picked\"\n",
      "         },\n",
      "         \"collection_date_sam\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"The collection date of the sample\"\n",
      "         },\n",
      "         \"geo_loc_name_country_calc\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the country where the sample was collected\"\n",
      "         },\n",
      "         \"geo_loc_name_country_continent_calc\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Name of the continent where the sample was collected\"\n",
      "         },\n",
      "         \"geo_loc_name_sam\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Full location of collection\"\n",
      "         },\n",
      "         \"ena_first_public_run\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Date when INSDC partner record was public\"\n",
      "         },\n",
      "         \"ena_last_update_run\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Date when INSDC partner record was updated\"\n",
      "         },\n",
      "         \"sample_name_sam\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"INSDC sample name\"\n",
      "         },\n",
      "         \"datastore_filetype\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Type of files available to download from SRA\"\n",
      "         },\n",
      "         \"datastore_provider\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Locations of where the files are available to download from\"\n",
      "         },\n",
      "         \"datastore_region\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Regions of where the data is located\"\n",
      "         },\n",
      "         \"attributes\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Full list of sample attributes in a nested(array) structure\"\n",
      "         },\n",
      "         \"attributes.k\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Attribute's name\"\n",
      "         },\n",
      "         \"attributes.v\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"Attribute's value\"\n",
      "         },\n",
      "         \"jattr\": {\n",
      "            \"type\": \"string\",\n",
      "            \"description\": \"JSON based string of the sample attributes\"\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x122125f40>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "searchClient.list_table_info('nih_sra_datastore.sra.metadata', verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42fe6c34",
   "metadata": {},
   "source": [
    "### Query the SRA table on the biosample id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1a93df24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"select * from nih_sra_datastore.sra.metadata where biosample='SAMN06062404'\""
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = f\"select * from nih_sra_datastore.sra.metadata where biosample='{response.json()['dbgap']['sra_sample_id']}'\"\n",
    "query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "09a9a724",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "48eac0c24c9e4b65a2b8e05cc08fc5e7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "IntProgress(value=0, max=300)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from ipywidgets import IntProgress\n",
    "from IPython.display import display\n",
    "max_count = 300\n",
    "\n",
    "progress_bar = IntProgress(min=0, max=max_count) # instantiate the bar\n",
    "display(progress_bar) # display the bar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ae2c7cc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = searchClient.run_query(query, return_type='dataframe', progessIndicator=progress_bar)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3b4db380",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>acc</th>\n",
       "      <td>SRR8016410</td>\n",
       "      <td>SRR5154616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assay_type</th>\n",
       "      <td>WGS</td>\n",
       "      <td>WGS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>center_name</th>\n",
       "      <td>Macrogen</td>\n",
       "      <td>Macrogen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>consent</th>\n",
       "      <td>DS-CVD-IRB-NPU-MDS</td>\n",
       "      <td>DS-CVD-IRB-NPU-MDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experiment</th>\n",
       "      <td>SRX4849106</td>\n",
       "      <td>SRX2472219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <td>NWD403016</td>\n",
       "      <td>NWD403016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>instrument</th>\n",
       "      <td>HiSeq X Ten</td>\n",
       "      <td>HiSeq X Ten</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>librarylayout</th>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PAIRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>libraryselection</th>\n",
       "      <td>RANDOM</td>\n",
       "      <td>RANDOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>librarysource</th>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>GENOMIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>platform</th>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>ILLUMINA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_acc</th>\n",
       "      <td>SRS1903809</td>\n",
       "      <td>SRS1903809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample</th>\n",
       "      <td>SAMN06062404</td>\n",
       "      <td>SAMN06062404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organism</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_study</th>\n",
       "      <td>SRP095692</td>\n",
       "      <td>SRP095692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>releasedate</th>\n",
       "      <td>2019-02-27T00:00:00.000Z</td>\n",
       "      <td>2019-02-27T00:00:00.000Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bioproject</th>\n",
       "      <td>PRJNA343581</td>\n",
       "      <td>PRJNA343581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mbytes</th>\n",
       "      <td>21901</td>\n",
       "      <td>29163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>loaddate</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>avgspotlen</th>\n",
       "      <td>300</td>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mbases</th>\n",
       "      <td>130478</td>\n",
       "      <td>130478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insertsize</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>library_name</th>\n",
       "      <td>NWD403016</td>\n",
       "      <td>NWD403016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosamplemodel_sam</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>collection_date_sam</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geo_loc_name_country_calc</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geo_loc_name_country_continent_calc</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geo_loc_name_sam</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ena_first_public_run</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ena_last_update_run</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name_sam</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>datastore_filetype</th>\n",
       "      <td>[crai, cram, vcf, vcf_index]</td>\n",
       "      <td>[bam, run.zq, sra]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>datastore_provider</th>\n",
       "      <td>[gs, s3]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>datastore_region</th>\n",
       "      <td>[gs.US, s3.us-east-1]</td>\n",
       "      <td>[gs.US, ncbi.dbgap, s3.us-east-1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>attributes</th>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>jattr</th>\n",
       "      <td>{\"sex_calc\": \"female\", \"assemblyname\": \"GRCh38...</td>\n",
       "      <td>{\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                     0  \\\n",
       "acc                                                                         SRR8016410   \n",
       "assay_type                                                                         WGS   \n",
       "center_name                                                                   Macrogen   \n",
       "consent                                                             DS-CVD-IRB-NPU-MDS   \n",
       "experiment                                                                  SRX4849106   \n",
       "sample_name                                                                  NWD403016   \n",
       "instrument                                                                 HiSeq X Ten   \n",
       "librarylayout                                                                   PAIRED   \n",
       "libraryselection                                                                RANDOM   \n",
       "librarysource                                                                  GENOMIC   \n",
       "platform                                                                      ILLUMINA   \n",
       "sample_acc                                                                  SRS1903809   \n",
       "biosample                                                                 SAMN06062404   \n",
       "organism                                                                  Homo sapiens   \n",
       "sra_study                                                                    SRP095692   \n",
       "releasedate                                                   2019-02-27T00:00:00.000Z   \n",
       "bioproject                                                                 PRJNA343581   \n",
       "mbytes                                                                           21901   \n",
       "loaddate                                                                          None   \n",
       "avgspotlen                                                                         300   \n",
       "mbases                                                                          130478   \n",
       "insertsize                                                                        None   \n",
       "library_name                                                                NWD403016    \n",
       "biosamplemodel_sam                                                                  []   \n",
       "collection_date_sam                                                               None   \n",
       "geo_loc_name_country_calc                                                         None   \n",
       "geo_loc_name_country_continent_calc                                               None   \n",
       "geo_loc_name_sam                                                                    []   \n",
       "ena_first_public_run                                                                []   \n",
       "ena_last_update_run                                                                 []   \n",
       "sample_name_sam                                                                     []   \n",
       "datastore_filetype                                        [crai, cram, vcf, vcf_index]   \n",
       "datastore_provider                                                            [gs, s3]   \n",
       "datastore_region                                                 [gs.US, s3.us-east-1]   \n",
       "attributes                           [{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...   \n",
       "jattr                                {\"sex_calc\": \"female\", \"assemblyname\": \"GRCh38...   \n",
       "\n",
       "                                                                                     1  \n",
       "acc                                                                         SRR5154616  \n",
       "assay_type                                                                         WGS  \n",
       "center_name                                                                   Macrogen  \n",
       "consent                                                             DS-CVD-IRB-NPU-MDS  \n",
       "experiment                                                                  SRX2472219  \n",
       "sample_name                                                                  NWD403016  \n",
       "instrument                                                                 HiSeq X Ten  \n",
       "librarylayout                                                                   PAIRED  \n",
       "libraryselection                                                                RANDOM  \n",
       "librarysource                                                                  GENOMIC  \n",
       "platform                                                                      ILLUMINA  \n",
       "sample_acc                                                                  SRS1903809  \n",
       "biosample                                                                 SAMN06062404  \n",
       "organism                                                                  Homo sapiens  \n",
       "sra_study                                                                    SRP095692  \n",
       "releasedate                                                   2019-02-27T00:00:00.000Z  \n",
       "bioproject                                                                 PRJNA343581  \n",
       "mbytes                                                                           29163  \n",
       "loaddate                                                                          None  \n",
       "avgspotlen                                                                         300  \n",
       "mbases                                                                          130478  \n",
       "insertsize                                                                        None  \n",
       "library_name                                                                 NWD403016  \n",
       "biosamplemodel_sam                                                                  []  \n",
       "collection_date_sam                                                               None  \n",
       "geo_loc_name_country_calc                                                         None  \n",
       "geo_loc_name_country_continent_calc                                               None  \n",
       "geo_loc_name_sam                                                                    []  \n",
       "ena_first_public_run                                                                []  \n",
       "ena_last_update_run                                                                 []  \n",
       "sample_name_sam                                                                     []  \n",
       "datastore_filetype                                                  [bam, run.zq, sra]  \n",
       "datastore_provider                                                      [gs, ncbi, s3]  \n",
       "datastore_region                                     [gs.US, ncbi.dbgap, s3.us-east-1]  \n",
       "attributes                           [{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...  \n",
       "jattr                                {\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8c530523",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sex': 'female',\n",
       " 'body_site': 'peripheral blood',\n",
       " 'repository': 'GeneSTAR',\n",
       " 'sample_use': ['Seq_DNA_SNP_CNV'],\n",
       " 'analyte_type': 'DNA',\n",
       " 'biosample_id': 'SAMN06062404',\n",
       " 'consent_code': 2,\n",
       " 'dbgap_status': 'Loaded',\n",
       " 'sra_sample_id': 'SAMN06062404',\n",
       " 'dbgap_sample_id': 2170791,\n",
       " 'study_accession': 'phs001218.v2.p1',\n",
       " 'dbgap_subject_id': 1797782,\n",
       " 'sra_data_details': {'runs': '2',\n",
       "  'bases': '260957674200',\n",
       "  'center': 'Macrogen',\n",
       "  'status': 'public',\n",
       "  'size_Gb': '49',\n",
       "  'platform': 'ILLUMINA',\n",
       "  'experiments': '2',\n",
       "  'experiment_type': 'WGS'},\n",
       " 'study_subject_id': 'phs001218.v2_GS08900974',\n",
       " 'consent_short_name': 'DS-CVD-IRB-NPU-MDS',\n",
       " 'study_with_consent': 'phs001218.c2',\n",
       " 'submitted_sample_id': 'NWD403016',\n",
       " 'submitted_subject_id': 'GS08900974',\n",
       " 'study_accession_with_consent': 'phs001218.v2.p1.c2'}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.json()['dbgap']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "62f5e68d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>attribute</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sex_calc</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>assemblyname</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bases</td>\n",
       "      <td>130478837100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bytes</td>\n",
       "      <td>30580440819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>consent_code</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>run_file_create_date</td>\n",
       "      <td>2018-10-18T05:31:00.000Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>run_file_version</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>alignment_provider_run</td>\n",
       "      <td>Sequencing Center</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>analyte_type_sam</td>\n",
       "      <td>DNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>biospecimen_repository_sam</td>\n",
       "      <td>GeneSTAR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>biospecimen_repository_sample_id_sam</td>\n",
       "      <td>NWD403016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>body_site_sam</td>\n",
       "      <td>peripheral blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>gap_accession_sam</td>\n",
       "      <td>phs001218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>study_design_sam</td>\n",
       "      <td>Prospective Longitudinal Cohort</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>study_name_sam</td>\n",
       "      <td>NHLBI TOPMed: Genetic Study of Atherosclerosis...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>submitted_subject_id_sam</td>\n",
       "      <td>GS08900974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>phs</td>\n",
       "      <td>phs001218.c2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               attribute  \\\n",
       "0                               sex_calc   \n",
       "1                           assemblyname   \n",
       "2                                  bases   \n",
       "3                                  bytes   \n",
       "4                           consent_code   \n",
       "5                   run_file_create_date   \n",
       "6                       run_file_version   \n",
       "7                 alignment_provider_run   \n",
       "8                       analyte_type_sam   \n",
       "9             biospecimen_repository_sam   \n",
       "10  biospecimen_repository_sample_id_sam   \n",
       "11                         body_site_sam   \n",
       "12                     gap_accession_sam   \n",
       "13                      study_design_sam   \n",
       "14                        study_name_sam   \n",
       "15              submitted_subject_id_sam   \n",
       "16                                   phs   \n",
       "\n",
       "                                                value  \n",
       "0                                              female  \n",
       "1                                    GCA_000001405.13  \n",
       "2                                        130478837100  \n",
       "3                                         30580440819  \n",
       "4                                                   2  \n",
       "5                            2018-10-18T05:31:00.000Z  \n",
       "6                                                   2  \n",
       "7                                   Sequencing Center  \n",
       "8                                                 DNA  \n",
       "9                                            GeneSTAR  \n",
       "10                                          NWD403016  \n",
       "11                                   peripheral blood  \n",
       "12                                          phs001218  \n",
       "13                    Prospective Longitudinal Cohort  \n",
       "14  NHLBI TOPMed: Genetic Study of Atherosclerosis...  \n",
       "15                                         GS08900974  \n",
       "16                                       phs001218.c2  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "    SELECT att.k attribute, att.v value\n",
    " FROM nih_sra_datastore.sra.metadata s, unnest(attributes) att\n",
    "where s.acc = 'SRR5154616' and att.k != 'primary_search' \"\"\"\n",
    "\n",
    "df = searchClient.run_query(query, return_type='dataframe', progessIndicator=progress_bar)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5753e2b",
   "metadata": {},
   "source": [
    "### Query same files in NCBI DRS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2e6b8521",
   "metadata": {},
   "outputs": [],
   "source": [
    "from fasp.loc import SRADRSClient\n",
    "dcl = SRADRSClient('https://locate.be-md.ncbi.nlm.nih.gov')\n",
    "drs_id = dcl.acc2drsID('SRR8016410')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2c2b023f",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_response = dcl.get_object(drs_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b96aa065",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'access_methods': [{'access_id': '01f5af5012b771d644dc9d66655e5b35371d09daad04096ed610580ef2f3af18', 'region': 'gs.US', 'type': 'https'}, {'access_id': 'a9dd5a3b85e5c3289692315cb59d1b679e4f0c1a49a8b242b30efd10619b50e8', 'region': 's3.us-east-1', 'type': 'https'}], 'checksums': [{'checksum': '2f5a2cbcaefefefd89cd6ffcfb84b959', 'type': 'md5'}], 'created_time': '2018-10-11T12:09:44Z', 'id': '2f5a2cbcaefefefd89cd6ffcfb84b959', 'name': 'NWD403016.freeze5.v1.vcf.gz.csi', 'self_url': 'drs://locate.be-md.ncbi.nlm.nih.gov/2f5a2cbcaefefefd89cd6ffcfb84b959', 'size': 1711170}\n",
      "gs.US\n",
      "s3.us-east-1\n",
      "________________________________________________________________________________\n",
      "{'access_methods': [{'access_id': 'd2be5fff49bec88fb8d2395bb95c14fc95aa02539a0564f3e199e7eeca0d701a', 'region': 'gs.US', 'type': 'https'}, {'access_id': '8a9085401febe6114a84743c3ea6f52760a9aca1661cdf22f96a6d6d3a5f99a1', 'region': 's3.us-east-1', 'type': 'https'}], 'checksums': [{'checksum': '308fa51b0b2ba6aab94315278341343d', 'type': 'md5'}], 'created_time': '2018-10-11T12:09:44Z', 'id': '308fa51b0b2ba6aab94315278341343d', 'name': 'NWD403016.b38.irc.v1.cram', 'self_url': 'drs://locate.be-md.ncbi.nlm.nih.gov/308fa51b0b2ba6aab94315278341343d', 'size': 20280571399}\n",
      "gs.US\n",
      "s3.us-east-1\n",
      "________________________________________________________________________________\n",
      "{'access_methods': [{'access_id': '76220dbc33720528f8efb77dc5f5ba7369cdb8ac557fed207d81e30f65689231', 'region': 'gs.US', 'type': 'https'}, {'access_id': '1318b1aaf92482904ed534f2991498bced6f4510137626bd5760947f7e24e2f9', 'region': 's3.us-east-1', 'type': 'https'}], 'checksums': [{'checksum': '80c20cdd223bfa7c5ae801a1a62f7151', 'type': 'md5'}], 'created_time': '2018-10-11T12:09:44Z', 'id': '80c20cdd223bfa7c5ae801a1a62f7151', 'name': 'NWD403016.freeze5.v1.vcf.gz', 'self_url': 'drs://locate.be-md.ncbi.nlm.nih.gov/80c20cdd223bfa7c5ae801a1a62f7151', 'size': 2682022077}\n",
      "gs.US\n",
      "s3.us-east-1\n",
      "________________________________________________________________________________\n",
      "{'access_methods': [{'access_id': '9faf02933584f88aa6ca26cf4c85cdf5cd64f1841127a1f4b6b5c1a24595ee73', 'region': 'gs.US', 'type': 'https'}, {'access_id': 'ee665f8f5e7e2ed9fd93ee0b2aa4488d6510d5dcb321ece92c4b4aa02120a26b', 'region': 's3.us-east-1', 'type': 'https'}], 'checksums': [{'checksum': 'f12a3897ce282c4201f6ea2fd4314a8d', 'type': 'md5'}], 'created_time': '2018-10-11T12:09:44Z', 'id': 'f12a3897ce282c4201f6ea2fd4314a8d', 'name': 'NWD403016.b38.irc.v1.cram.crai', 'self_url': 'drs://locate.be-md.ncbi.nlm.nih.gov/f12a3897ce282c4201f6ea2fd4314a8d', 'size': 1440693}\n",
      "gs.US\n",
      "s3.us-east-1\n",
      "________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "for r in run_response['contents']:\n",
    "    file = dcl.get_object(r['id'])\n",
    "    print(file)\n",
    "    for a in file['access_methods']:\n",
    "        print(a['region'])\n",
    "        #dcl.get_access_url(r['id'], a['access_id'])\n",
    "    print('_'*80)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf59c1bb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
