{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running the VUS cooccurence workflow via WES on TopMed COPDGene data\n",
    "Other than the COPDGene vcf file which was the target of the analysis, the files used in the example were uploaded to the BioDataCatalyst Seven Bridges server.\n",
    "\n",
    "Prior to the workflow the COPDGene_phs000951_TOPMed_WGS_freeze.8.chr13.hg38.c1.vcf was filtered down to the range of BRCA2 by using BCFTools."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/forei/imfenvs/fasp/lib/python3.9/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.\n",
      "  warnings.warn(msg)\n"
     ]
    }
   ],
   "source": [
    "from fasp.workflow import sbWESClient\n",
    "cl = sbWESClient('https://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/ga4gh/wes/v1', 'forei/fasp-vus',\n",
    "                     '~/.keys/sbbdc_key.json', debug=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setting up the WES run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    \"project\": \"forei/fasp-vus\",\n",
    "    \"inputs\": {\n",
    "        'gnomad_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617c5ce7e6261a31b6d12de3',\n",
    "                        'name': 'gnomad_chr13_brca2.vcf',\n",
    "                        'class': 'File'},\n",
    "\n",
    "        'vcf_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617c77bce6261a31b6d12f0a',\n",
    "                     'name': 'COPDGene_phs000951_TOPMed_WGS_freeze.8.chr13.hg38.c1.filtered.vcf',\n",
    "                     'class': 'File'},\n",
    "\n",
    "        'pathogenicity_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617c5ce6e6261a31b6d12ddc',\n",
    "                               'name': 'clinvar_brca2.tsv',\n",
    "                               'class': 'File'}\n",
    "        }\n",
    "    }\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have formulated the body in the way that it can be passed to a client function as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'b20b60f9-5125-4088-b61f-de791c9278b0'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import json\n",
    "run_id= cl.runGenericWorkflow(\n",
    "    workflow_url=\"sbg://forei/fasp-vus/cooccurrence/2\",\n",
    "    workflow_params = json.dumps(params),\n",
    "    workflow_type = \"CWL\",\n",
    "    workflow_type_version = \"sbg:draft-2\"\n",
    ")\n",
    "run_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COMPLETE\n",
      "0:18:29\n"
     ]
    }
   ],
   "source": [
    "import dateutil.parser\n",
    "print(cl.getTaskStatus(run_id))\n",
    "log = cl.GetRunLog(run_id)\n",
    "if log['run_log']['start_time']:\n",
    "    start = dateutil.parser.isoparse(log['run_log']['start_time'])\n",
    "    end = dateutil.parser.isoparse(log['run_log']['end_time'])\n",
    "    duration = end - start\n",
    "    print(str(duration))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'request': {'tags': {},\n",
       "  'workflow_params': {'name': 'cooccurrence run - 11-01-21 01:44:19',\n",
       "   'project': 'forei/fasp-vus',\n",
       "   'inputs': {'p2': None,\n",
       "    'save_files': None,\n",
       "    'gene': None,\n",
       "    'chromosome': None,\n",
       "    'data_directory': None,\n",
       "    'ensembl_release': None,\n",
       "    'pathology_file': None,\n",
       "    'gnomad_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f3fc9fe2ce002310283b8',\n",
       "     'name': 'gnomad_chr13_brca2.vcf',\n",
       "     'class': 'File'},\n",
       "    'hg_version': None,\n",
       "    'vcf_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f3fc9fe2ce002310283ba',\n",
       "     'name': '_1_COPDGene_phs000951_TOPMed_WGS_freeze.8.chr13.hg38.c1.filtered.vcf',\n",
       "     'class': 'File'},\n",
       "    'phased': None,\n",
       "    'pathogenicity_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4673fe2ce0023102842c',\n",
       "     'name': 'clinvar_brca2.tsv',\n",
       "     'class': 'File'}}},\n",
       "  'workflow_type': 'CWL',\n",
       "  'workflow_engine_params': {},\n",
       "  'workflow_url': 'sbg://forei/fasp-vus/cooccurrence/2'},\n",
       " 'state': 'COMPLETE',\n",
       " 'outputs': {'vpi_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b11',\n",
       "   'name': '_2_BRCA2-vpi.json',\n",
       "   'class': 'File'},\n",
       "  'pathology_output': None,\n",
       "  'all_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b0f',\n",
       "   'name': '_2_BRCA2-all.json',\n",
       "   'class': 'File'},\n",
       "  'out_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b17',\n",
       "   'name': '_3_BRCA2-cooccurrences.json',\n",
       "   'class': 'File'},\n",
       "  'ipv_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b14',\n",
       "   'name': '_2_BRCA2-ipv.json',\n",
       "   'class': 'File'},\n",
       "  'tout_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b13',\n",
       "   'name': '_2_BRCA2-tout.json',\n",
       "   'class': 'File'}},\n",
       " 'run_id': 'b20b60f9-5125-4088-b61f-de791c9278b0',\n",
       " 'run_log': {'name': 'cooccurrence run - 11-01-21 01:44:19',\n",
       "  'cmd': None,\n",
       "  'start_time': '2021-11-01T01:44:20Z',\n",
       "  'end_time': '2021-11-01T02:02:49Z',\n",
       "  'stdout': None,\n",
       "  'stderr': None,\n",
       "  'exit_code': 0},\n",
       " 'task_logs': [{'name': 'cooccurrence',\n",
       "   'cmd': '/usr/bin/python3 /federated-analysis/app/cooccurrence/cooccurrenceFinder.py --g BRCA2 --vcf /sbgenomics/Projects/2aa9e16d-ede5-43d6-9605-04fab05f34b3/_1_COPDGene_phs000951_TOPMed_WGS_freeze.8.chr13.hg38.c1.filtered.vcf --h 38 --e 99 --c 13 --p True --p2 0.001 --vpf /sbgenomics/Projects/2aa9e16d-ede5-43d6-9605-04fab05f34b3/clinvar_brca2.tsv --d /var/tmp/pyensembl-cache --save True --gf /sbgenomics/Projects/2aa9e16d-ede5-43d6-9605-04fab05f34b3/gnomad_chr13_brca2.vcf',\n",
       "   'start_time': '2021-11-01T01:48:08Z',\n",
       "   'end_time': '2021-11-01T02:02:46Z',\n",
       "   'stdout': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac4e6261a31b6d14af3',\n",
       "   'stderr': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac4e6261a31b6d14af3',\n",
       "   'exit_code': 0}]}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.GetRunLog(run_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting the results - via DRS\n",
    "Once the run is complete, further steps can use DRS to obtain the file output from the workflow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'vpi_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b11',\n",
       "  'name': '_2_BRCA2-vpi.json',\n",
       "  'class': 'File'},\n",
       " 'pathology_output': None,\n",
       " 'all_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b0f',\n",
       "  'name': '_2_BRCA2-all.json',\n",
       "  'class': 'File'},\n",
       " 'out_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b17',\n",
       "  'name': '_3_BRCA2-cooccurrences.json',\n",
       "  'class': 'File'},\n",
       " 'ipv_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b14',\n",
       "  'name': '_2_BRCA2-ipv.json',\n",
       "  'class': 'File'},\n",
       " 'tout_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b13',\n",
       "  'name': '_2_BRCA2-tout.json',\n",
       "  'class': 'File'}}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "runLog = cl.GetRunLog(run_id)\n",
    "runLog['outputs']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b17'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resultsDRSID = runLog['outputs']['out_file']['path']\n",
    "resultsDRSID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the BioDataCatalyst SB DRS Server to retrieve the results files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from fasp.loc import sbbdcDRSClient\n",
    "drsClient = sbbdcDRSClient('~/.keys/sevenbridges_keys.json', 's3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DRS GetObject\n",
    "Here's how we then get details of the file. Note that here only the id portion of the DRS URI is being passed. It is the job of a metaresolver to look at that URI and to determine where to send the id. As noted, we are passing up on the opportunity to use a metaresolver and putting in the id manually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': '617f4ac9e6261a31b6d14b17',\n",
       " 'name': '_3_BRCA2-cooccurrences.json',\n",
       " 'size': 81104,\n",
       " 'checksums': [{'type': 'etag',\n",
       "   'checksum': 'cad46dcd72010874721bd86a5297c77c-1'}],\n",
       " 'self_uri': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617f4ac9e6261a31b6d14b17',\n",
       " 'created_time': '2021-11-01T02:02:49Z',\n",
       " 'updated_time': '2021-11-01T02:02:49Z',\n",
       " 'mime_type': 'application/json',\n",
       " 'access_methods': [{'type': 's3',\n",
       "   'region': 'us-east-1',\n",
       "   'access_id': 'aws-us-east-1'}]}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sbDRSID = resultsDRSID.split('/')[-1]\n",
    "fileDetails = drsClient.getObject(sbDRSID)\n",
    "fileDetails"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = drsClient.getAccessURL(sbDRSID,'s3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Downloading the file\n",
    "Now we can use the url obtained to download the file. We'll create a small function to encapsulate the download."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import os\n",
    "def download(url, file_path):\n",
    "    with open(os.path.expanduser(file_path), \"wb\") as file:\n",
    "        response = requests.get(url)\n",
    "        file.write(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "fullPath = fileDetails['name']\n",
    "download(url, fullPath)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(fullPath) as json_file:\n",
    "    data = json.load(json_file)\n",
    "# delete the local copy of the file\n",
    "#os.remove(fullPath)\n",
    "# and look at the contents\n",
    "# data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vus</th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>n</th>\n",
       "      <th>k</th>\n",
       "      <th>likelihood</th>\n",
       "      <th>maxPop</th>\n",
       "      <th>maxPopFreq</th>\n",
       "      <th>cohortFreq</th>\n",
       "      <th>no_pathogenic_coocurrs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(13, 32315831, 'G', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>9345</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002811</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.914742</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(13, 32318080, 'C', 'T')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10195</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025604</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.997944</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(13, 32318598, 'T', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>9047</td>\n",
       "      <td>1</td>\n",
       "      <td>0.003733</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.885572</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(13, 32319654, 'A', 'G')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>8119</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009026</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.794734</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(13, 32321240, 'G', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10196</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025580</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.998042</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>(13, 32325741, 'C', 'T')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10194</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025628</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.997847</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>(13, 32331128, 'G', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10194</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025628</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.997847</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>(13, 32333969, 'A', 'G')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10194</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025628</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.997847</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>(13, 32338918, 'A', 'G')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10194</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025628</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.997847</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>(13, 32340868, 'G', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10193</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025653</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.997749</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>(13, 32343709, 'GA', 'G')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>3580</td>\n",
       "      <td>1</td>\n",
       "      <td>0.678061</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.350431</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>(13, 32345879, 'G', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10194</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025628</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.997847</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>(13, 32346707, 'T', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10194</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025628</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.997847</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>(13, 32353757, 'C', 'T')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>5163</td>\n",
       "      <td>1</td>\n",
       "      <td>0.150344</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.505384</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>(13, 32354180, 'T', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10202</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025434</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>(13, 32354271, 'G', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>8080</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009367</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.790916</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>(13, 32355250, 'T', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10202</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025434</td>\n",
       "      <td>eas</td>\n",
       "      <td>0.977087</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>(13, 32356799, 'G', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10216</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025097</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>(13, 32359800, 'A', 'G')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10202</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025434</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>(13, 32362509, 'T', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>8062</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009529</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.789154</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>(13, 32365177, 'C', 'T')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7875</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011385</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.770850</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>(13, 32366208, 'C', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7873</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011407</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.770654</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>(13, 32367579, 'A', 'AGATGGCTTG')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7883</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011299</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.771633</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>(13, 32371874, 'A', 'G')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10202</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025434</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>(13, 32372278, 'C', 'T')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10202</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025434</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>(13, 32375245, 'T', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>8016</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009956</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.784652</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>(13, 32375563, 'C', 'CT')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7971</td>\n",
       "      <td>1</td>\n",
       "      <td>0.010391</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.780247</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>(13, 32376120, 'A', 'G')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7787</td>\n",
       "      <td>1</td>\n",
       "      <td>0.012379</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.762236</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>(13, 32379251, 'T', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7834</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011838</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.766836</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>(13, 32380424, 'G', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7992</td>\n",
       "      <td>1</td>\n",
       "      <td>0.010186</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.782302</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>(13, 32380604, 'T', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10202</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025434</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>(13, 32381260, 'C', 'T')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7827</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011917</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.766151</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>(13, 32385062, 'C', 'T')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7826</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011928</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.766053</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>(13, 32385702, 'A', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7830</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011883</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.766445</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>(13, 32386226, 'T', 'G')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>8009</td>\n",
       "      <td>1</td>\n",
       "      <td>0.010022</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.783966</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>(13, 32387052, 'G', 'GC')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10202</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025434</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>(13, 32387908, 'A', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7816</td>\n",
       "      <td>1</td>\n",
       "      <td>0.012043</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.765074</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>(13, 32387951, 'C', 'T')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7826</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011928</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.766053</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>(13, 32388121, 'C', 'CT')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>9942</td>\n",
       "      <td>2</td>\n",
       "      <td>0.032573</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.973179</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>(13, 32388181, 'T', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7829</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011894</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.766347</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>(13, 32388666, 'C', 'CT')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>5653</td>\n",
       "      <td>1</td>\n",
       "      <td>0.094317</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.553348</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>(13, 32389570, 'C', 'T')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7987</td>\n",
       "      <td>1</td>\n",
       "      <td>0.010234</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.781813</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>(13, 32391514, 'T', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10202</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025434</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>(13, 32392589, 'CA', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>9560</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002291</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.935787</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>(13, 32393329, 'A', 'G')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7827</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011917</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.766151</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>(13, 32394454, 'G', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>8006</td>\n",
       "      <td>1</td>\n",
       "      <td>0.010051</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.783673</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>(13, 32394470, 'A', 'G')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7827</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011917</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.766151</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>(13, 32396599, 'G', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7828</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011906</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.766249</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>(13, 32397288, 'T', 'C')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10202</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025434</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>(13, 32397569, 'AT', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7827</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011917</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.766151</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>(13, 32399885, 'C', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10216</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025097</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>(13, 32400151, 'T', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>9620</td>\n",
       "      <td>2</td>\n",
       "      <td>0.044252</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.941660</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>(13, 32344166, 'GA', 'G')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7460</td>\n",
       "      <td>1</td>\n",
       "      <td>0.016898</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.730227</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>(13, 32388666, 'C', 'CTT')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>5162</td>\n",
       "      <td>1</td>\n",
       "      <td>0.150487</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.505286</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>(13, 32323151, 'ATT', 'A')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>4106</td>\n",
       "      <td>1</td>\n",
       "      <td>0.411052</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.401919</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>(13, 32343709, 'G', 'GA')</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.001</td>\n",
       "      <td>635</td>\n",
       "      <td>1</td>\n",
       "      <td>11.176525</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.062157</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  vus        p1     p2      n  k  likelihood  \\\n",
       "0            (13, 32315831, 'G', 'A')  0.000049  0.001   9345  1    0.002811   \n",
       "1            (13, 32318080, 'C', 'T')  0.000049  0.001  10195  2    0.025604   \n",
       "2            (13, 32318598, 'T', 'C')  0.000049  0.001   9047  1    0.003733   \n",
       "3            (13, 32319654, 'A', 'G')  0.000049  0.001   8119  1    0.009026   \n",
       "4            (13, 32321240, 'G', 'C')  0.000049  0.001  10196  2    0.025580   \n",
       "5            (13, 32325741, 'C', 'T')  0.000049  0.001  10194  2    0.025628   \n",
       "6            (13, 32331128, 'G', 'A')  0.000049  0.001  10194  2    0.025628   \n",
       "7            (13, 32333969, 'A', 'G')  0.000049  0.001  10194  2    0.025628   \n",
       "8            (13, 32338918, 'A', 'G')  0.000049  0.001  10194  2    0.025628   \n",
       "9            (13, 32340868, 'G', 'C')  0.000049  0.001  10193  2    0.025653   \n",
       "10          (13, 32343709, 'GA', 'G')  0.000049  0.001   3580  1    0.678061   \n",
       "11           (13, 32345879, 'G', 'A')  0.000049  0.001  10194  2    0.025628   \n",
       "12           (13, 32346707, 'T', 'C')  0.000049  0.001  10194  2    0.025628   \n",
       "13           (13, 32353757, 'C', 'T')  0.000049  0.001   5163  1    0.150344   \n",
       "14           (13, 32354180, 'T', 'C')  0.000049  0.001  10202  2    0.025434   \n",
       "15           (13, 32354271, 'G', 'A')  0.000049  0.001   8080  1    0.009367   \n",
       "16           (13, 32355250, 'T', 'C')  0.000049  0.001  10202  2    0.025434   \n",
       "17           (13, 32356799, 'G', 'A')  0.000049  0.001  10216  2    0.025097   \n",
       "18           (13, 32359800, 'A', 'G')  0.000049  0.001  10202  2    0.025434   \n",
       "19           (13, 32362509, 'T', 'C')  0.000049  0.001   8062  1    0.009529   \n",
       "20           (13, 32365177, 'C', 'T')  0.000049  0.001   7875  1    0.011385   \n",
       "21           (13, 32366208, 'C', 'A')  0.000049  0.001   7873  1    0.011407   \n",
       "22  (13, 32367579, 'A', 'AGATGGCTTG')  0.000049  0.001   7883  1    0.011299   \n",
       "23           (13, 32371874, 'A', 'G')  0.000049  0.001  10202  2    0.025434   \n",
       "24           (13, 32372278, 'C', 'T')  0.000049  0.001  10202  2    0.025434   \n",
       "25           (13, 32375245, 'T', 'C')  0.000049  0.001   8016  1    0.009956   \n",
       "26          (13, 32375563, 'C', 'CT')  0.000049  0.001   7971  1    0.010391   \n",
       "27           (13, 32376120, 'A', 'G')  0.000049  0.001   7787  1    0.012379   \n",
       "28           (13, 32379251, 'T', 'C')  0.000049  0.001   7834  1    0.011838   \n",
       "29           (13, 32380424, 'G', 'A')  0.000049  0.001   7992  1    0.010186   \n",
       "30           (13, 32380604, 'T', 'A')  0.000049  0.001  10202  2    0.025434   \n",
       "31           (13, 32381260, 'C', 'T')  0.000049  0.001   7827  1    0.011917   \n",
       "32           (13, 32385062, 'C', 'T')  0.000049  0.001   7826  1    0.011928   \n",
       "33           (13, 32385702, 'A', 'C')  0.000049  0.001   7830  1    0.011883   \n",
       "34           (13, 32386226, 'T', 'G')  0.000049  0.001   8009  1    0.010022   \n",
       "35          (13, 32387052, 'G', 'GC')  0.000049  0.001  10202  2    0.025434   \n",
       "36           (13, 32387908, 'A', 'C')  0.000049  0.001   7816  1    0.012043   \n",
       "37           (13, 32387951, 'C', 'T')  0.000049  0.001   7826  1    0.011928   \n",
       "38          (13, 32388121, 'C', 'CT')  0.000049  0.001   9942  2    0.032573   \n",
       "39           (13, 32388181, 'T', 'C')  0.000049  0.001   7829  1    0.011894   \n",
       "40          (13, 32388666, 'C', 'CT')  0.000049  0.001   5653  1    0.094317   \n",
       "41           (13, 32389570, 'C', 'T')  0.000049  0.001   7987  1    0.010234   \n",
       "42           (13, 32391514, 'T', 'C')  0.000049  0.001  10202  2    0.025434   \n",
       "43          (13, 32392589, 'CA', 'C')  0.000049  0.001   9560  1    0.002291   \n",
       "44           (13, 32393329, 'A', 'G')  0.000049  0.001   7827  1    0.011917   \n",
       "45           (13, 32394454, 'G', 'A')  0.000049  0.001   8006  1    0.010051   \n",
       "46           (13, 32394470, 'A', 'G')  0.000049  0.001   7827  1    0.011917   \n",
       "47           (13, 32396599, 'G', 'C')  0.000049  0.001   7828  1    0.011906   \n",
       "48           (13, 32397288, 'T', 'C')  0.000049  0.001  10202  2    0.025434   \n",
       "49          (13, 32397569, 'AT', 'A')  0.000049  0.001   7827  1    0.011917   \n",
       "50           (13, 32399885, 'C', 'A')  0.000049  0.001  10216  2    0.025097   \n",
       "51           (13, 32400151, 'T', 'A')  0.000049  0.001   9620  2    0.044252   \n",
       "52          (13, 32344166, 'GA', 'G')  0.000049  0.001   7460  1    0.016898   \n",
       "53         (13, 32388666, 'C', 'CTT')  0.000049  0.001   5162  1    0.150487   \n",
       "54         (13, 32323151, 'ATT', 'A')  0.000049  0.001   4106  1    0.411052   \n",
       "55          (13, 32343709, 'G', 'GA')  0.000049  0.001    635  1   11.176525   \n",
       "\n",
       "   maxPop maxPopFreq  cohortFreq  no_pathogenic_coocurrs  \n",
       "0    None       None    0.914742                       1  \n",
       "1    None       None    0.997944                       1  \n",
       "2    None       None    0.885572                       1  \n",
       "3    None       None    0.794734                       1  \n",
       "4    None       None    0.998042                       1  \n",
       "5    None       None    0.997847                       1  \n",
       "6    None       None    0.997847                       1  \n",
       "7    None       None    0.997847                       1  \n",
       "8    None       None    0.997847                       1  \n",
       "9    None       None    0.997749                       1  \n",
       "10   None       None    0.350431                       1  \n",
       "11   None       None    0.997847                       1  \n",
       "12   None       None    0.997847                       1  \n",
       "13   None       None    0.505384                       1  \n",
       "14   None       None    0.998630                       1  \n",
       "15   None       None    0.790916                       1  \n",
       "16    eas   0.977087    0.998630                       1  \n",
       "17   None       None    1.000000                       1  \n",
       "18   None       None    0.998630                       1  \n",
       "19   None       None    0.789154                       1  \n",
       "20   None       None    0.770850                       1  \n",
       "21   None       None    0.770654                       1  \n",
       "22   None       None    0.771633                       1  \n",
       "23   None       None    0.998630                       1  \n",
       "24   None       None    0.998630                       1  \n",
       "25   None       None    0.784652                       1  \n",
       "26   None       None    0.780247                       1  \n",
       "27   None       None    0.762236                       1  \n",
       "28   None       None    0.766836                       1  \n",
       "29   None       None    0.782302                       1  \n",
       "30   None       None    0.998630                       1  \n",
       "31   None       None    0.766151                       1  \n",
       "32   None       None    0.766053                       1  \n",
       "33   None       None    0.766445                       1  \n",
       "34   None       None    0.783966                       1  \n",
       "35   None       None    0.998630                       1  \n",
       "36   None       None    0.765074                       1  \n",
       "37   None       None    0.766053                       1  \n",
       "38   None       None    0.973179                       1  \n",
       "39   None       None    0.766347                       1  \n",
       "40   None       None    0.553348                       1  \n",
       "41   None       None    0.781813                       1  \n",
       "42   None       None    0.998630                       1  \n",
       "43   None       None    0.935787                       1  \n",
       "44   None       None    0.766151                       1  \n",
       "45   None       None    0.783673                       1  \n",
       "46   None       None    0.766151                       1  \n",
       "47   None       None    0.766249                       1  \n",
       "48   None       None    0.998630                       1  \n",
       "49   None       None    0.766151                       1  \n",
       "50   None       None    1.000000                       1  \n",
       "51   None       None    0.941660                       1  \n",
       "52   None       None    0.730227                       1  \n",
       "53   None       None    0.505286                       1  \n",
       "54   None       None    0.401919                       1  \n",
       "55   None       None    0.062157                       1  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# flatten cooccurrence output\n",
    "flat_vus = []\n",
    "for k, v in data['cooccurring vus'].items():\n",
    "    pathogenic_count = len(v['pathogenic variants'])\n",
    "    ## this is a pythonic way of merging dicts - it is cryptic\n",
    "    z = {**{\"vus\":k}, **v['likelihood data'], **v['allele frequencies'], **{\"no_pathogenic_coocurrs\":pathogenic_count}}\n",
    "    flat_vus.append(z)\n",
    "\n",
    "# turn the array of dicts into a data frame    \n",
    "import pandas as pd\n",
    "flat_df = pd.DataFrame(flat_vus)\n",
    "flat_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vus</th>\n",
       "      <th>count</th>\n",
       "      <th>maxPop</th>\n",
       "      <th>maxPopFreq</th>\n",
       "      <th>cohortFreq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(13, 32318080, 'C', 'T')</td>\n",
       "      <td>9765</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.955854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(13, 32318598, 'T', 'C')</td>\n",
       "      <td>4653</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.455462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(13, 32319654, 'A', 'G')</td>\n",
       "      <td>3060</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.299530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(13, 32321240, 'G', 'C')</td>\n",
       "      <td>9766</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.955951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(13, 32325741, 'C', 'T')</td>\n",
       "      <td>9762</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.955560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549</th>\n",
       "      <td>(13, 32360422, 'C', 'T')</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.000098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550</th>\n",
       "      <td>(13, 32378219, 'T', 'C')</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.000098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>551</th>\n",
       "      <td>(13, 32391658, 'A', 'T')</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.000098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552</th>\n",
       "      <td>(13, 32365109, 'C', 'CTT')</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.000098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553</th>\n",
       "      <td>(13, 32390265, 'G', 'T')</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.000098</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>554 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            vus  count maxPop maxPopFreq  cohortFreq\n",
       "0      (13, 32318080, 'C', 'T')   9765   None       None    0.955854\n",
       "1      (13, 32318598, 'T', 'C')   4653   None       None    0.455462\n",
       "2      (13, 32319654, 'A', 'G')   3060   None       None    0.299530\n",
       "3      (13, 32321240, 'G', 'C')   9766   None       None    0.955951\n",
       "4      (13, 32325741, 'C', 'T')   9762   None       None    0.955560\n",
       "..                          ...    ...    ...        ...         ...\n",
       "549    (13, 32360422, 'C', 'T')      1   None       None    0.000098\n",
       "550    (13, 32378219, 'T', 'C')      1   None       None    0.000098\n",
       "551    (13, 32391658, 'A', 'T')      1   None       None    0.000098\n",
       "552  (13, 32365109, 'C', 'CTT')      1   None       None    0.000098\n",
       "553    (13, 32390265, 'G', 'T')      1   None       None    0.000098\n",
       "\n",
       "[554 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# homozygous vus output\n",
    "homozygous_vus = []\n",
    "for k, v in data['homozygous vus'].items():\n",
    "    ## this is a pythonic way of merging dicts - it is cryptic\n",
    "    z = {**{\"vus\":k}, **v}\n",
    "    homozygous_vus.append(z)\n",
    "\n",
    "# turn the array of dicts into a data frame    \n",
    "import pandas as pd\n",
    "hz_df = pd.DataFrame(homozygous_vus)\n",
    "hz_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## To do\n",
    "- Submit the pathogenicity file from the local system\n",
    "- Either  access the gnomad file from Gnomad, or supply it from the local system\n",
    "\n",
    "## Done\n",
    "- Make the container available to other WES servers by adding the Docker container to Docker Hub instead of the Seven Bridges docker repository\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
