{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## Running the VUS cooccurrence workflow on TCGA germline variants\n",
    "This notebook runs the cooccurrence on a real dataset and with a comprehensive set of variants for BRCA2 obtained from ClinVar.  \n",
    "\n",
    "This paper described producing vcfs of germline mutations across all TCGA projects\n",
    "https://gdc.cancer.gov/about-data/publications/PanCanAtlas-Germline-AWG\n",
    "The vcfs are available in the Genomic Data Commons.\n",
    "\n",
    "The TCGA data uses build 37. So we need to produce \n",
    "a pathogencity file\n",
    "a gnomad file based on build 37. This is Gnomad release 2.1.1. Another notebook describes using BCFTools to filter the Gnomad data. See SBCGC_WES_bcftools."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/forei/imfenvs/fasp/lib/python3.9/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.\n",
      "  warnings.warn(msg)\n"
     ]
    }
   ],
   "source": [
    "from fasp.workflow import sbcgcWESClient\n",
    "\n",
    "cl = sbcgcWESClient('forei/fasp-vus', debug=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setting up the WES run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    \"project\": \"forei/fasp-vus\",\n",
    "    \"inputs\": {\n",
    "        'save_files': 'True',\n",
    "        'ensembl_release': '75',\n",
    "        'nomad_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617e05ebd1cbfe46094811ba',\n",
    "                        'name': 'gnomad_grch37_chr13_brca2.vcf',\n",
    "                        'class': 'File'},\n",
    "\n",
    "        'cooccurence_app': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f288d9c5e581c1b2032b7',\n",
    "                        'name': 'cooccurrenceFinder7.py',\n",
    "                        'class': 'File'},\n",
    "\n",
    "        'hg_version': '37',\n",
    "        'vcf_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617dede89c5e581c1b200dfa',\n",
    "                     'name': '_1_PCA.r1.TCGAbarcode.merge.tnSwapCorrected.10389.filtered.vcf',\n",
    "                     'class': 'File'},\n",
    "\n",
    "        'pathogenicity_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617c5ce6e6261a31b6d12ddc',\n",
    "                               'name': 'clinvar_BRCA2_full.txt',\n",
    "                               'class': 'File'}\n",
    "\n",
    "        }\n",
    "    }\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have formulated the body in the way that it can be passed to a client function as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'75ccf738-def9-4e89-b585-9de64b6b65f2'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import json\n",
    "run_id= cl.runGenericWorkflow(\n",
    "    workflow_url=\"sbg://forei/fasp-vus/cooccurence/21\",\n",
    "    workflow_params = json.dumps(params),\n",
    "    workflow_type = \"CWL\",\n",
    "    workflow_type_version = \"sbg:draft-2\"\n",
    ")\n",
    "run_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COMPLETE\n",
      "0:08:51\n"
     ]
    }
   ],
   "source": [
    "import dateutil.parser\n",
    "print(cl.getTaskStatus(run_id))\n",
    "log = cl.GetRunLog(run_id)\n",
    "if log['run_log']['start_time']:\n",
    "    start = dateutil.parser.isoparse(log['run_log']['start_time'])\n",
    "    end = dateutil.parser.isoparse(log['run_log']['end_time'])\n",
    "    duration = end - start\n",
    "    print(str(duration))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'request': {'tags': {},\n",
       "  'workflow_params': {'name': 'cooccurrence run - 10-31-21 23:55:30',\n",
       "   'project': 'forei/fasp-vus',\n",
       "   'inputs': {'p2': None,\n",
       "    'save_files': 'True',\n",
       "    'gene': None,\n",
       "    'chromosome': None,\n",
       "    'pathology_file': None,\n",
       "    'nomad_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617e05ebd1cbfe46094811ba',\n",
       "     'basename': 'gnomad_grch37_chr13_brca2.vcf',\n",
       "     'nameext': '.vcf',\n",
       "     'class': 'File',\n",
       "     'nameroot': 'gnomad_grch37_chr13_brca2'},\n",
       "    'vcf_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617dede89c5e581c1b200dfa',\n",
       "     'basename': '_1_PCA.r1.TCGAbarcode.merge.tnSwapCorrected.10389.filtered.vcf',\n",
       "     'nameext': '.vcf',\n",
       "     'class': 'File',\n",
       "     'nameroot': '_1_PCA.r1.TCGAbarcode.merge.tnSwapCorrected.10389.filtered'},\n",
       "    'phased': None,\n",
       "    'data_directory': None,\n",
       "    'ensembl_release': '75',\n",
       "    'hg_version': '37',\n",
       "    'pathogenicity_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f27419c5e581c1b2032b1',\n",
       "     'basename': 'clinvar_BRCA2_full.txt',\n",
       "     'nameext': '.txt',\n",
       "     'class': 'File',\n",
       "     'nameroot': 'clinvar_BRCA2_full'},\n",
       "    'cooccurence_app': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f288d9c5e581c1b2032b7',\n",
       "     'basename': 'cooccurrenceFinder7.py',\n",
       "     'nameext': '.py',\n",
       "     'class': 'File',\n",
       "     'nameroot': 'cooccurrenceFinder7'}}},\n",
       "  'workflow_type': 'CWL',\n",
       "  'workflow_engine_params': {},\n",
       "  'workflow_url': 'sbg://forei/fasp-vus/cooccurence/21'},\n",
       " 'state': 'COMPLETE',\n",
       " 'outputs': {'vpi_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483051',\n",
       "   'basename': '_12_BRCA2-vpi.json',\n",
       "   'nameext': '.json',\n",
       "   'class': 'File',\n",
       "   'nameroot': '_12_BRCA2-vpi'},\n",
       "  'pathology_output': None,\n",
       "  'all_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483053',\n",
       "   'basename': '_12_BRCA2-all.json',\n",
       "   'nameext': '.json',\n",
       "   'class': 'File',\n",
       "   'nameroot': '_12_BRCA2-all'},\n",
       "  'out_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483055',\n",
       "   'basename': '_15_BRCA2-cooccurrences.json',\n",
       "   'nameext': '.json',\n",
       "   'class': 'File',\n",
       "   'nameroot': '_15_BRCA2-cooccurrences'},\n",
       "  'ipv_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483059',\n",
       "   'basename': '_12_BRCA2-ipv.json',\n",
       "   'nameext': '.json',\n",
       "   'class': 'File',\n",
       "   'nameroot': '_12_BRCA2-ipv'},\n",
       "  'tout_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483057',\n",
       "   'basename': '_12_BRCA2-tout.json',\n",
       "   'nameext': '.json',\n",
       "   'class': 'File',\n",
       "   'nameroot': '_12_BRCA2-tout'}},\n",
       " 'run_id': '75ccf738-def9-4e89-b585-9de64b6b65f2',\n",
       " 'run_log': {'name': 'cooccurrence run - 10-31-21 23:55:30',\n",
       "  'cmd': None,\n",
       "  'start_time': '2021-10-31T23:55:32Z',\n",
       "  'end_time': '2021-11-01T00:04:23Z',\n",
       "  'stdout': None,\n",
       "  'stderr': None,\n",
       "  'exit_code': 0},\n",
       " 'task_logs': [{'name': 'cooccurence',\n",
       "   'cmd': '/usr/bin/python3 /sbgenomics/Projects/df88aa55-c90f-47f3-8073-45fc61166e15/cooccurrenceFinder7.py --g BRCA2 --vcf /sbgenomics/Projects/df88aa55-c90f-47f3-8073-45fc61166e15/_1_PCA.r1.TCGAbarcode.merge.tnSwapCorrected.10389.filtered.vcf --h 37 --e 75 --c 13 --p True --p2 0.001 --vpf /sbgenomics/Projects/df88aa55-c90f-47f3-8073-45fc61166e15/clinvar_BRCA2_full.txt --d /var/tmp/pyensembl-cache --save True --gf /sbgenomics/Projects/df88aa55-c90f-47f3-8073-45fc61166e15/gnomad_grch37_chr13_brca2.vcf',\n",
       "   'start_time': '2021-10-31T23:58:46Z',\n",
       "   'end_time': '2021-11-01T00:04:21Z',\n",
       "   'stdout': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f039c5e581c1b2032f5',\n",
       "   'stderr': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f039c5e581c1b2032f5',\n",
       "   'exit_code': 0}]}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.GetRunLog(run_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting the results - via DRS\n",
    "Once the run is complete, further steps can use DRS to obtain the file output from the workflow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'vpi_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483051',\n",
       "  'basename': '_12_BRCA2-vpi.json',\n",
       "  'nameext': '.json',\n",
       "  'class': 'File',\n",
       "  'nameroot': '_12_BRCA2-vpi'},\n",
       " 'pathology_output': None,\n",
       " 'all_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483053',\n",
       "  'basename': '_12_BRCA2-all.json',\n",
       "  'nameext': '.json',\n",
       "  'class': 'File',\n",
       "  'nameroot': '_12_BRCA2-all'},\n",
       " 'out_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483055',\n",
       "  'basename': '_15_BRCA2-cooccurrences.json',\n",
       "  'nameext': '.json',\n",
       "  'class': 'File',\n",
       "  'nameroot': '_15_BRCA2-cooccurrences'},\n",
       " 'ipv_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483059',\n",
       "  'basename': '_12_BRCA2-ipv.json',\n",
       "  'nameext': '.json',\n",
       "  'class': 'File',\n",
       "  'nameroot': '_12_BRCA2-ipv'},\n",
       " 'tout_file': {'path': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483057',\n",
       "  'basename': '_12_BRCA2-tout.json',\n",
       "  'nameext': '.json',\n",
       "  'class': 'File',\n",
       "  'nameroot': '_12_BRCA2-tout'}}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "runLog = cl.GetRunLog(run_id)\n",
    "runLog['outputs']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483055'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resultsDRSID = runLog['outputs']['out_file']['path']\n",
    "resultsDRSID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the CGC DRS Server to retrieve the results files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from fasp.loc import sbcgcDRSClient\n",
    "drsClient = sbcgcDRSClient('~/.keys/sevenbridges_keys.json', 's3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DRS GetObject\n",
    "Here's how we then get details of the file. Note that here only the id portion of the DRS URI is being passed. It is the job of a metaresolver to look at that URI and to determine where to send the id. As noted, we are passing up on the opportunity to use a metaresolver and putting in the id manually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': '617f2f06d1cbfe4609483055',\n",
       " 'name': '_15_BRCA2-cooccurrences.json',\n",
       " 'size': 38886,\n",
       " 'checksums': [{'type': 'etag',\n",
       "   'checksum': '774cd65cc59dbb738843e2b40fa81f45-1'}],\n",
       " 'self_uri': 'drs://cgc-ga4gh-api.sbgenomics.com/617f2f06d1cbfe4609483055',\n",
       " 'created_time': '2021-11-01T00:04:22Z',\n",
       " 'updated_time': '2021-11-01T00:04:22Z',\n",
       " 'mime_type': 'application/json',\n",
       " 'access_methods': [{'type': 's3',\n",
       "   'region': 'us-east-1',\n",
       "   'access_id': 'aws-us-east-1'}]}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sbDRSID = resultsDRSID.split('/')[-1]\n",
    "fileDetails = drsClient.getObject(sbDRSID)\n",
    "fileDetails"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = drsClient.getAccessURL(sbDRSID,'s3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Downloading the file\n",
    "Now we can use the url obtained to download the file. We'll create a small function to encapsulate the download."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import os\n",
    "def download(url, file_path):\n",
    "    with open(os.path.expanduser(file_path), \"wb\") as file:\n",
    "        response = requests.get(url)\n",
    "        file.write(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "fullPath = fileDetails['name']\n",
    "download(url, fullPath)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(fullPath) as json_file:\n",
    "    data = json.load(json_file)\n",
    "# delete the local copy of the file\n",
    "#os.remove(fullPath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vus</th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>n</th>\n",
       "      <th>k</th>\n",
       "      <th>likelihood</th>\n",
       "      <th>maxPop</th>\n",
       "      <th>maxPopFreq</th>\n",
       "      <th>cohortFreq</th>\n",
       "      <th>no_pathogenic_coocurrs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(13, 32890572, 'G', 'A')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>4541</td>\n",
       "      <td>1</td>\n",
       "      <td>0.145945</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.437097</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(13, 32900933, 'T', 'A')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>1022</td>\n",
       "      <td>1</td>\n",
       "      <td>1.507573</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.098373</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(13, 32905265, 'G', 'A')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>6165</td>\n",
       "      <td>6</td>\n",
       "      <td>11.485790</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.593416</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(13, 32911888, 'A', 'G')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>5369</td>\n",
       "      <td>1</td>\n",
       "      <td>0.084252</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.516797</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(13, 32913055, 'A', 'G')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10385</td>\n",
       "      <td>7</td>\n",
       "      <td>2.074215</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.999615</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>(13, 32915005, 'G', 'C')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10382</td>\n",
       "      <td>7</td>\n",
       "      <td>2.078348</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.999326</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>(13, 32915410, 'CAATT', 'C')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>5233</td>\n",
       "      <td>1</td>\n",
       "      <td>0.092209</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.503706</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>(13, 32920844, 'T', 'C')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>6657</td>\n",
       "      <td>5</td>\n",
       "      <td>2.789874</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.640774</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>(13, 32929232, 'A', 'G')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>4120</td>\n",
       "      <td>1</td>\n",
       "      <td>0.192979</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.396573</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>(13, 32929387, 'T', 'C')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10386</td>\n",
       "      <td>7</td>\n",
       "      <td>2.072839</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.999711</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>(13, 32936646, 'T', 'C')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>7950</td>\n",
       "      <td>3</td>\n",
       "      <td>0.134087</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.765232</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>(13, 32953388, 'T', 'C')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>6668</td>\n",
       "      <td>3</td>\n",
       "      <td>0.313926</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.641833</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>(13, 32959199, 'C', 'T')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>5919</td>\n",
       "      <td>1</td>\n",
       "      <td>0.058490</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.569737</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>(13, 32959534, 'A', 'G')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>3592</td>\n",
       "      <td>1</td>\n",
       "      <td>0.273947</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.345750</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>(13, 32971425, 'T', 'C')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>1967</td>\n",
       "      <td>2</td>\n",
       "      <td>2.391919</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.189335</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>(13, 32959839, 'A', 'C')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>4997</td>\n",
       "      <td>1</td>\n",
       "      <td>0.107840</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.480990</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>(13, 32903685, 'C', 'T')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>3001</td>\n",
       "      <td>1</td>\n",
       "      <td>0.405489</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.288863</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>(13, 32910328, 'T', 'C')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>2941</td>\n",
       "      <td>1</td>\n",
       "      <td>0.421958</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.283088</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>(13, 32930936, 'G', 'A')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>3019</td>\n",
       "      <td>3</td>\n",
       "      <td>3.534920</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.290596</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>(13, 32958649, 'G', 'A')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>485</td>\n",
       "      <td>1</td>\n",
       "      <td>2.152924</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.046684</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>(13, 32968607, 'A', 'G')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>1651</td>\n",
       "      <td>1</td>\n",
       "      <td>0.993153</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.158918</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>(13, 32918941, 'A', 'G')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>1301</td>\n",
       "      <td>1</td>\n",
       "      <td>1.252788</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.125229</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>(13, 32918967, 'G', 'A')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>982</td>\n",
       "      <td>1</td>\n",
       "      <td>1.548123</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.094523</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>(13, 32968591, 'G', 'A')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>1383</td>\n",
       "      <td>1</td>\n",
       "      <td>1.186443</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.133122</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>(13, 32962045, 'A', 'C')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>204</td>\n",
       "      <td>1</td>\n",
       "      <td>2.594216</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.019636</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>(13, 32962088, 'C', 'T')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>144</td>\n",
       "      <td>1</td>\n",
       "      <td>2.699582</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.013861</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>(13, 32893791, 'A', 'G')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>61</td>\n",
       "      <td>1</td>\n",
       "      <td>2.852431</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.005872</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>(13, 32961189, 'G', 'GC')</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>0.001</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "      <td>2.635860</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.017326</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             vus        p1     p2      n  k  likelihood  \\\n",
       "0       (13, 32890572, 'G', 'A')  0.000337  0.001   4541  1    0.145945   \n",
       "1       (13, 32900933, 'T', 'A')  0.000337  0.001   1022  1    1.507573   \n",
       "2       (13, 32905265, 'G', 'A')  0.000337  0.001   6165  6   11.485790   \n",
       "3       (13, 32911888, 'A', 'G')  0.000337  0.001   5369  1    0.084252   \n",
       "4       (13, 32913055, 'A', 'G')  0.000337  0.001  10385  7    2.074215   \n",
       "5       (13, 32915005, 'G', 'C')  0.000337  0.001  10382  7    2.078348   \n",
       "6   (13, 32915410, 'CAATT', 'C')  0.000337  0.001   5233  1    0.092209   \n",
       "7       (13, 32920844, 'T', 'C')  0.000337  0.001   6657  5    2.789874   \n",
       "8       (13, 32929232, 'A', 'G')  0.000337  0.001   4120  1    0.192979   \n",
       "9       (13, 32929387, 'T', 'C')  0.000337  0.001  10386  7    2.072839   \n",
       "10      (13, 32936646, 'T', 'C')  0.000337  0.001   7950  3    0.134087   \n",
       "11      (13, 32953388, 'T', 'C')  0.000337  0.001   6668  3    0.313926   \n",
       "12      (13, 32959199, 'C', 'T')  0.000337  0.001   5919  1    0.058490   \n",
       "13      (13, 32959534, 'A', 'G')  0.000337  0.001   3592  1    0.273947   \n",
       "14      (13, 32971425, 'T', 'C')  0.000337  0.001   1967  2    2.391919   \n",
       "15      (13, 32959839, 'A', 'C')  0.000337  0.001   4997  1    0.107840   \n",
       "16      (13, 32903685, 'C', 'T')  0.000337  0.001   3001  1    0.405489   \n",
       "17      (13, 32910328, 'T', 'C')  0.000337  0.001   2941  1    0.421958   \n",
       "18      (13, 32930936, 'G', 'A')  0.000337  0.001   3019  3    3.534920   \n",
       "19      (13, 32958649, 'G', 'A')  0.000337  0.001    485  1    2.152924   \n",
       "20      (13, 32968607, 'A', 'G')  0.000337  0.001   1651  1    0.993153   \n",
       "21      (13, 32918941, 'A', 'G')  0.000337  0.001   1301  1    1.252788   \n",
       "22      (13, 32918967, 'G', 'A')  0.000337  0.001    982  1    1.548123   \n",
       "23      (13, 32968591, 'G', 'A')  0.000337  0.001   1383  1    1.186443   \n",
       "24      (13, 32962045, 'A', 'C')  0.000337  0.001    204  1    2.594216   \n",
       "25      (13, 32962088, 'C', 'T')  0.000337  0.001    144  1    2.699582   \n",
       "26      (13, 32893791, 'A', 'G')  0.000337  0.001     61  1    2.852431   \n",
       "27     (13, 32961189, 'G', 'GC')  0.000337  0.001    180  1    2.635860   \n",
       "\n",
       "   maxPop maxPopFreq  cohortFreq  no_pathogenic_coocurrs  \n",
       "0    None       None    0.437097                       1  \n",
       "1    None       None    0.098373                       1  \n",
       "2    None       None    0.593416                       6  \n",
       "3    None       None    0.516797                       1  \n",
       "4    None       None    0.999615                       7  \n",
       "5    None       None    0.999326                       7  \n",
       "6    None       None    0.503706                       1  \n",
       "7    None       None    0.640774                       5  \n",
       "8    None       None    0.396573                       1  \n",
       "9    None       None    0.999711                       7  \n",
       "10   None       None    0.765232                       3  \n",
       "11   None       None    0.641833                       3  \n",
       "12   None       None    0.569737                       1  \n",
       "13   None       None    0.345750                       1  \n",
       "14   None       None    0.189335                       2  \n",
       "15   None       None    0.480990                       1  \n",
       "16   None       None    0.288863                       1  \n",
       "17   None       None    0.283088                       1  \n",
       "18   None       None    0.290596                       3  \n",
       "19   None       None    0.046684                       1  \n",
       "20   None       None    0.158918                       1  \n",
       "21   None       None    0.125229                       1  \n",
       "22   None       None    0.094523                       1  \n",
       "23   None       None    0.133122                       1  \n",
       "24   None       None    0.019636                       1  \n",
       "25   None       None    0.013861                       1  \n",
       "26   None       None    0.005872                       1  \n",
       "27   None       None    0.017326                       1  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# flatten cooccurrence output\n",
    "flat_vus = []\n",
    "for k, v in data['cooccurring vus'].items():\n",
    "    pathogenic_count = len(v['pathogenic variants'])\n",
    "    ## this is a pythonic way of merging dicts - it is cryptic\n",
    "    z = {**{\"vus\":k}, **v['likelihood data'], **v['allele frequencies'], **{\"no_pathogenic_coocurrs\":pathogenic_count}}\n",
    "    flat_vus.append(z)\n",
    "\n",
    "# turn the array of dicts into a data frame    \n",
    "import pandas as pd\n",
    "flat_df = pd.DataFrame(flat_vus)\n",
    "flat_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vus</th>\n",
       "      <th>count</th>\n",
       "      <th>maxPop</th>\n",
       "      <th>maxPopFreq</th>\n",
       "      <th>cohortFreq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(13, 32890572, 'G', 'A')</td>\n",
       "      <td>663</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.063817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(13, 32900933, 'T', 'A')</td>\n",
       "      <td>367</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.035326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(13, 32905265, 'G', 'A')</td>\n",
       "      <td>6098</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.586967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(13, 32911888, 'A', 'G')</td>\n",
       "      <td>968</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.093175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(13, 32913055, 'A', 'G')</td>\n",
       "      <td>10239</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.985562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>(13, 32893724, 'G', 'A')</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.000096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>(13, 32965764, 'C', 'T')</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.000096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>(13, 32905262, 'C', 'G')</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.000096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>(13, 32942101, 'C', 'T')</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.000096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>(13, 32950555, 'G', 'C')</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.000096</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>258 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          vus  count maxPop maxPopFreq  cohortFreq\n",
       "0    (13, 32890572, 'G', 'A')    663   None       None    0.063817\n",
       "1    (13, 32900933, 'T', 'A')    367   None       None    0.035326\n",
       "2    (13, 32905265, 'G', 'A')   6098   None       None    0.586967\n",
       "3    (13, 32911888, 'A', 'G')    968   None       None    0.093175\n",
       "4    (13, 32913055, 'A', 'G')  10239   None       None    0.985562\n",
       "..                        ...    ...    ...        ...         ...\n",
       "253  (13, 32893724, 'G', 'A')      1   None       None    0.000096\n",
       "254  (13, 32965764, 'C', 'T')      1   None       None    0.000096\n",
       "255  (13, 32905262, 'C', 'G')      1   None       None    0.000096\n",
       "256  (13, 32942101, 'C', 'T')      1   None       None    0.000096\n",
       "257  (13, 32950555, 'G', 'C')      1   None       None    0.000096\n",
       "\n",
       "[258 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# homozygous vus output\n",
    "homozygous_vus = []\n",
    "for k, v in data['homozygous vus'].items():\n",
    "    ## this is a pythonic way of merging dicts - it is cryptic\n",
    "    z = {**{\"vus\":k}, **v}\n",
    "    homozygous_vus.append(z)\n",
    "\n",
    "# turn the array of dicts into a data frame    \n",
    "import pandas as pd\n",
    "hz_df = pd.DataFrame(homozygous_vus)\n",
    "hz_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## To do\n",
    "- Submit the pathogenicity file from the local system\n",
    "- Either  access the gnomad file from Gnomad, or supply it from the local system\n",
    "\n",
    "## Done\n",
    "- Make the container available to other WES servers by adding the Docker container to Docker Hub instead of the Seven Bridges docker repository\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
