{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running the VUS cooccurence workflow via WES on Cavatica\n",
    "This notebook tests runs the BRCA Challenge federated_analysis on the Kids First Osteosarcoma study on Cavatica.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/forei/imfenvs/fasp/lib/python3.9/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.\n",
      "  warnings.warn(msg)\n"
     ]
    }
   ],
   "source": [
    "from fasp.workflow import cavaticaWESClient, sb\n",
    "\n",
    "cl = cavaticaWESClient('forei/fasp-vus', debug=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setting up the WES run\n",
    "This uses the DRS ids of each file to specify the inputs to the cooccurrence analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "project = 'forei/fasp-vus'\n",
    "instances = { \"KF\":  {\n",
    "    \"wesClient\": cavaticaWESClient(project)\n",
    "    \"worklow_url\":\"sbg://forei/fasp-vus/cooccurrence/2\",\n",
    "    'vcf_file': {\n",
    "        'path': 'drs://cavatica-ga4gh-api.sbgenomics.com/61848448d2f88b031ebde426',\n",
    "        'name': 'ab010347-8ae3-4c36-833b-9743e6157524.CGP.filtered.deNovo.vep.merged.vcf.gz'},\n",
    "    'pathogenicity_file': {\n",
    "        'path': 'drs://cavatica-ga4gh-api.sbgenomics.com/618475b3d2f88b031ebd8de4',\n",
    "        'name': 'clinvar_BRCA2_mixed.tsv'}\n",
    "    },\n",
    "\n",
    "\"bdc\":  {\n",
    "    \"wesClient\": sbWESClient('https://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/ga4gh/wes/v1', project,\n",
    "                     '~/.keys/sbbdc_key.json')\n",
    "    \"worklow_url\":\"sbg://forei/fasp-vus/cooccurrence/2\",\n",
    "    'vcf_file': {\n",
    "        'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617c77bce6261a31b6d12f0a',\n",
    "        'name': 'COPDGene_phs000951_TOPMed_WGS_freeze.8.chr13.hg38.c1.filtered.vcf'},\n",
    "    'pathogenicity_file': {\n",
    "        'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617c5ce6e6261a31b6d12ddc',\n",
    "        'name': 'clinvar_brca2.tsv'}\n",
    "    }\n",
    "\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "unmatched '}' (3533571355.py, line 8)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"/var/folders/wz/jjbxsnr13v7dkw2jkbpmqd_dly65gq/T/ipykernel_85105/3533571355.py\"\u001b[0;36m, line \u001b[0;32m8\u001b[0m\n\u001b[0;31m    }\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m unmatched '}'\n"
     ]
    }
   ],
   "source": [
    "params = {\n",
    "    \"project\": \"forei/fasp-vus\",\n",
    "    \"inputs\": {\n",
    "        'vcf_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617c77bce6261a31b6d12f0a',\n",
    "                     'name': 'COPDGene_phs000951_TOPMed_WGS_freeze.8.chr13.hg38.c1.filtered.vcf',\n",
    "                     'class': 'File'},\n",
    "\n",
    "        'pathogenicity_file': {'path': 'drs://ga4gh-api.sb.biodatacatalyst.nhlbi.nih.gov/617c5ce6e6261a31b6d12ddc',\n",
    "                               'name': 'clinvar_brca2.tsv',\n",
    "                               'class': 'File'}\n",
    "        }\n",
    "    }\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    \"project\": \"forei/fasp-vus\",\n",
    "    \"inputs\": {\n",
    "    'vcf_file': {'path': 'drs://cavatica-ga4gh-api.sbgenomics.com/61848448d2f88b031ebde426',\n",
    "     'name': 'ab010347-8ae3-4c36-833b-9743e6157524.CGP.filtered.deNovo.vep.merged.vcf.gz',\n",
    "     'class': 'File'},\n",
    "    'pathogenicity_file': {'path': 'drs://cavatica-ga4gh-api.sbgenomics.com/618475b3d2f88b031ebd8de4',\n",
    "     'name': 'clinvar_BRCA2_mixed.tsv',\n",
    "     'class': 'File'}\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have formulated the body in the way that it can be passed to a client function as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'527f63d1-5a55-40bc-930d-bc7411e706a7'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import json\n",
    "run_id= cl.runGenericWorkflow(\n",
    "    workflow_url=\"sbg://forei/fasp-vus/cooccurrence/2\",\n",
    "    workflow_params = json.dumps(params),\n",
    "    workflow_type = \"CWL\",\n",
    "    workflow_type_version = \"sbg:draft-2\",\n",
    "    verbose=False\n",
    ")\n",
    "run_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check until complete (about 4 mins, your time may vary)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'COMPLETE'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl.getTaskStatus(run_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting the results - via DRS\n",
    "Once the run is complete, further steps can use DRS to obtain the file output from the workflow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'vpi_file': {'path': 'drs://cavatica-ga4gh-api.sbgenomics.com/6188354bd2f88b031ee29d83',\n",
       "  'name': '_4_BRCA2-vpi.json',\n",
       "  'class': 'File'},\n",
       " 'pathology_output': None,\n",
       " 'all_file': {'path': 'drs://cavatica-ga4gh-api.sbgenomics.com/6188354bd2f88b031ee29d8a',\n",
       "  'name': '_4_BRCA2-all.json',\n",
       "  'class': 'File'},\n",
       " 'out_file': {'path': 'drs://cavatica-ga4gh-api.sbgenomics.com/6188354bd2f88b031ee29d82',\n",
       "  'name': '_16_BRCA2-cooccurrences.json',\n",
       "  'class': 'File'},\n",
       " 'ipv_file': {'path': 'drs://cavatica-ga4gh-api.sbgenomics.com/6188354bd2f88b031ee29d84',\n",
       "  'name': '_4_BRCA2-ipv.json',\n",
       "  'class': 'File'},\n",
       " 'tout_file': {'path': 'drs://cavatica-ga4gh-api.sbgenomics.com/6188354bd2f88b031ee29d88',\n",
       "  'name': '_4_BRCA2-tout.json',\n",
       "  'class': 'File'}}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "runLog = cl.GetRunLog(run_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'drs://cavatica-ga4gh-api.sbgenomics.com/6188354bd2f88b031ee29d82'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resultsDRSID = runLog['outputs']['out_file']['path']\n",
    "resultsDRSID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the Cavatica DRS Server to retrieve the results files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from fasp.loc import cavaticaDRSClient\n",
    "drsClient = cavaticaDRSClient('~/.keys/sevenbridges_keys.json', 's3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DRS GetObject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': '6188354bd2f88b031ee29d82',\n",
       " 'name': '_16_BRCA2-cooccurrences.json',\n",
       " 'size': 21058,\n",
       " 'checksums': [{'type': 'etag',\n",
       "   'checksum': '7527cbd2250415cbc1d307c4fb7a1ad9-1'}],\n",
       " 'self_uri': 'drs://cavatica-ga4gh-api.sbgenomics.com/6188354bd2f88b031ee29d82',\n",
       " 'created_time': '2021-11-07T20:21:31Z',\n",
       " 'updated_time': '2021-11-07T20:21:31Z',\n",
       " 'mime_type': 'application/json',\n",
       " 'access_methods': [{'type': 's3',\n",
       "   'region': 'us-east-1',\n",
       "   'access_id': 'aws-us-east-1'}]}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sbDRSID = resultsDRSID.split('/')[-1]\n",
    "\n",
    "fileDetails = drsClient.getObject(sbDRSID)\n",
    "fileDetails"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = drsClient.getAccessURL(sbDRSID,'s3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Downloading the file\n",
    "Now we can use the url obtained to download the file. We'll create a small function to encapsulate the download."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import os\n",
    "def download(url, file_path):\n",
    "    with open(os.path.expanduser(file_path), \"wb\") as file:\n",
    "        response = requests.get(url)\n",
    "        file.write(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "fullPath =  fileDetails['name']\n",
    "download(url, fullPath)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cooccurring vus': {\"('13', 32315831, 'G', 'A')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 28,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.09701474531924571},\n",
       "   'allele frequencies': {'maxPop': nan,\n",
       "    'maxPopFreq': nan,\n",
       "    'cohortFreq': 0.8484848484848485},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32318080, 'C', 'T')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32318598, 'T', 'C')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 32,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10271219643185438},\n",
       "   'allele frequencies': {'maxPop': nan,\n",
       "    'maxPopFreq': nan,\n",
       "    'cohortFreq': 0.9696969696969697},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32321240, 'G', 'C')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32325741, 'C', 'T')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32331128, 'G', 'A')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32333969, 'A', 'G')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32338918, 'A', 'G')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32340868, 'G', 'C')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32342270, 'CAAA', 'CA')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 24,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.09163333212918408},\n",
       "   'allele frequencies': {'maxPop': nan,\n",
       "    'maxPopFreq': nan,\n",
       "    'cohortFreq': 0.7272727272727273},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32345879, 'G', 'A')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32346707, 'T', 'C')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32354180, 'T', 'C')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32355250, 'T', 'C')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32356799, 'G', 'A')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32359800, 'A', 'G')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32366208, 'C', 'A')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 25,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.09295003262470186},\n",
       "   'allele frequencies': {'maxPop': nan,\n",
       "    'maxPopFreq': nan,\n",
       "    'cohortFreq': 0.7575757575757576},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32371874, 'A', 'G')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 32,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10271219643185438},\n",
       "   'allele frequencies': {'maxPop': nan,\n",
       "    'maxPopFreq': nan,\n",
       "    'cohortFreq': 0.9696969696969697},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32372278, 'C', 'T')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32377590, 'CAAA', 'CAA')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 30,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.09982283094571447},\n",
       "   'allele frequencies': {'maxPop': nan,\n",
       "    'maxPopFreq': nan,\n",
       "    'cohortFreq': 0.9090909090909091},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32380604, 'T', 'A')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32383835, 'G', 'C')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 13,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.07832433946884151},\n",
       "   'allele frequencies': {'maxPop': nan,\n",
       "    'maxPopFreq': nan,\n",
       "    'cohortFreq': 0.3939393939393939},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32387052, 'G', 'GC')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32388121, 'CTT', 'CT')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 14,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.07944979997751686},\n",
       "   'allele frequencies': {'maxPop': nan,\n",
       "    'maxPopFreq': nan,\n",
       "    'cohortFreq': 0.42424242424242425},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32391514, 'T', 'C')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32397288, 'T', 'C')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32399885, 'C', 'A')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 33,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10418809168519826},\n",
       "   'allele frequencies': {'maxPop': nan, 'maxPopFreq': nan, 'cohortFreq': 1.0},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32400151, 'T', 'A')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 31,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.10125720823961136},\n",
       "   'allele frequencies': {'maxPop': nan,\n",
       "    'maxPopFreq': nan,\n",
       "    'cohortFreq': 0.9393939393939394},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]},\n",
       "  \"('13', 32359222, 'GAAA', 'GA')\": {'likelihood data': {'p1': 0.015151515151515152,\n",
       "    'p2': 0.001,\n",
       "    'n': 10,\n",
       "    'k': 1,\n",
       "    'likelihood': 0.07504271525921115},\n",
       "   'allele frequencies': {'maxPop': nan,\n",
       "    'maxPopFreq': nan,\n",
       "    'cohortFreq': 0.30303030303030304},\n",
       "   'pathogenic variants': [['13', 32376792, 'G', 'T']]}},\n",
       " 'homozygous vus': {\"('13', 32315831, 'G', 'A')\": {'count': 16,\n",
       "   'cohortFreq': 0.48484848484848486},\n",
       "  \"('13', 32318080, 'C', 'T')\": {'count': 29,\n",
       "   'cohortFreq': 0.8787878787878788},\n",
       "  \"('13', 32318598, 'T', 'C')\": {'count': 17,\n",
       "   'cohortFreq': 0.5151515151515151},\n",
       "  \"('13', 32321240, 'G', 'C')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32325741, 'C', 'T')\": {'count': 28,\n",
       "   'cohortFreq': 0.8484848484848485},\n",
       "  \"('13', 32331128, 'G', 'A')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32333969, 'A', 'G')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32338918, 'A', 'G')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32340868, 'G', 'C')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32342270, 'CAAA', 'CA')\": {'count': 16,\n",
       "   'cohortFreq': 0.48484848484848486},\n",
       "  \"('13', 32345879, 'G', 'A')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32346707, 'T', 'C')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32354180, 'T', 'C')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32355250, 'T', 'C')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32356799, 'G', 'A')\": {'count': 33, 'cohortFreq': 1.0},\n",
       "  \"('13', 32359800, 'A', 'G')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32371580, 'A', 'G')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32371874, 'A', 'G')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32372278, 'C', 'T')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32380604, 'T', 'A')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32385702, 'A', 'C')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32387052, 'G', 'GC')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32388121, 'CTT', 'CT')\": {'count': 11,\n",
       "   'cohortFreq': 0.3333333333333333},\n",
       "  \"('13', 32391514, 'T', 'C')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32392589, 'CAAA', 'CAA')\": {'count': 15,\n",
       "   'cohortFreq': 0.45454545454545453},\n",
       "  \"('13', 32397288, 'T', 'C')\": {'count': 30,\n",
       "   'cohortFreq': 0.9090909090909091},\n",
       "  \"('13', 32399885, 'C', 'A')\": {'count': 33, 'cohortFreq': 1.0},\n",
       "  \"('13', 32400151, 'T', 'A')\": {'count': 20,\n",
       "   'cohortFreq': 0.6060606060606061},\n",
       "  \"('13', 32358885, 'G', 'T')\": {'count': 5,\n",
       "   'cohortFreq': 0.15151515151515152},\n",
       "  \"('13', 32364618, 'C', 'CT')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32366868, 'CAA', 'CA')\": {'count': 6,\n",
       "   'cohortFreq': 0.18181818181818182},\n",
       "  \"('13', 32368798, 'GT', 'G')\": {'count': 6,\n",
       "   'cohortFreq': 0.18181818181818182},\n",
       "  \"('13', 32371795, 'C', 'T')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32377590, 'CAAA', 'CAA')\": {'count': 12,\n",
       "   'cohortFreq': 0.36363636363636365},\n",
       "  \"('13', 32385397, 'A', 'G')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32396449, 'A', 'G')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32323151, 'ATTTTTTTT', 'ATTTTTTTTT')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32388666, 'CT', 'CTT')\": {'count': 13,\n",
       "   'cohortFreq': 0.3939393939393939},\n",
       "  \"('13', 32315889, 'C', 'T')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32316090, 'G', 'T')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32316435, 'G', 'A')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32317661, 'T', 'C')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32318623, 'A', 'C')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32319654, 'A', 'G')\": {'count': 9, 'cohortFreq': 0.2727272727272727},\n",
       "  \"('13', 32325430, 'G', 'C')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32326796, 'T', 'A')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32328257, 'T', 'C')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32334210, 'C', 'T')\": {'count': 6,\n",
       "   'cohortFreq': 0.18181818181818182},\n",
       "  \"('13', 32334492, 'A', 'G')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32337751, 'A', 'G')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32341273, 'CAATT', 'C')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32341387, 'C', 'A')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32345500, 'T', 'G')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32347807, 'A', 'G')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32350510, 'G', 'C')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32353467, 'C', 'CAGAG')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32353470, 'A', 'C')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32353471, 'AAGC', 'A')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32354271, 'G', 'A')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32355095, 'A', 'G')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32359222, 'GAAA', 'GA')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32359789, 'A', 'G')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32360774, 'G', 'A')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32362509, 'T', 'C')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32364744, 'A', 'G')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32365177, 'C', 'T')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32366208, 'C', 'A')\": {'count': 8,\n",
       "   'cohortFreq': 0.24242424242424243},\n",
       "  \"('13', 32367579, 'A', 'AGATGGCTTG')\": {'count': 6,\n",
       "   'cohortFreq': 0.18181818181818182},\n",
       "  \"('13', 32368819, 'G', 'GT')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32369919, 'G', 'T')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32372668, 'G', 'A')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32372996, 'CT', 'CTT')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32375245, 'T', 'C')\": {'count': 8,\n",
       "   'cohortFreq': 0.24242424242424243},\n",
       "  \"('13', 32375563, 'CT', 'CTT')\": {'count': 5,\n",
       "   'cohortFreq': 0.15151515151515152},\n",
       "  \"('13', 32376120, 'A', 'G')\": {'count': 6,\n",
       "   'cohortFreq': 0.18181818181818182},\n",
       "  \"('13', 32376521, 'CAA', 'CAAA')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32376974, 'A', 'C')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32376978, 'C', 'T')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32377317, 'C', 'T')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32379251, 'T', 'C')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32380424, 'G', 'A')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32381260, 'C', 'T')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32383102, 'C', 'CTGTAATCCAGCACTTTGGGAGGCCAAGGTGGG')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32383682, 'G', 'A')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32383835, 'G', 'C')\": {'count': 5,\n",
       "   'cohortFreq': 0.15151515151515152},\n",
       "  \"('13', 32384329, 'T', 'G')\": {'count': 5,\n",
       "   'cohortFreq': 0.15151515151515152},\n",
       "  \"('13', 32384512, 'G', 'A')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32385062, 'C', 'T')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32386226, 'T', 'G')\": {'count': 8,\n",
       "   'cohortFreq': 0.24242424242424243},\n",
       "  \"('13', 32386567, 'AAAAACCATTTGTATGAAGCAGAAAAAAAATG', 'A')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32387908, 'A', 'C')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32387951, 'C', 'T')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32388181, 'T', 'C')\": {'count': 6,\n",
       "   'cohortFreq': 0.18181818181818182},\n",
       "  \"('13', 32389570, 'C', 'T')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32393329, 'A', 'G')\": {'count': 8,\n",
       "   'cohortFreq': 0.24242424242424243},\n",
       "  \"('13', 32394454, 'G', 'A')\": {'count': 6,\n",
       "   'cohortFreq': 0.18181818181818182},\n",
       "  \"('13', 32394470, 'A', 'G')\": {'count': 7,\n",
       "   'cohortFreq': 0.21212121212121213},\n",
       "  \"('13', 32395960, 'CTTT', 'CTTTTT')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32396599, 'G', 'C')\": {'count': 8,\n",
       "   'cohortFreq': 0.24242424242424243},\n",
       "  \"('13', 32397569, 'AT', 'A')\": {'count': 6,\n",
       "   'cohortFreq': 0.18181818181818182},\n",
       "  \"('13', 32399302, 'A', 'G')\": {'count': 3,\n",
       "   'cohortFreq': 0.09090909090909091},\n",
       "  \"('13', 32399599, 'ATT', 'AT')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32315226, 'G', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32329548, 'C', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32336191, 'T', 'C')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32343709, 'GA', 'G')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32344804, 'A', 'G')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32344830, 'G', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32345389, 'A', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32349322, 'C', 'CA')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32349553, 'C', 'T')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32351120, 'C', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32353757, 'C', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32353975, 'G', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32354065, 'A', 'G')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32354267, 'T', 'C')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32361269, 'T', 'C')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32365149, 'T', 'G')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32365721, 'CT', 'CTTTT')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32366751, 'C', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32368799, 'T', 'G')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32368810, 'G', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32369205, 'TAA', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32369961, 'G', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32371400, 'A', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32377459, 'A', 'C')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32379045, 'A', 'G')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32381972, 'G', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32382146, 'G', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32395894, 'T', 'C')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32399139, 'A', 'G')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32399786, 'CTT', 'CTTT')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32315655, 'A', 'G')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32317569, 'A', 'G')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32322709, 'G', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32329806, 'T', 'TTA')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32338162, 'T', 'C')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32342160, 'C', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32346481, 'G', 'A')\": {'count': 4,\n",
       "   'cohortFreq': 0.12121212121212122},\n",
       "  \"('13', 32347999, 'C', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32365165, 'G', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32368811, 'T', 'G')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32382620, 'G', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32385652, 'T', 'C')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32390367, 'C', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32391173, 'G', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32393237, 'TC', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32393672, 'C', 'G')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32395111, 'A', 'G')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32398875, 'A', 'C')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32330360, 'G', 'A')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32348194, 'A', 'C')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32350055, 'A', 'G')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32385099, 'C', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32385103, 'C', 'A')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32344166, 'GAA', 'GAAA')\": {'count': 2,\n",
       "   'cohortFreq': 0.06060606060606061},\n",
       "  \"('13', 32349814, 'CAAA', 'CA')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32317314, 'CAAGT', 'C')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32332592, 'A', 'C')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304},\n",
       "  \"('13', 32350747, 'C', 'T')\": {'count': 1,\n",
       "   'cohortFreq': 0.030303030303030304}}}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import json\n",
    "with open(fullPath) as json_file:\n",
    "    data = json.load(json_file)\n",
    "# delete the local copy of the file\n",
    "os.remove(fullPath)\n",
    "# and look at the contents\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vus</th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>n</th>\n",
       "      <th>k</th>\n",
       "      <th>likelihood</th>\n",
       "      <th>maxPop</th>\n",
       "      <th>maxPopFreq</th>\n",
       "      <th>cohortFreq</th>\n",
       "      <th>no_pathogenic_coocurrs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>('13', 32315831, 'G', 'A')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>0.097015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.848485</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>('13', 32318080, 'C', 'T')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>('13', 32318598, 'T', 'C')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>0.102712</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>('13', 32321240, 'G', 'C')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>('13', 32325741, 'C', 'T')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>('13', 32331128, 'G', 'A')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>('13', 32333969, 'A', 'G')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>('13', 32338918, 'A', 'G')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>('13', 32340868, 'G', 'C')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>('13', 32342270, 'CAAA', 'CA')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>0.091633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.727273</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>('13', 32345879, 'G', 'A')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>('13', 32346707, 'T', 'C')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>('13', 32354180, 'T', 'C')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>('13', 32355250, 'T', 'C')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>('13', 32356799, 'G', 'A')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>('13', 32359800, 'A', 'G')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>('13', 32366208, 'C', 'A')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "      <td>0.092950</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>('13', 32371874, 'A', 'G')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>0.102712</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.969697</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>('13', 32372278, 'C', 'T')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>('13', 32377590, 'CAAA', 'CAA')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "      <td>0.099823</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>('13', 32380604, 'T', 'A')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>('13', 32383835, 'G', 'C')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>0.078324</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.393939</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>('13', 32387052, 'G', 'GC')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>('13', 32388121, 'CTT', 'CT')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>0.079450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.424242</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>('13', 32391514, 'T', 'C')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>('13', 32397288, 'T', 'C')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>('13', 32399885, 'C', 'A')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.104188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>('13', 32400151, 'T', 'A')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "      <td>0.101257</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.939394</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>('13', 32359222, 'GAAA', 'GA')</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>0.075043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.303030</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                vus        p1     p2   n  k  likelihood  \\\n",
       "0        ('13', 32315831, 'G', 'A')  0.015152  0.001  28  1    0.097015   \n",
       "1        ('13', 32318080, 'C', 'T')  0.015152  0.001  33  1    0.104188   \n",
       "2        ('13', 32318598, 'T', 'C')  0.015152  0.001  32  1    0.102712   \n",
       "3        ('13', 32321240, 'G', 'C')  0.015152  0.001  33  1    0.104188   \n",
       "4        ('13', 32325741, 'C', 'T')  0.015152  0.001  33  1    0.104188   \n",
       "5        ('13', 32331128, 'G', 'A')  0.015152  0.001  33  1    0.104188   \n",
       "6        ('13', 32333969, 'A', 'G')  0.015152  0.001  33  1    0.104188   \n",
       "7        ('13', 32338918, 'A', 'G')  0.015152  0.001  33  1    0.104188   \n",
       "8        ('13', 32340868, 'G', 'C')  0.015152  0.001  33  1    0.104188   \n",
       "9    ('13', 32342270, 'CAAA', 'CA')  0.015152  0.001  24  1    0.091633   \n",
       "10       ('13', 32345879, 'G', 'A')  0.015152  0.001  33  1    0.104188   \n",
       "11       ('13', 32346707, 'T', 'C')  0.015152  0.001  33  1    0.104188   \n",
       "12       ('13', 32354180, 'T', 'C')  0.015152  0.001  33  1    0.104188   \n",
       "13       ('13', 32355250, 'T', 'C')  0.015152  0.001  33  1    0.104188   \n",
       "14       ('13', 32356799, 'G', 'A')  0.015152  0.001  33  1    0.104188   \n",
       "15       ('13', 32359800, 'A', 'G')  0.015152  0.001  33  1    0.104188   \n",
       "16       ('13', 32366208, 'C', 'A')  0.015152  0.001  25  1    0.092950   \n",
       "17       ('13', 32371874, 'A', 'G')  0.015152  0.001  32  1    0.102712   \n",
       "18       ('13', 32372278, 'C', 'T')  0.015152  0.001  33  1    0.104188   \n",
       "19  ('13', 32377590, 'CAAA', 'CAA')  0.015152  0.001  30  1    0.099823   \n",
       "20       ('13', 32380604, 'T', 'A')  0.015152  0.001  33  1    0.104188   \n",
       "21       ('13', 32383835, 'G', 'C')  0.015152  0.001  13  1    0.078324   \n",
       "22      ('13', 32387052, 'G', 'GC')  0.015152  0.001  33  1    0.104188   \n",
       "23    ('13', 32388121, 'CTT', 'CT')  0.015152  0.001  14  1    0.079450   \n",
       "24       ('13', 32391514, 'T', 'C')  0.015152  0.001  33  1    0.104188   \n",
       "25       ('13', 32397288, 'T', 'C')  0.015152  0.001  33  1    0.104188   \n",
       "26       ('13', 32399885, 'C', 'A')  0.015152  0.001  33  1    0.104188   \n",
       "27       ('13', 32400151, 'T', 'A')  0.015152  0.001  31  1    0.101257   \n",
       "28   ('13', 32359222, 'GAAA', 'GA')  0.015152  0.001  10  1    0.075043   \n",
       "\n",
       "    maxPop  maxPopFreq  cohortFreq  no_pathogenic_coocurrs  \n",
       "0      NaN         NaN    0.848485                       1  \n",
       "1      NaN         NaN    1.000000                       1  \n",
       "2      NaN         NaN    0.969697                       1  \n",
       "3      NaN         NaN    1.000000                       1  \n",
       "4      NaN         NaN    1.000000                       1  \n",
       "5      NaN         NaN    1.000000                       1  \n",
       "6      NaN         NaN    1.000000                       1  \n",
       "7      NaN         NaN    1.000000                       1  \n",
       "8      NaN         NaN    1.000000                       1  \n",
       "9      NaN         NaN    0.727273                       1  \n",
       "10     NaN         NaN    1.000000                       1  \n",
       "11     NaN         NaN    1.000000                       1  \n",
       "12     NaN         NaN    1.000000                       1  \n",
       "13     NaN         NaN    1.000000                       1  \n",
       "14     NaN         NaN    1.000000                       1  \n",
       "15     NaN         NaN    1.000000                       1  \n",
       "16     NaN         NaN    0.757576                       1  \n",
       "17     NaN         NaN    0.969697                       1  \n",
       "18     NaN         NaN    1.000000                       1  \n",
       "19     NaN         NaN    0.909091                       1  \n",
       "20     NaN         NaN    1.000000                       1  \n",
       "21     NaN         NaN    0.393939                       1  \n",
       "22     NaN         NaN    1.000000                       1  \n",
       "23     NaN         NaN    0.424242                       1  \n",
       "24     NaN         NaN    1.000000                       1  \n",
       "25     NaN         NaN    1.000000                       1  \n",
       "26     NaN         NaN    1.000000                       1  \n",
       "27     NaN         NaN    0.939394                       1  \n",
       "28     NaN         NaN    0.303030                       1  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# flatten cooccurrence output\n",
    "flat_vus = []\n",
    "for k, v in data['cooccurring vus'].items():\n",
    "    pathogenic_count = len(v['pathogenic variants'])\n",
    "    ## this is a pythonic way of merging dicts - it is cryptic\n",
    "    z = {**{\"vus\":k}, **v['likelihood data'], **v['allele frequencies'], **{\"no_pathogenic_coocurrs\":pathogenic_count}}\n",
    "    flat_vus.append(z)\n",
    "\n",
    "# turn the array of dicts into a data frame    \n",
    "import pandas as pd\n",
    "flat_df = pd.DataFrame(flat_vus)\n",
    "flat_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## To do\n",
    "- Submit the pathogenicity file from the local system\n",
    "- Either  access the gnomad file from Gnomad, or supply it from the local system\n",
    "\n",
    "### Done\n",
    "- Make the container available to other WES servers by adding the Docker container to Docker Hub instead of the Seven Bridges docker repository\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
